---
title: "Greedy immune factors model"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
library(rjson)
# because of NaN
aa = readr::read_file("~/project/deeplearning/icb/data/tcga/scrna.v2/tensorboardLog/20190717-195353/metrics_val_last_weights.json")
result <- fromJSON(gsub( aa, pattern="NaN", replacement = "null"))
result = sapply(result, function(tt) ifelse(is.null(tt), NA, tt))
result.dt = data.table(name=names(result), metric=unlist(result))
result.dt = result.dt[-seq(nrow(result.dt)-1, nrow(result.dt))]
result.dt = result.dt[order(metric)]
result.dt[,feat:=gsub(name, pattern = "_[0-9]+_correlation$",replacement = "")]
msi.feat = result.dt[grep(name, pattern="MSI")]$feat[-(1:2)]
neoantigen.feat = result.dt[grep(name, pattern="Neoantigens")]$feat
extra.feat = c("TIL.Regional.Fraction", "Intratumor.Heterogeneity", "Aneuploidy.Score", "Silent.Mutation.Rate", "Nonsilent.Mutation.Rate")
rescued.feat = c(msi.feat, neoantigen.feat, extra.feat)

# selected.phenotypes = c()

```

## expression association 



```{r}

write.dataset = function(output.dir, sample.name, dataset) {
    dir.create(output.dir)
    write.table(file=paste0(output.dir, "/samples_name.txt"),x = sample.name,
                row.names = F, col.names =T,  sep="\t", quote=F )
    write.table(file=paste0(output.dir, "/dataset.txt"),x = dataset,
                row.names = F, col.names =T,  sep="\t", quote=F )
    rand_inx = sample(nrow(dataset))
    dataset_shuffle = dataset[rand_inx,]
    train.inx = 1:ceiling(.85 * nrow(dataset_shuffle))
    val.inx = ceiling(.85 * nrow(dataset_shuffle)):nrow(dataset_shuffle)
    
    write.table(file=paste0(output.dir, "/dataset_train.txt"),x = dataset_shuffle[train.inx,],
                row.names = F, col.names =T,  sep="\t", quote=F )
    write.table(file=paste0(output.dir, "/dataset_val.txt"),x = dataset_shuffle[val.inx,],
                row.names = F, col.names =T,  sep="\t", quote=F )
    
}

dataset_ssgsea = "/liulab/asahu/data/ssgsea/xiaoman/TCGA_ALLTPM.txt"
# dataset_ssgsea = "/liulab/asahu/data/ssgsea/xiaoman/TCGA_gdc/EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv"
dataset_ssgsea = fread(dataset_ssgsea)
sample.name = fread("~/project/deeplearning/icb/data/tcga/scrna.v2/samples_name.txt")
expression_mat= t(as.matrix(dataset_ssgsea[,seq(2,ncol(dataset_ssgsea)),with=F]))
head(rownames(expression_mat))

colnames(expression_mat) = dataset_ssgsea$gene_name
patient.name = colnames(dataset_ssgsea)[-1]
patient.name = gsub(patient.name, pattern="-", replacement=".")
patient.name = substring(patient.name, 1, 16)
rownames(expression_mat) = patient.name
head(patient.name)
head(sample.name$x)
expression_match = expression_mat[sample.name$x,]
## select 20 genes per phenotype and take PC
rescued.pheno = tcga.dataset[,rescued.feat, with=F]
library(avinash)
cors.list = cor(rescued.pheno, expression_match, use = "pairwise.complete.obs")

top.cors = lapply(seq(nrow(cors.list)), function(tt) {
    xx = cors.list[tt,] 
    xx[order( abs(xx), decreasing = T)[1:50]] 
    }
)

names(top.cors)  = rownames(cors.list)


# save(file="/liulab/asahu/data/ssgsea/xiaoman/ICB_datasets/top_correlated_genes_with_ICB_biomarkers.RData", top.cors)
# save(file="/liulab/asahu/data/ssgsea/xiaoman/ICB_datasets/correlated_genes_with_ICB_biomarkers.RData", cors)

top.genes = unique(unlist(lapply(top.cors, names)))
# top.genes.extra = c("PMS2", "MSH6", "EPCAM", "MSH2")
# top.genes = unique(c(top.genes, top.genes.extra))
# setdiff(top.genes, common.genes)
top.expression = expression_match[,top.genes]


dataset.new = cbind(tcga.dataset, top.expression)
output.dir = "~/project/deeplearning/icb/data/tcga/scrna.v3/"
write.dataset(output.dir = output.dir, dataset = dataset.new, sample.name = sample.name)

which(colnames(dataset.new) %in% rescued.feat)

temp_out = get_sel_pca(top.expression, top.genes, scale=F)
ev = apply(temp_out$pca_out, 2, sd)^2
ev = ev/sum(ev)
cumsum(ev)

pca_sel_obj = temp_out$pca_obj
pca_sel_obj$len_selected = 10
aa1  = prcomp(top.expression)

save(file=paste0(output.dir, "/pca_sel_obj.RData"), pca_sel_obj, top.genes)
pca_top = temp_out$pca_out[,seq(pca_sel_obj$len_selected)]

```

## Immune genes
```{r}
T_Cell_extra = c("IL10", "IDO", "TGFB1", "TGFB2", "TGFBR1", "TGFBR1", "CD37", "TLR", "Arginase")
APC_2 = c("A2AR", "VISTA", "B7_h3", "PDL1", "PDL2", "CD80", "CD86", "Galectin_9", "Ox40L", "CD40", "B7RP1", "CD70", "HVEM", "GITRL", "TNFSF9", "CD155", "CD112")
T_Cell_1 = c("CTLA4", "TIM3", "OX40", "CD40L", "ICOS", "CD27", "BTLA", "LAG3", "TCR", "KIR", "GITR", "TNFRSF9", "CD226", "TIGIT")
checkpoint.genes = unique(c(T_Cell_extra, APC_2, T_Cell_1))
load("/liulab/asahu/data/ssgsea/xiaoman/getz/all.tcga.genes.RData")
setdiff(checkpoint.genes, all.genes)
checkpoint.genes.1 = intersect(checkpoint.genes, all.genes)

checkpoint.genes.rescue = c("IDO1", "IDO2", "ARG1", "ARG2", "ADORA2A", "ADORA1", "VSIR", "CD276", "VTCN1", "JAK2", "STAT3", "CD80", "ICOSLG", "ICOS", "PVR", "CD226", "HAVCR2", "CD4", "PRF1", "FOXP3", "CD28", "LCK", "B2M")
pd1.genes = grep("^PDCD1",  all.genes, value=T)
Galectin_9.genes = grep("^LGALS",  all.genes, value=T)
jak.genes =c("JAK1", "JAK2", "JAK3")
stat.genes = grep("^STAT[0-9]",  all.genes, value=T)
TNF.genes  =c(grep("^TNFR",  all.genes, value=T), grep("^TNFS",  all.genes, value=T))
il2.gene = c("IL2", "PTPN2", grep("^IL2R",  all.genes, value=T))
il7.gene = grep("^IL7",  all.genes, value=T)
il4.gene = grep("^IL7",  all.genes, value=T)
il6.gene = grep("^IL6",  all.genes, value=T)
il10.gene = grep("^IL10",  all.genes, value=T)
HAVC.gene =grep("HAVC",  all.genes, value=T)
gzm.genes  =grep("^GZM",  all.genes, value=T)
traf.genes = grep("^TRAF",  all.genes, value=T)
nfk.genes = grep("^NFK",  all.genes, value=T)
cd40.genes = grep("^CD40",  all.genes, value=T)
igh.genes = grep("^IGH",  all.genes, value=T)
cd3.genes = grep("^CD3[A-Z]*$",  all.genes, value=T)
tra.genes = grep("^TR[A-B][C,D,V]",  all.genes, value=T)
kir.genes = grep("^KIR",  all.genes, value=T)
tgf.genes =grep("^TGF",  all.genes, value=T)
antigen.presentation.genes = grep("^HLA",  all.genes, value=T)
traf.genes = grep("^TR[A-B]F",  all.genes, value=T)
serpin.genes = grep("^SERPINB[1-9]$",  all.genes, value=T)
vegf.genes = grep("^VEGF",  all.genes, value=T)
tap.genes = c("TAP1", "TAP2", "TAPBP")

checkpoint.genes.semifinal = unique(c(checkpoint.genes.1, checkpoint.genes.rescue, pd1.genes, Galectin_9.genes, stat.genes, TNF.genes, il2.gene, il7.gene, il4.gene, il6.gene, il10.gene, HAVC.gene, gzm.genes, traf.genes, nfk.genes, cd40.genes, igh.genes, cd3.genes,tra.genes, kir.genes, tgf.genes, antigen.presentation.genes, serpin.genes, vegf.genes))
 
immune.genes = unique(intersect(checkpoint.genes.semifinal, all.genes))                                    
nontranscriptome.immune.faactors = tcga.dataset[,c(102:103, 110:123, 128:137),with=F]                                     
tcga.immune.factors = tcga.dataset[,c(66:137, 178:188),with=F]
tcga.immune.genes = expression_match[,immune.genes]

```

### immune factors 

```{r}
immune.factors = tcga.dataset[,c(66:96, 100:137), with=F]
library(avinash)
cors.list = cor(immune.factors, expression_match, use = "pairwise.complete.obs")
aaa = data.table(rownames(cors.list), seq(nrow(cors.list)))
# aaa = aaa[V1 %in% c(genes.curr, paste0(genes.curr, ".ccf"))]
topmost.cors.genes = sapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[abs(xx) > 0.1]
    names(which.max(abs(xx)))
    }
)

top.cors = lapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[order( abs(xx), decreasing = T)[1:10]] 
    xx[abs(xx) >0.1]
    names(xx)
    }
)
if.topmost.cors = unique(unlist(topmost.cors.genes))
if.top.cors = unique(unlist(top.cors))

```


## Protein association
Check association with at least one immune fator.
```{r}
rrpa = fread("/liulab/asahu/data/ssgsea/xiaoman/TCGA_gdc/TCGA-RPPA-pancan-clean.txt")
rrpa[,SampleID:=gsub(substring(SampleID,1,16), pattern="-", replacement = ".")]
rrpa.match = rrpa[match(sample.name$x, SampleID)]
rrpa.mat = rrpa.match[,-(1:2),with=F]
temp = calc.cor(rrpa.mat, nontranscriptome.immune.factors, use = "pairwise.complete.obs")
# aa = plot.heatmap(temp$estimate[,-(21:22)])
pdf("~/liulab_home/data/ssgsea/xiaoman/TCGA_gdc/plots/top_if_rrpa.pdf",width=18, height = 7)
Heatmap(t(temp$estimate[,-(21:22)]), name = "top factors", column_title = "protien", row_title = "immune factors",
        row_names_gp = gpar(fontsize = 7), # Text size for row names
        column_names_gp = gpar(fontsize = 5) # Text size for row names
        )
dev.off()


temp2 = calc.cor(rrpa.mat, tcga.immune.factors, use = "pairwise.complete.obs")
pdf("~/liulab_home/data/ssgsea/xiaoman/TCGA_gdc/plots/if_rrpa.pdf",width=18, height = 7)
aa = temp2$estimate
aa = aa[which(apply(aa, 1, sd, na.rm=T) > 0), which(apply(aa, 2, sd, na.rm=T) > 0)] 
Heatmap(t(aa), name = "top factors", column_title = "factors", row_title = "protien",
        row_names_gp = gpar(fontsize = 7), # Text size for row names
        column_names_gp = gpar(fontsize = 5) # Text size for row names
        )
dev.off()

temp3 = calc.cor(rrpa.mat, tcga.immune.genes, use = "pairwise.complete.obs")
aa = temp3$estimate
aa = aa[which(apply(aa, 1, sd, na.rm=T) > 0), which(apply(aa, 2, sd, na.rm=T) > 0)] 
pdf("~/liulab_home/data/ssgsea/xiaoman/TCGA_gdc/plots/immune_genes_rrpa.pdf",width=18, height = 7)
Heatmap(t(aa), name = "Immune genes", column_title = "Protein", row_title = "Immune genes",
        row_names_gp = gpar(fontsize = 4), # Text size for row names
        column_names_gp = gpar(fontsize = 5) # Text size for row names
        )
dev.off()

protien.sel1 = which(rowSums(abs(temp$estimate)> 0.3 , na.rm = T) > 0)
protien.sel2 = which(rowSums(abs(temp2$estimate)> 0.4 , na.rm = T) > 0)
protien.sel3 = which(rowSums(abs(temp2$estimate)> 0.4 , na.rm = T) > 0)
protien.sel = unique(c(names(protien.sel1), names(protien.sel2), names(protien.sel3)))


rrpa.dataset = rrpa.mat[,protien.sel,with=F]
setnames(rrpa.dataset, seq(ncol(rrpa.dataset)), paste(colnames(rrpa.dataset), "prot", sep=".")) 

cors.list = cor(rrpa.dataset, expression_match, use = "pairwise.complete.obs")
cors.list = cors.list[colnames(rrpa.dataset),]
aaa = data.table(rownames(cors.list), seq(nrow(cors.list)))
topmost.cors.genes = sapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[abs(xx) > 0.15]
    names(which.max(abs(xx)))
    }
)

top.cors = lapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[order( abs(xx), decreasing = T)[1:10]] 
    xx[abs(xx) >0.15]
    names(xx)
    }
)
rrpa.topmost.cors = unique(unlist(topmost.cors.genes))
rrpa.top.cors = unique(unlist(top.cors))
```

## Mutation data(Havel et al. 2019) : 

 STK11, B2M, PTEN, PTPN2, APLNR, SWI/SNF (PBAF)
```{r}
mut.genes = c("STK11", "B2M", "PTEN", "PTPN2", "APLNR")
setdiff(mut.genes, all.genes)
baf.complex.genes=fread("/liulab/asahu/data/ssgsea/xiaoman/baf_complex.csv")$V2
baf.complex.genes =baf.complex.genes[-1]
all.mut.genes = unique(c(mut.genes, baf.complex.genes))

mutation = fread("/liulab/asahu/data/ssgsea/xiaoman/TCGA_gdc/TCGA_consolidated.abs_mafs_truncated.fixed.txt")
head(mutation$Hugo_Symbol)
# names(table(mutation.dt$Consequence))
synonymous.mut.type = c("downstream_gene_variant",   "intron_variant",       "stop_retained_variant",    "synonymous_variant",       "upstream_gene_variant")   
 
mutation[,SampleID:=gsub(substring(sample,1,16), pattern="-", replacement = ".")]
mut.samples = unique(mutation$SampleID)
mutation.dt = mutation[Hugo_Symbol %in% all.mut.genes]
mutation.dt = mutation.dt[!(Consequence %in% synonymous.mut.type)]
mutation.dt =mutation.dt[SampleID%in%sample.name$x]
mut.bin.mat = table(mutation.dt[,.( SampleID , Hugo_Symbol)])
mut.ccf.mat = xtabs(ccf_CI95_low~ SampleID + Hugo_Symbol, data=mutation.dt)
mut.cff.mat = mut.ccf.mat[,colnames(mut.bin.mat)]

mut.bin = matrix(NA, nrow=length(sample.name$x), ncol = ncol(mut.bin.mat), dimnames = list(sample.name$x, colnames(mut.bin.mat)))
mut.bin[intersect(mut.samples, sample.name$x),] =0
mut.bin[rownames(mut.bin.mat),] = mut.bin.mat

mut.ccf = matrix(NA, nrow=length(sample.name$x), ncol = ncol(mut.ccf.mat), dimnames = list(sample.name$x, colnames(mut.ccf.mat)))
mut.ccf[intersect(mut.samples, sample.name$x),] =0
mut.ccf[rownames(mut.ccf.mat),] = mut.ccf.mat
colnames(mut.ccf) = paste0(colnames(mut.ccf.mat), ".ccf")
mut.dataset = cbind( mut.ccf, mut.bin)
genes.curr1 = names(sort(colSums(mut.bin.mat),decreasing = T))[1:6]
genes.curr = unique(c(mut.genes, genes.curr1))

cors.list = cor(mut.dataset, expression_match, use = "pairwise.complete.obs")
rownames(cors.list) = colnames(mut.dataset)
aaa = data.table(rownames(cors.list), seq(nrow(cors.list)))
aaa = aaa[V1 %in% c(genes.curr, paste0(genes.curr, ".ccf"))]


topmost.cors.genes = sapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[abs(xx) > 0.1]
    names(which.max(abs(xx)))
    }
)

top.cors = lapply(aaa$V1, function(tt) {
    xx = cors.list[tt,] 
    xx = xx[order( abs(xx), decreasing = T)[1:10]] 
    xx[abs(xx) >0.1]
    names(xx)
    }
)
mut.topmost.cors = unique(unlist(topmost.cors.genes))
mut.top.cors = unique(unlist(top.cors))

```

##aggregation of genes
PCA
```{r}
topmost.cors.genes  = unique(c(if.topmost.cors, rrpa.topmost.cors, mut.topmost.cors))
top.cors.genes  = unique(c(if.top.cors, rrpa.top.cors, mut.top.cors))
top.cors.genes = setdiff(top.cors.genes, topmost.cors.genes)

top.cors.genes = setdiff(top.cors.genes, colnames(tcga.dataset))

tcga.exp.norm = normalize.expression(expression_match, num.samp.thr =0) 
mat = apply(tcga.exp.norm[,top.cors.genes], 2, function(tt) (tt-mean(tt))/sd(tt))
pcs=tcga.dataset[,2:51,with=F]
residual.pca = function(mat, control){
  apply(mat, 2, function(tt) {
    control$cov = tt
    (lm(cov~., data= control))$residuals
  })
}

uu = residual.pca(mat, control = pcs)
top.gene.pcs = prcomp(uu)
ww = (top.gene.pcs$sdev^2)
ww1 = cumsum(ww/sum(ww))
ww1[1:100] # only 30% of explained variance is captured

# xx = princomp(mat)
# ww = (xx$sdev^2)
# ww1 = cumsum(ww/sum(ww))

```
## Create dataset and write dataset files 
```{r}

top.pcs = top.gene.pcs$x[,1:50]
tcga.cols = colnames(tcga.dataset)
colnames(top.pcs) = paste0("PC", 51:100)
pc.data = cbind(tcga.dataset[,1:51,with=F], top.pcs)

extra.genes.inx = c("TGFB1", "TGFBR2", "KLRC1")
top.genes.extra = c("PMS2", "MSH6", "EPCAM", "MSH2")
input.output.genes = unique(c(extra.genes.inx, top.genes.extra, tcga.cols[c(62, 89:90, 189:510)])) #101
input.only.genes = setdiff(topmost.cors.genes, input.output.genes)
if.cols = tcga.cols[c(66:88, 91:137, 178:188)]
if.cols = setdiff(if.cols, input.output.genes)
if.data  = tcga.dataset[,if.cols, with=F] #78
setnames(if.data, 68:78, paste0("ERV", 1:11))

protien.data = rrpa.dataset
mut.data = mut.dataset
colnames(mut.data) = paste0(colnames(mut.data), ".mut")
survival.data = tcga.dataset[,c(170:177), with=F]

dataset.new = cbind(pc.data, tcga.exp.norm[,input.only.genes], tcga.exp.norm[,input.output.genes], if.data, protien.data, mut.data, survival.data)
output.dir = "~/project/deeplearning/icb/data/tcga/scrna.v4.pcs/"
write.dataset(output.dir = output.dir, dataset = dataset.new, sample.name = sample.name)

```

