---
title: "Oxphos analysis + citokines analysis of Aviv cohort"
output: html_document
---

## 


```{r}
aviv = readRDS("/liulab/xmwang/oxphos_proj/loading_data/scRNAseq/Mel.all.data.QC.rds")

mal.oxphos = fread("/homes6/xmwang/OXPHOS/scr.oxphos_cell.csv")
```



## citokines
```{r}
all.genes = aviv$genes
citokines = fread("/liulab/xmwang/oxphos_proj/loading_data/surface/cytokine.txt", header=F)
surface.genes = fread("/liulab/xmwang/oxphos_proj/loading_data/surface/ExpressionLigRec.txt", header=T)
citokines = intersect(citokines$V2, all.genes)
surface.genes = setdiff(intersect(surface.genes$ApprovedSymGbol, all.genes), citokines)
```


```{r}
source("../epoch-142/source.proj.R")
data.cytokines =  t(aviv$tpm)[aviv$cell.types =="Mal",citokines]


set.seed(1234)
n_neighbors =30; learning_rate =.1; min_dist = 0.001; pca = NULL
    umap.model <- umap(data.cytokines, 
                       pca = pca, n_neighbors =  n_neighbors, learning_rate = learning_rate,  min_dist = min_dist, n_threads = 32, ret_model=T, metric = "cosine", n_epochs = 100)
  
  d_umap_1 = as.data.table(umap.model$embedding)
  # d_umap_1$col=phenotype_sel.mod$oxphos_score
  d_umap_1$col= mal.oxphos[match(rownames(data.cytokines), object)]$Cor

  p=ggplot(d_umap_1, aes(x=V1, y=V2)) +
    geom_point(size=1,aes(color=col), alpha=0.9) +
    # guides(colour=guide_legend(override.aes=list(size=2))) +
    xlab("umap_1") + ylab("umap_2") +
    theme_classic() +
           theme(axis.text.x=element_blank(), axis.text.y=element_blank()) +
      scale_color_gradient2_tableau(palette = "Orange-Blue Diverging") +
                 guides( color =guide_colourbar(title="oxphos-level"))  
  
 print(p)
 cwd  = "~/project/oxphos/"
 dir.create(cwd, recursive = T)
 ggsave("~/project/oxphos/cytokines_clustering.pdf", p)
 
tsne.dir = sprintf("%s/tnse", cwd)
 dir.create(tsne.dir, recursive = T)
 cell.type.curr = "BLCA"



```



## Kmeans
```{r}
set.seed(100)
data.curr = d_umap_1
data.curr[,UMAP1:=V1]
data.curr[,UMAP2:=V2]
setnames(data.curr, "col", "oxphos")

embedding.mat  = data.curr[,.SD, .SDcols=names(data.curr) %like% "^UMAP"]
data.DI = embedding.mat 
clust.kmeans <- kmeans(data.DI, centers=2, nstart = 25)
table(clust.kmeans$cluster)

data.curr$cluster1 = (clust.kmeans$cluster)
data.curr[,cluster:= as.factor(cluster1)]
p=ggplot(data.curr, aes(x=UMAP1, y=UMAP2)) +
    geom_point(size=2,aes(color=cluster), alpha=0.8) +
    guides(colour=guide_legend(override.aes=list(size=4))) +
    xlab("Dim1") + ylab("Dim2") +
    # ggtitle(label = title) +
    theme_light(base_size=20) +
    scale_color_tableau() + 
    theme_classic() +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank()) 
# if(!is.null(filename)) ggsave(file=filename, p)

print(p)
filename = sprintf("%s/kmean_clustering.pdf", tsne.dir,  "response")
ggsave(file=filename, p)
```



## Diffusion map 

```{r}
library(destiny)
library(Biobase)
library(ggthemes)
library(scater)
set.seed(1234)
dm.obj= DiffusionMap(data.cytokines, density_norm=F, rotate = F)
data.curr$diffusionmap1 = eigenvectors(dm.obj)[, 1] 
data.curr$diffusionmap2 = eigenvectors(dm.obj)[, 2] 
p.dm1 = ggplot(data.curr, aes(x = diffusionmap1, y = diffusionmap2, colour = oxphos)) +
    geom_point(alpha=0.7) + scale_color_gradient2_tableau(palette = "Orange-Blue Diverging")+ 
    xlab("Diffusion component 1") + 
    ylab("Diffusion component 2") +
    theme_classic()

print(p.dm1)
ggsave(filename = sprintf("%s/%s_%s_diffusion.pdf", tsne.dir, cell.type.curr, "oxphos"),p.dm1)

p.dm2 = ggplot(data.curr, aes(x = diffusionmap1, y = diffusionmap2, colour = cluster)) +
    geom_point() + 
    scale_color_tableau() + 
    xlab("Diffusion component 1") + 
    ylab("Diffusion component 2") +
    theme_classic()

ggsave(filename = sprintf("%s/%s_%s_diffusion_cluster.pdf", tsne.dir, cell.type.curr, "oxphos"),p.dm2)

dp = DPT(dm.obj)
data.curr$dpt = dp$dpt
p.dm3 <- ggplot(data.curr, aes(x = diffusionmap1, y = diffusionmap2, colour = dpt)) +
    geom_point() + 
    xlab("Diffusion component 1") + 
    ylab("Diffusion component 2")+
    theme_classic()

library(cowplot)
p <- plot_grid(p.dm1, p.dm2, p.dm3, nrow=2)
print(p)

ggsave(filename = sprintf("%s/%s_%s_diffusion_cluster_combined.pdf", tsne.dir, cell.type.curr, "oxphos"),p)
```
```{r}
data.curr$sample.name = rownames(data.cytokines)
save(file=paste0(cwd, "/data.curr.RData"), data.curr)
```

## Slingshot 
Adopted from [Slingshot Vignette](https://bioconductor.org/packages/release/bioc/vignettes/slingshot/inst/doc/vignette.html)
```{r}
library(slingshot)
library(gam)
# ?sli ngshot
# slingshot.out = slingshot(embedding.mat, clusterLabels = data.curr$cluster,  reducedDim = 'PCA')
embedding.mat = data.cytokines
mat = as.matrix(embedding.mat)
rownames(mat) = data.curr$sample.name
sce = SingleCellExperiment(assays = list(tpm= t(mat)))
dm = cbind(DC1 = dm.obj$DC1, DC2 = dm.obj$DC2)
dm2 = eigenvectors(dm.obj)[, 1:2]
reducedDims(sce)= SimpleList(DiffMap = dm, DiffMap2 = dm2)
colData(sce)$graph = data.curr$cluster
slingshot.out = slingshot(sce, clusterLabels = "graph",  reducedDim = 'DiffMap2')

library(RColorBrewer)
colors <- colorRampPalette(brewer.pal(11,'Spectral'))(100)
plotcol <- colors[cut(slingshot.out$slingPseudotime_1, breaks=100)]

plot(reducedDims(slingshot.out)$DiffMap, col = as.factor(data.curr$response), pch=16, asp = 1)
lines(SlingshotDataSet(slingshot.out), lwd=2, col='black')

require(gam)
require(clusterExperiment)
t <- slingshot.out$slingPseudotime_1


# fit a GAM with a loess term for pseudotime
gam.pval <- apply(t(embedding.mat),1,function(z){
    d <- data.frame(z=z, t=t)
    tmp <- gam(z ~ lo(t), data=d)
    p <- summary(tmp)[4][[1]][1,5]
    p
})
## Loading required package: clusterExperiment
topgenes <- names(sort(gam.pval[gam.pval<1E-10], decreasing = FALSE))
heatdata <- assay(sce)[topgenes,order(t, na.last = NA)]
# rownames(heatdata) = 
heatclus <- slingshot.out$graph[order(t, na.last = NA)]
ce <- ClusterExperiment(heatdata, heatclus)
pdf(paste0(tsne.dir, "/ClusterExperiment_heatmap_markers.pdf"), width=10, height=16)

plotHeatmap(ce, clusterSamplesData = "orderSamplesValue",
            visualizeData = 'transformed')
dev.off()
pdf(paste0(tsne.dir, "/ClusterExperiment_heatmap_markers.pdf"), width=10, height=16)

library(ComplexHeatmap)
Heatmap(heatdata, 
        name = "Cytokines", #title of legend
        column_title = "Variables", row_title = "Samples",
        row_names_gp = gpar(fontsize = 7), # Text size for row names
        column_names_gp = gpar(fontsize = 6) # Text size for row names
)
dev.off()

```



## Create Seurat object

```{r}
library(Seurat)
library(dplyr)
all.genes = colnames(data.cytokines)
sco = CreateSeuratObject(counts =t(data.cytokines), project = "cytokine", min.cells = 2, min.features = 10)
# sco[["percent.mt"]] <- PercentageFeatureSet(sco, pattern = "^MT-")
# VlnPlot(sco, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# 
 xx = log(as.matrix(sco[["RNA"]]@counts) +1)
 sco[["RNA"]]@data = xx
#  
#  # Identification of highly variable features 
 sco <- FindVariableFeatures(sco, selection.method = "vst", nfeatures = 150)
# 
sco <- ScaleData(sco, features = all.genes)

```


# Change the clustering  (https://github.com/satijalab/seurat/issues/252)[here]
```{r}
sco.curr = sco
sco.curr@meta.data$deepImmune.clust = data.curr$cluster[match(colnames(sco.curr), rownames(data.cytokines))]
Idents(object = sco.curr) <- sco.curr@meta.data$deepImmune.clust
sco.curr.markers <- FindAllMarkers(sco.curr, only.pos = TRUE, min.pct = 0.6, logfc.threshold = 0.3)
# sco.markers=FindMarkers(sco.curr,2,thresh.use = 2,test.use = "roc")
# top2 = sco.curr.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)
# VlnPlot(sco.curr,features=top2$gene)


## heatmap plot 
top10 <- sco.curr.markers %>% group_by(cluster) %>% top_n(n = 50, wt = -log(p_val_adj)) %>% subset(p_val_adj < .05)  
p = DoHeatmap(sco.curr, features = top10$gene) + NoLegend()
curr.dir = paste0(tsne.dir, "/markers/")
dir.create(curr.dir)
filename = paste0(curr.dir, "heatmap_markers.pdf")
ggsave(file=filename, p, width=16, height=16)
print(p)
write.table(file = paste0(curr.dir, "markers.txt"), x = top10, sep="\t", row.names = F, col.names = T, quote = F)

```

```{r}
top5 <- sco.curr.markers %>% group_by(cluster) %>% top_n(n = 15, wt = -log(p_val_adj)) %>% subset(p_val_adj < 1E-3) 
p1= VlnPlot(sco.curr, features = top5$gene)
print(p1)
filename = paste0(curr.dir, "/vlnplt_markers.pdf")
ggsave(file=filename, p1, width=16, height=16)
```


```{r}



umap.mat = as.matrix(data.curr[,.(UMAP_1=UMAP1, UMAP_2=UMAP2)])
rownames(umap.mat) = data.curr$sample.name
sco.curr[["dim"]] <- CreateDimReducObject(embeddings = umap.mat, key = "UMAP_", assay = DefaultAssay(sco.curr))
DimPlot(sco.curr, reduction = "dim", pt.size = 5)

feat.p = FeaturePlot(sco.curr,  reduction = "dim",features = top5$gene)
print(feat.p)
filename = paste0(curr.dir, "/featureplot_markers.pdf")
ggsave(file=filename, feat.p, width=16, height=16)


```

## Evaluate the markers in Genentech data
1. characterize those two clusters based on immune phenotypes.  with immune phenotypes i mean... inflammed, IC/TC levels, subtype and survival
2. Compared witin the cluster if oxphos is associated with survival. 

```{r}
library(pROC)
# library(IMvigor210CoreBiologies)
load("/liulab/asahu/data/ssgsea/xiaoman/Genentech/genentech_cds.RData") # data name is cds
cds = cds[,colnames(sco.curr)]
which(!(colnames(sco.curr)==colnames(cds)))
stopifnot(identical(colnames(sco.curr), colnames(cds)))
genentech.exp = counts(cds)
genentech.feat = fData(cds)
genentech.pheno = pData(cds)

# genentech.response.old = genentech.response = genentech.pheno$binaryResponse
# levels(genentech.response) = c(1,0)
# 
# top10 = data.table(top10)

sco.curr@meta.data = cbind(sco.curr@meta.data, genentech.pheno[colnames(sco.curr),])
p = DimPlot(sco.curr, reduction = "dim", group.by = c("Best Confirmed Overall Response", "Immune phenotype", "Lund2"), pt.size = 3)
DimPlot(sco.curr, reduction = "dim", group.by = c("TCGA Subtype", "binaryResponse"), pt.size = 3)
DimPlot(sco.curr, reduction = "dim", group.by = c("TCGA Subtype", "Best Confirmed Overall Response"), pt.size = 3)

print(p)
```

## plot pie chart 

```{r}
library(ggplot2)
library(ggthemes)
# Barplot
# intersect(colnames(data.curr), colnames(genentech.pheno))
data.curr = cbind(data.curr, genentech.pheno[colnames(sco.curr),])

bp<- ggplot(data.curr, aes(x=data.curr$`Immune phenotype`, y=as.factor(cluster), fill=group))+
geom_bar(width = 1, stat = "identity")
bp

myplot <- data.curr[!is.na(Lund2)] %>% ggplot(. , aes(Lund2, group = cluster, fill = cluster)) + 
          # geom_bar(aes(y = ..prop.., fill = factor(..x..)), stat="count") + 
          geom_bar(aes(y = ..prop..), stat="count",position=position_dodge()) + 
          scale_y_continuous(labels=scales::percent) +
          ylab("relative frequencies") 
          # facet_grid(~cluster)
          # facet_grid(~data.curr$`Immune phenotype`)

avi.dt = cbind( data.curr[,.(cluster)], genentech.pheno[colnames(sco.curr), colnames(genentech.pheno)[-c(18:22)]])
avi.dt1 = melt(avi.dt,1)
avi.dt1 = avi.dt1[!is.na(value)][(value!="NE")]

myplot <- ggplot(avi.dt1 , aes(value, group = interaction(cluster, variable), fill = cluster)) + 
          # geom_bar(aes(y = ..prop.., fill = factor(..x..)), stat="count") + 
          geom_bar(aes(y = ..prop..), stat="count",position=position_dodge()) + 
          scale_y_continuous(labels=scales::percent) +
          ylab("relative frequencies") +
        facet_wrap(~variable, scales = "free", ncol=5)
myplot
```
## For each subclass (type + response) compare oxphos
```{r}

data.curr$ESRRA = dataset_ssgsea_mat[colnames(sco.curr), "ESRRA"]
new.data = cbind(data.curr$oxphos, genentech.pheno[colnames(sco.curr), c("Neoantigen burden per MB", "FMOne mutation burden per MB")])
sco.curr[["RNA"]]@data = rbind(sco.curr[["RNA"]]@data, t(new.data))

xx = VlnPlot(sco.curr, feature = top5$gene[4], split.by =  "binaryResponse", group.by = "Lund2")
print(xx)

# avi.dt = cbind(data.curr, genentech.pheno[colnames(sco.curr),])
myplot <- ggplot(data.curr[!is.na(binaryResponse)] , aes(x=Lund2, y = oxphos)) +
  # geom_violin(aes(fill=binaryResponse), alpha=0.6, kernel= "cosine") +
  geom_boxplot(aes(fill=binaryResponse), alpha=0.6) +
  geom_point(position=position_dodge(width=0.75),aes(group=binaryResponse)) 

myplot = myplot + stat_compare_means( aes(group = binaryResponse), label = "p.format")
print(myplot)


library(lme4)
data.curr[,response.bin:=ifelse(binaryResponse=="SD/PD", 0,1)]
 xx = glmer( response.bin  ~ oxphos + (1|Lund2), data=data.curr[Lund2%in%c("Genomically unstable", "Infiltrated")], family = binomial, control = glmerControl(optimizer = "bobyqa"), nAGQ = 10)
 xx = glmer(  os ~ oxphos + (1|Lund2), data=data.curr, family = gaussian)
 xx1 = glmer(  os ~ 1, data=data.curr, family = gaussian)

```

