---
title: "Oxphos mechanism"
output: html_notebook
---

## In cells where ESRRAi is effective, ESRRA is concomitant with SIRT1 (and TNFAPI3) 


```{r}
genes.curr = c("ACO2", "ESRRA", "SIRT1", "TNFAIP3", "NFKB2", "RELA", "RELB", "CYLD","SOCS1", "TANK", "TRAF1", "SENP6", "TOLLIP", "IL6", "IL1B", "TNF")
```

```{r}
#tcga correlation
tcga.expression =fread("/liulab/asahu/data/ssgsea/xiaoman/TCGA_ALLTPM.txt")
library(corrplot)
# genes.curr = c("ESRRA", "SIRT1", "TNFAIP3", "NFKB2", "RELA", "RELB", "ACO2")
setdiff(genes.curr, tcga.expression$gene_name )
exp = t(tcga.expression[match(genes.curr,gene_name ) , -1, with=F])
tt= exp
colnames(tt) = genes.curr

cwd = "~/project/oxphos/mechanism"
dir.create(cwd,recursive=T)

pdf(sprintf("%s/corplot_expression_tcga_nfkb.pdf", cwd))
print(corrplot(cor(tt, method="spearman"), method = "circle",  tl.cex=1, diag = F, order = "original"))
dev.off()
```
##  Explanation
The data shows ESRRA is co-expressed NK-KB genes, and not SIRT1. The correlaton with TNFAIP3 is small. Another important observation while TNFAPI3 is more correlated with RELB than RELA. The opposite is true for ESRRA. This suggests, if ESRRA regulates NF-KB, then it is unlikely to be mediated by TNFAPI3 or SIRT1. 



##  NK-KB
We next asked whether ESRRA regulates NK-KB pathways. We substantiate ESRRA correlation with NK-KB in single cell transcriptome. We controlled for the patients, build linear model to explore these relationships. 

```{r}
library(Seurat)
BCC <- readRDS('/liulab/asahu/data/MAESTRO/GSE129785_BCC_All_MAESTRO_SeuratObj.rds')
BCC_RNA <- subset(BCC, cells = rownames(BCC@meta.data[which(BCC@meta.data[,'tech']=='RNA'),]))
BCC_ATAC <- subset(BCC, cells = rownames(BCC@meta.data[which(BCC@meta.data[,'tech']=='ATAC'),]))
rna.metadata = BCC_RNA@meta.data
 subset.inx = rna.metadata$assign.ident == "Marcrophages"


# genes.curr = c("ACO2", "ESRRA", "SIRT1", "TNFAIP3", "NFKB2", "RELA", "RELB")
tt = t(as.matrix(BCC_RNA[["RNA"]]@data[genes.curr , subset.inx]))
pdf(sprintf("%s/corplot_expression_BCC_nfkb.pdf", cwd))
print(corrplot(cor(tt, method="spearman"), method = "circle", type = "lower", tl.cex=1, diag = F))
dev.off()
```

```{r}
# install.packages("hdf5r")
scc <- Read10X_h5("~/project/deeplearning/icb/data/sc/GSE123814_human_aPD1/GSE123813_bcc_scRNA_counts.h5")
output.dir1 = "~/project/deeplearning/icb/data/sc/GSE123814_human_aPD1/dca"

## samples x genes
dir.create(output.dir1)
write.table(file=paste0(output.dir1, "/count.mat.txt"), x = t(scc),
    row.names = T, col.names =T,  sep="\t", quote=F)
```




```{r}
dca.scc =fread("~/project/deeplearning/icb/data/sc/GSE123814_human_aPD1/dca/dca_impute/mean_norm.tsv")
metadata.dca = fread("~/project/deeplearning/icb/data/sc/GSE123814_human_aPD1//GSE123813_bcc_metadata.txt")
genes.curr =dca.scc$V1
dca.scc = as.matrix(dca.scc[,-1,with=F])
rownames(dca.scc) =genes.curr

match.inx = match(colnames(BCC_RNA), metadata.dca$cell.id)
dca.scc  = dca.scc[match.inx,]

subset.inx = rna.metadata$assign.ident == "Marcrophages"

tt = dca.scc[subset.inx, genes.curr ]
pdf(sprintf("%s/corplot_dca_expression_BCC_nfkb.pdf", cwd))
print(corrplot(cor(tt, method="pearson"), method = "circle", type = "lower", tl.cex=1, diag = F))
# corrplot(cor(tt, method="spearman"), method = "circle", type = "lower", tl.cex=1, diag = F)
dev.off()
```

