---
title: "TCR embedding from https://github.com/tbepler/protein-sequence-embedding-iclr2019/"
output: html_notebook
---
## Write fasta files and check other file format 
1. Directory data /liulab/asahu/softwares/protein-sequence-embedding-iclr2019/
```{r}
cwd = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz"
dir.create(cwd)
setwd(cwd)

tab.file = fread("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/SCOPe/astral-scopedom-seqres-gd-sel-gs-bib-95-2.06.test.sampledpairs.txt")

library("seqinr")
ref <- read.fasta(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/SCOPe/astral-scopedom-seqres-gd-sel-gs-bib-95-2.06.train.fa", seqtype = "AA")

```

## ICB expression

```{r}

load("~/project/deeplearning/icb/data/Getz_scRNA/dataset_ssgsea_temp.RData")
load("~/project/deeplearning/icb/data/Getz_scRNA/headers.RData")
load("~/project/deeplearning/icb/data/Getz_scRNA/phenotype_sel.mod.RData")
icb.expression = t(dataset_ssgsea_temp[,2:16292, with=F])
colnames(icb.expression) =  dataset_ssgsea_temp$gene_name
rownames(icb.expression) =  gsub(unlist(headers[1])[-1], pattern="-", replacement=".")
phenotype_sel.mod$sample.name = gsub(phenotype_sel.mod$sample.name, pattern="-", replacement=".")
length(intersect(rownames(icb.expression), phenotype_sel.mod$sample.name))
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
# identical(phenotype_sel.mod$sample.name, dataset.sample.name)
rm(dataset_ssgsea_temp, icb.expression)
```

## For cell type specific genes
```{r}

library(Biobase)
# library(IMvigor210CoreBiologies)
load("/liulab/asahu/data/ssgsea/xiaoman/Genentech/genentech_cds.RData") # data name is cds
genentech.exp = counts(cds)
genentech.feat = fData(cds)
genentech.pheno = pData(cds)
genentech.response.old = genentech.response = genentech.pheno$binaryResponse
levels(genentech.response) = c(1,0)

dataset_phenotype = fread("/liulab/asahu/data/ssgsea/xiaoman/Avin/clinical_ICB_oxphos.txt")
genentech.phenotype_sel.mod = dataset_phenotype[match(colnames(genentech.exp), bcr_patient_barcode)]
genentech.phenotype_sel.mod[, Response:=as.double(Response)]
genentech.phenotype_sel.mod[is.na(Response) & (vital_status == 1) & (survive < 3)]$Response = 0
genentech.phenotype_sel.mod[is.na(Response) & (survive > 7)]$Response = 1
genentech.response  = new.genentech.response  = genentech.phenotype_sel.mod$Response

```

## Getz fasta file
```{r}
source("source.proj.R")
trust4.cdr3 = get.trust4.cdr3()
trust4.copy = trust4.cdr3
# trust4.copy[,fa.label:=paste0(">", )]
trust4.copy$aa = trust4.copy[,{
    start = ifelse(V=="*", 1 + ( str.len %%3), 1)
    end = ifelse(J=="*", floor(str.len/3) *3, str.len)
   aa = translate.seq(stringr::str_sub(CDR3,start, end))
   ifelse(V=="*", stringi::stri_c("X", aa), 
          ifelse(J=="*", stringi::stri_c("X", aa), aa))
}]
trust4.copy[,myid:=paste0("V",seq(.N))]
trust4.copy[,fa.file:=paste(paste0(">",myid),  "c.2.1.2", V, J, C, cellid)]
trust4.copy[,aa.len:=stringr::str_length(aa)]
## for PFAM
# >S0E8J4_GIBF5/356-387 S0E8J4.1 PF16503.4;zn-ribbon_14;
trust4.copy[,fa.file:=paste(paste0(">",myid),  "/356-387 S0E8J4.1 PF16503.4;zn-ribbon_14;")]
trust4.filtered = trust4.copy[! (is.na(aa.len) |is.na(sample.name.short))]
```


```{r}
trust4.fa = c(t(trust4.filtered[,.(fa.file, aa)]))
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.fa", x= trust4.fa, row.names = F, quote =F, col.names = F)
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.100.fa", x= trust4.fa[1:100], row.names = F, quote =F, col.names = F)
```
## PFAM Amino acid embedding 
```{r}

```

## Using different stat
1. Take mean 
2. Use last layer 
3. Use center 


## Find the top predicted cdr3 for each sample.
##  Current features
```{r}
library(avinash)
my.unique = function(tt) unique(tt[tt!="*"])
trust4.filtered$currid = seq(nrow(trust4.filtered))
trust4.cdr3.filtered = trust4.filtered[, .SD[abs.score == max(abs.score)],
                                 by=.(is.alpha, cellid)][, .SD[which.min(str.len)], 
                                                         by = .(is.alpha, cellid)][abs.score>=1]
trust4.cdr3.filtered[, cdr3.type:=my.unique( sapply(.SD, function(tt) substr(tt,1,3))), .SDcols=c("V", "J", "C"), by = seq_len(nrow(trust4.cdr3.filtered))]
phenotype_sel.curr = phenotype_sel.mod[match(trust4.cdr3.filtered$sample.name, sample.name)]
resp = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/GSE120575_patient_ID_single_cells.txt", skip=19)
colnames(resp)[1:35] = paste0("V",1:35)
resp$V2 = gsub(resp$V2, pattern="-", replacement=".")
resp.matched=resp[match(trust4.cdr3.filtered$sample.name, resp$V2)]
if.genes = get.checkpoint.genes()
icb.expression.curr = icb.expression.matched[trust4.cdr3.filtered$sample.name,intersect(if.genes, colnames(icb.expression.matched))]
icb.expression.curr.all = icb.expression.matched[trust4.cdr3.filtered$sample.name,]
if.curr = get.phenotype(sample.subet = trust4.cdr3.filtered$sample.name)
tcell.types = sort(unique(phenotype_sel.curr$assign.ident.2))[c(2:3,5:7, 10:11)]
trust4.cdr3.filtered = cbind(trust4.cdr3.filtered, resp.matched[,.(patient=V5, response=V6, treatment=V7)]) 


# only t.cells
filter1 = substr(trust4.cdr3.filtered$cdr3.type,1,3) == "TRB" & (phenotype_sel.curr$assign.ident.2 %in% tcell.types)
trust4.cdr3.filtered.curr = trust4.cdr3.filtered[filter1,]
icb.expression.curr.curr = icb.expression.curr.all[filter1,]
phenotype_sel.curr.curr = phenotype_sel.curr[filter1,]

library(Seurat)
sco.curr = CreateSeuratObject(counts =t(icb.expression.curr.curr), project = "nir", min.cells = 2, min.features = 200)
sco.curr[["percent.mt"]] <- PercentageFeatureSet(sco.curr, pattern = "^MT-")
# We filter cells that have >5% mitochondrial counts
# sco.curr<- subset(sco.curr, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
xx = log(as.matrix(sco.curr[["RNA"]]@counts) +1)
sco.curr[["RNA"]]@data = xx
# sco.curr = sco[,trust4.cdr3.filtered.curr$sample.name]
sco.curr <- ScaleData(sco.curr, features = rownames(sco.curr))
sco.curr <- FindVariableFeatures(sco.curr, selection.method = "vst", nfeatures = 2000)
```

```{r}
library(dplyr)
options(warn=-1)
modes = c("last", "mean", "middle", "SIF")
modes = c("SIF")
library(reticulate)
np <- import("numpy")
trust.files = list.files("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/", pattern = ".sav.fvec.npy", full.names = T)
trust.files = c("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.facdr3b_no_pretrain_epoch100.sav.fvec.npy", "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.famypfam_no_pretrain_epoch100.sav.fvec.npy")
for (file1 in trust.files) {
    print(paste("processing", file1))
    trust4.pfam <- np$load(file1)
    
    for (mode in modes) {
        dir.name = gsub(".sav.fvec.npy", gsub("getz.cdr3.fa", basename(file1), replacement = ""), replacement = "")
        curr.dir = dirname(file1)
        
        dir.curr = sprintf("%s/outputs/%s/%s/", curr.dir, dir.name, mode)
        dir.create(dir.curr, recursive = T)
        # out = plot.tcr.embedding(trust.curr=trust4.pfam, save.dir=dir.curr, title="pfam", mode=mode)
        out1 = plot.tcr.embedding.2(trust.curr=trust4.pfam, save.dir=sprintf("%s/v2/",dir.curr), title="pfam", mode=mode )
    }
}


options(warn=0)
```


## positional sif

```{r}
source("source.proj.R")
positional.sifs = list(smooth.sif=function(xx, a=.001) a/(a+xx), log.sif=function(xx) -log(xx), smooth.sif2=function(xx, a=.001) a/(a+xx^2),  log.sif2=function(xx) -2*log(xx))
remove.pc = F; do.kmeans = T
trunc.postional.sif = function(xx, lead=3, lag =4, a=.001){
    xx = a/(a+xx)
    xx[seq(lead)] = 0
    # len.xx = length(xx)
    xx[length(xx) - seq(lag-1,0)] = 0
    xx
}
positional.sifs = list(trunc.postional.sif=trunc.postional.sif)
np <- import("numpy")
trust.files = list.files("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/", pattern = ".sav.fvec.npy", full.names = T)
# trust.files = list.files("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/", pattern = ".sav.fvec.npy", full.names = T)
for (file1 in trust.files[1]) {
    print(paste("processing", file1))
    trust4.pfam <- np$load(file1)
    
    for (mode.inx in seq_along(positional.sifs)) {
        dir.name = gsub(".sav.fvec.npy", gsub("getz.cdr3.fa", basename(file1), replacement = ""), replacement = "")
        curr.dir = dirname(file1)
        
        dir.curr = sprintf("%s/positional.sif.witout.pc.kmeans/%s/%s/", curr.dir, dir.name, names(positional.sifs)[mode.inx])
        dir.create(dir.curr, recursive = T)
        mode = positional.sifs[mode.inx]
        # out = plot.tcr.embedding(trust.curr=trust4.pfam, save.dir=dir.curr, title="pfam", mode=mode)
        out1 = plot.tcr.embedding.2(trust.curr=trust4.pfam, save.dir=sprintf("%s/v2/",dir.curr), title="pfam", mode=mode, plot.seurat.marker = T)
    }
}

```


```{r}
library(avinash)
trust4.pfam.agg = aggregate.embedding(trust4.pfam, trust4.filtered, mode="last")
trust4.pfam.agg.filtered = trust4.pfam.agg[trust4.cdr3.filtered$currid,]

n_neighbors =15; learning_rate =1; min_dist = .01; pca = 50

umap.all.p = plotUMAP(data = trust4.pfam.agg.filtered, col=NULL, color.col=NULL, size=1, do.discrete=T, 
                      n_neighbors = n_neighbors, learning_rate = learning_rate, min_dist = min_dist, pca=pca,
                      title= "PFAM last",
                      filename=NULL , n_epochs = 100, metric = "euclidean")

umap.all.p = plotUMAP(data = trust4.pfam.agg.filtered, umap.model = umap.all.p[[1]], col=NULL, color.col=NULL, size=1, do.discrete=T, 
                      n_neighbors = n_neighbors, learning_rate = learning_rate, min_dist = min_dist, pca=pca,
                      title= "PFAM last",
                      filename=NULL , n_epochs = 100, metric = "euclidean")
print(umap.all.p[[2]])
```

## 

```{r}

cols = which(colnames(trust4.cdr3.filtered) %in% c("V","J","C"))

trust4.cdr3.filtered[, cdr3.type:=my.unique( sapply(.SD, function(tt) substr(tt,1,3))), .SDcols=cols, by = seq_len(nrow(trust4.cdr3.filtered))]
n_neighbors =15; learning_rate =1; min_dist = .01; pca = 50
umap.all.p = plotUMAP(data = trust4.pfam.agg.filtered,  
                      # umap.model = umap.all.p[[1]],
                      col=NULL, color.col=as.factor(trust4.cdr3.filtered$cdr3.type), size=1, do.discrete=T, 
                      n_neighbors = n_neighbors, learning_rate = learning_rate, min_dist = min_dist, pca=pca,
                      title= "PFAM last",
                      filename=NULL , n_epochs = 100, metric = "euclidean")

print(umap.all.p[[2]])
```

```{r}
data.umap = data.table(umap.all.p[[1]]$embedding)
p1 = plot.sequence.and.clustering(data.clust=data.umap, text = trust4.cdr3.filtered$aa, color.col = as.factor(trust4.cdr3.filtered$cdr3.type), num.plot.seq =100, text.size =2) 
# p1
library(RColorBrewer)
xx1 = color.clusters.features( data=as.data.frame(phenotype_sel.curr), cluster=data.umap,  color.cols =c("assign.ident", 
                                                                                                                  "assign.ident.2"),
                              title="",size=2, filename=NULL, normalize=F, do.discrete=T)
print(xx1$assign.ident.2)
```






## TCGA fasta file

```{r}

load("~/project/deeplearning/data/tcga/tcga_extended_tcr.V3.RData")

tcga_extended_tcr[,type:=substr(reportgene, 1,3)]
# tcga_cdr3_trbv_complete = tcga_extended_tcr[ (status=='complete') & (type =="TRB") ]
tcga_extended_tcr[,aa.len:=nchar(cdr3aa)]
tcga_cdr3_trbv_complete = tcga_extended_tcr[ (status=='complete') & (type =="TRB") & (nchar(cdr3aa) <= 25)]
tcga_cdr3_trbv_complete[,cdr3aa:=as.character(cdr3aa)]
tcga_cdr3_trbv_complete[,sample.name:=substr(sample,1,12)]
tcga_cdr3_trbv_complete[,sample.name.match:=gsub(sample.name,pattern="\\-", replacement = ".")]
tcga_cdr3_trbv_complete = tcga_cdr3_trbv_complete[!grepl("\\*", cdr3aa),]

tcga_cdr3_trbv_complete[,myid:=paste0("V",seq(.N))]
tcga_cdr3_trbv_complete[,fa.file:=paste(paste0(">",myid),  "/356-387 S0E8J4.1 PF16503.4;zn-ribbon_14;")]
dir.create("/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/tcga")
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/tcga/tcga.cdr3.fa", x=  c(t(tcga_cdr3_trbv_complete[,.(fa.file, cdr3aa)])), row.names = F, quote =F, col.names = F)
train.inx = sample(seq(nrow(tcga_cdr3_trbv_complete)), size = nrow(tcga_cdr3_trbv_complete) * .8)
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/tcga/tcga.train.cdr3.fa", x=  c(t(tcga_cdr3_trbv_complete[train.inx,.(fa.file, cdr3aa)])), row.names = F, quote =F, col.names = F)
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/tcga/tcga.test.cdr3.fa", x=  c(t(tcga_cdr3_trbv_complete[!train.inx,.(fa.file, cdr3aa)])), row.names = F, quote =F, col.names = F)

# write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.100.fa", x= trust4.fa[1:100], row.names = F, quote =F, col.names = F)

```


```{r}
trust4.copy = trust4.cdr3
# trust4.copy[,fa.label:=paste0(">", )]
trust4.copy$aa = trust4.copy[,{
    start = ifelse(V=="*", 1 + ( str.len %%3), 1)
    end = ifelse(J=="*", floor(str.len/3) *3, str.len)
   aa = translate.seq(stringr::str_sub(CDR3,start, end))
   ifelse(V=="*", stringi::stri_c("X", aa), 
          ifelse(J=="*", stringi::stri_c("X", aa), aa))
}]
trust4.copy[,myid:=paste0("V",seq(.N))]
trust4.copy[,fa.file:=paste(paste0(">",myid),  "c.2.1.2", V, J, C, cellid)]
trust4.copy[,aa.len:=stringr::str_length(aa)]
## for PFAM
# >S0E8J4_GIBF5/356-387 S0E8J4.1 PF16503.4;zn-ribbon_14;
trust4.copy[,fa.file:=paste(paste0(">",myid),  "/356-387 S0E8J4.1 PF16503.4;zn-ribbon_14;")]
trust4.filtered = trust4.copy[! (is.na(aa.len) |is.na(sample.name.short))]
trust4.fa = c(t(trust4.filtered[,.(fa.file, aa)]))
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.fa", x= trust4.fa, row.names = F, quote =F, col.names = F)
write.table(file = "/liulab/asahu/softwares/protein-sequence-embedding-iclr2019/data/Getz/getz.cdr3.100.fa", x= trust4.fa[1:100], row.names = F, quote =F, col.names = F)
```



## ssgsea in Getz data
```{r}
library('org.Hs.eg.db')
entrezid = mapIds(org.Hs.eg.db, colnames(icb.expression.matched), 'ENTREZID', 'SYMBOL')
icb.expression.entrez = icb.expression.matched[,!is.na(entrezid)]
colnames(icb.expression.entrez) = entrezid[!is.na(entrezid)]
icb.expr.ez.mat = t(icb.expression.entrez)
save(file="/liulab/asahu/data/ssgsea/xiaoman/getz/icb.expr.RData", icb.expr.ez.mat)



library(rslurm)
calculate.ssgsea <- function(expr) {
    library(GSVA)
    library(GSVAdata)
    # data(commonPickrellHuang)
    data(c2BroadSets)
    # esrnaseq.ssgsea <- gsva(expr, c2BroadSets, method = "ssgsea", kcdf="Poisson",  min.sz=5, max.sz=500, mx.diff=TRUE, ssgsea.norm=TRUE, verbose=FALSE, parallel.sz=1 )
    esrnaseq.ssgsea <- gsva(expr, c2BroadSets,  kcdf="Poisson",  min.sz=5, max.sz=500, mx.diff=TRUE, ssgsea.norm=TRUE, verbose=FALSE, parallel.sz=1 )
 
    esrnaseq.ssgsea
}


slurm_ssgsea <- function(col.index, ncores=32, expr.file="/liulab/asahu/data/ssgsea/xiaoman/getz/icb.expr.RData") {
    load(expr.file)
    expr.mat = icb.expr.ez.mat[,col.index]
    rm(icb.expr.ez.mat); gc()
    library(doParallel)
    library(itertools)
    ncores <- 32
    cl <- makePSOCKcluster(ncores)
    registerDoParallel(cl)
    X3 <- foreach(m=isplitCols(expr.mat, chunks=ncores), .combine='cbind') %dopar% {
        calculate.ssgsea(m)
    }
    stopCluster(cl) 
    X3
}


slurm_ssgsea <- function(col.index, ncores=32, expr.file="/liulab/asahu/data/ssgsea/xiaoman/getz/icb.expr.RData") {
    load(expr.file)
    expr.mat = icb.expr.ez.mat[,col.index]
    rm(icb.expr.ez.mat); gc()
    calculate.ssgsea(expr.mat)
}

cpus_per_node = 1
num.nodes = 64
# n = ncol(icb.expr.ez.mat)
n = 15116
k = ceiling(n/num.nodes)
index_list = split(seq(n), rep(1:ceiling(n/k), each=k)[1:n])
# index_list= list(1:10, 11:20)
sjob <- slurm_apply(function(i) slurm_ssgsea(index_list[[i]]), 
                    data.frame(i = seq_along(index_list)),
                    add_objects = c("slurm_ssgsea", "index_list", "calculate.ssgsea"),
                    nodes = num.nodes, cpus_per_node = cpus_per_node, submit = T)
res_raw <- get_slurm_out(sjob, outtype = 'raw')
# data(genderGenesEntrez)
# 
# MSY <- GeneSet(msYgenesEntrez, geneIdType=EntrezIdentifier(),
#                collectionType=BroadCollection(category="c2"), setName="MSY")
# XiE <- GeneSet(XiEgenesEntrez, geneIdType=EntrezIdentifier(),
#                collectionType=BroadCollection(category="c2"), setName="XiE")
# 
# canonicalC2BroadSets <- GeneSetCollection(c(canonicalC2BroadSets, MSY, XiE))
esrnaseq <- gsva(pickrellCountsArgonneCQNcommon_eset, canonicalC2BroadSets, min.sz=5, max.sz=500,
                 kcdf="Poisson", mx.diff=TRUE, verbose=FALSE, parallel.sz=1, ssgsea.norm=TRUE)



corsrowsgs <- sapply(1:nrow(esmicro),
                     function(i, esmicro, esrnaseq) cor(esmicro[i, ], esrnaseq[i, ], method="spearman"),
                     exprs(esmicro), exprs(esrnaseq))

```

