---
title: "BCR analysis"
output: html_notebook
---

## Change fasta reference file 

```{r}
# install.packages("seqinr")
library("seqinr")
ref <- read.fasta(file =  "/liulab/lsong/projects/bcr/bcrtcr.fa")
annotation <- getAnnot(ref)
annotation.dt = data.table(do.call(rbind, lapply(annotation, 
                                                 function(tt) unlist(strsplit(tt, split = " "))
)))
annotation.dt[,gene:=gsub(V1, pattern = ">", replacement = "")]
library(biomaRt)
ensembl <- useMart("ensembl", dataset= "hsapiens_gene_ensembl")
affy2ensembl = getBM(attributes = c("ensembl_transcript_id", "hgnc_symbol", "ensembl_transcript_id_version", "transcript_length"), mart=ensembl)
affy2ensembl = as.data.table(affy2ensembl)
setdiff(annotation.dt$gene, affy2ensembl$hgnc_symbol)
intersect(annotation.dt$gene, affy2ensembl$hgnc_symbol)
dt1 = affy2ensembl[hgnc_symbol %in% annotation.dt$gene]
aa = biomaRt::listAttributes(ensembl)
grep("ensembl Transcript ID", aa$description, ignore.case = T, value=T)

# Verify if ENSEMBL transcript id and length are same 
verify = F
if (verify) {
    
    bb =fread("/liulab/asahu/temp1.xx",header = F)
    cc = bb[,gsub("\\.[0-9]+", V1, replacement = "")]
    setdiff(dt1$ensembl_transcript_id_version, bb$V1)
    grep("ENST00000631448", bb$V1,  value=T)
    intersect(dt1$ensembl_transcript_id, cc)
    intersect(dt1$ensembl_transcript_id_version, bb$V1)
    dt1[ensembl_transcript_id=="ENST00000390425"]$ensembl_transcript_id_version
    bb[cc=="ENST00000390425"]
}

ensembl.id = dt1[match(annotation.dt$gene, dt1$hgnc_symbol)]$ensembl_transcript_id_version
annotation.dt$ensembl = ensembl.id
annotation.dt$start = 1
annotation.dt[,end:=as.numeric(V4)-as.numeric(V3) +1]
annotation.dt[,new.annot:=paste(paste0(">", gene), ensembl, start, end, V5)]
new.annot = lapply(annotation.dt$new.annot, function(tt) tt)
names(new.annot) = NULL 
ref1 = copy(ref)
# getAnnot(ref1) = new.annot
for (ii in seq_along(ref1)) {
    attr(ref1[[ii]], "Annot") = new.annot[[ii]]
    
}
xx = getAnnot(ref1)

dir.create("/liulab/asahu/projects/bcr")
new.names = gsub("^>", unlist(new.annot), replacement = "" ) 
write.fasta(sequences = ref1, names =new.names, nbchar = 10000, file.out = "/liulab/asahu/projects/bcr/bcrtcr.fa")



```

## Using the hg19 ensembl reference

```{r}
map = fread("/liulab/asahu/data/ssgsea/xiaoman/hg19_hugo2ensmblMap.txt")
bam.header =fread("/liulab/asahu/projects/bcr/A10_P1_M53_L001.header", header=F)
bam.header[,ensembl:=gsub("^SN:", V2, replacement = "")]
bam.header = bam.header[-nrow(bam.header)]
setdiff(annotation.dt$gene, map$Gene.name)
map.ig = map[Gene.name %in% ]
setnames(map,colnames(map), gsub(colnames(map), pattern=" ", replacement = "."))
intersect(map$Transcript.stable.ID.version, bb$V1)

mixcr = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/mixcr_combined.txt")
load("/liulab/asahu/files.txt") # uu 
mixcr = mixcr[cellname %in% uu$xx]
mixcr1  = mixcr[grep("TRBD1", allDHitsWithScore)]
mixcr1  = mixcr[grepl("TRBJ1", allJHitsWithScore) & grepl("TRBD1", allDHitsWithScore)]
nn = affy2ensembl[grepl("TRBD1",hgnc_symbol) |  grepl("TRBJ1",hgnc_symbol) ]
nn$ensembl_transcript_id %in% cc

mixcr.sel = mixcr[cellname == "A1_P8_M82_L001"]

write.table(file="/liulab/asahu/mixcr_A1_P8_M82_L001.out", mixcr.sel,quote=F)


map.bam = map[Transcript.stable.ID.version %in% bam.header$ensembl]
gene.intersect = intersect(map.bam$Gene.name, annotation.dt$gene)
gene.absent = setdiff(annotation.dt$gene, map.bam$Gene.name)
ensembl.id.dt = map.bam[match(gene.intersect, map.bam$Gene.name)]
set.seed(12321)
map.bam.diff = map.bam[!(Gene.name %in% annotation.dt$gene)]

rest.ensembl.id.dt = map.bam.diff[sample(nrow(map.bam.diff), length(gene.absent))]
rest.ensembl.id.dt$Gene.name = gene.absent

id.dt = rbind(ensembl.id.dt, rest.ensembl.id.dt)
annotation.dt$ensembl = id.dt[match(annotation.dt$gene, id.dt$Gene.name)]$Transcript.stable.ID.version

annotation.dt$start = 1
annotation.dt[,end:=as.numeric(V4)-as.numeric(V3) +1]
annotation.dt[,new.annot:=paste(paste0(">", gene), ensembl, start, end, V5)]
stopifnot(sum(is.na(annotation.dt$ensembl)) ==0 )
stopifnot(any(annotation.dt$ensembl %in% bam.header$ensembl))
new.annot = lapply(annotation.dt$new.annot, function(tt) tt)
names(new.annot) = NULL 
ref1 = copy(ref)
# getAnnot(ref1) = new.annot
for (ii in seq_along(ref1)) {
    attr(ref1[[ii]], "Annot") = new.annot[[ii]]
    
}
xx = getAnnot(ref1)

dir.create("/liulab/asahu/projects/bcr")
new.names = gsub("^>", unlist(new.annot), replacement = "" ) 
write.fasta(sequences = ref1, names =new.names, nbchar = 10000, file.out = "/liulab/asahu/projects/bcr/bcrtcr.hg19.fa")
```

## check files missing

```{r}
aa = list.files(path = "/rna-seq/readcounts", pattern = ".results", full.names = T, recursive = T)
xx = basename(aa)                  
xx = gsub(".rsem.genes.results", xx, replacement = "")
save(file="/liulab/asahu/temp1.RData", xx)
xx = gsub(xx, )
sum(grepl("_L001", xx))
xx[!grepl("_L001", xx)]
"H2_P5_M17_T1" %in% phenotype_sel.mod$sample.name
"H2_P5_M17" %in% phenotype_sel.mod$sample.name
grep("H9_P6_M17", xx, value = T)
tt = sapply(phenotype_sel.mod$sample.name, nchar)
vv = phenotype_sel.mod$sample.name[tt>15]
grep("A10_P1_MM",xx, value=T)
sum(xx %in% phenotype_sel.mod$sample.name)


process.cell.name <- function(tt) {
    tt1 = strsplit(tt, split = "\\.|_")
    paste(tt1[[1]][1:3], collapse = "_")
}


phenotype_sel.mod$patient.name=resp[match(phenotype_sel.mod$sample.name, V2)]$V5
phenotype_sel.mod$sample.name.processed = sapply(phenotype_sel.mod$sample.name,  process.cell.name)
xx.processed =  sapply(xx,  process.cell.name)
temp = phenotype_sel.mod
temp[, present:=ifelse((sample.name %in% xx) | (sample.name.processed %in% xx.processed), 1, 0)]
head(temp[present==0]$sample.name)
# phenotype_sel.mod$sample.name[14339]
grep("G6_P1_M39", phenotype_sel.mod$sample.name, value=T)
grep("A10_P1_M23", xx, value=T)
# temp[present==0]$patient.name 
# sum(resp$V2  %in% xx | resp$V2 %in% xx.processed)
sort(table(temp[present==0]$patient.name))
sum(temp[present==0]$patient.name == "Pre_P8")
tt1 = table(temp[present==0]$patient.name)
tt2 = table(temp$patient.name)
tt3 = sort(tt1/tt2[names(tt1)])
t(t(tt3))
```

## Run trust
Run in Kraken 
```{r}

run_trust <- function(file) {
    filename = basename(file)
    dir.prefix = gsub('*.toTranscriptome.sorted.bam', filename, replacement = "")
    dir.name = sprintf("/rna-seq/trust4/outputs/%s", dir.prefix)
    dir.create(dir.name, recursive = T)
    setwd(dir.name)
    command1 = sprintf("cd %s; /liulab/asahu/projects/bcr/bcr_git/run-trust4 -b  %s -f /liulab/asahu/projects/bcr/bcrtcr.hg19.fa --ref /liulab/asahu/projects/bcr/IMGT+C.fa", dir.name, file)
    out = system(command1, wait = T)
    print(command1)
    out 
}

dir.create("/rna-seq/trust4/dbExtract", recursive = T)
files = grep("Pre", list.files(path="/rna-seq/dbGAP_tar", pattern = "*.tar", full.names = T), value=T)
require(doMC)
require(foreach)
registerDoMC(cores = 48)
out.all = list()
for (file1 in files) {
    setwd("/rna-seq/trust4/dbExtract")
    command = sprintf("tar -xf %s --wildcards --no-anchored \'*.toTranscriptome.sorted.bam\' -C /rna-seq/trust4/dbExtract", file1)
    uu = tryCatch(system(command, wait = T),
             error = function(e) NULL)
    bam.files = list.files(path="/rna-seq/trust4/dbExtract", pattern = "*.toTranscriptome.sorted.bam", recursive=T, full.names=T)
    out = NULL
    if(!is.null(bam.files))
        out = foreach(bam.file = bam.files) %dopar% {
            run_trust(bam.file)
        }
    out.all[[file1]] = out 
    bam.files = NULL
    system("rm -rf /rna-seq/trust4/dbExtract/*", wait=T)
    print(sprintf("%s finished", file1))
}

<<<<<<< HEAD


=======
>>>>>>> e11a50b592485b4d1b2a8337dffcf608ce843ac8
out.all = list()
for (file1 in files) {
    command = sprintf("tar -xf %s --wildcards --no-anchored \'*.toTranscriptome.sorted.bam\' -C /rna-seq/trust4/dbExtract", file1)
    uu = tryCatch(system(command, wait = T),
             error = function(e) NULL)
    bam.files = list.files(path="/rna-seq/trust4/dbExtract", pattern = "*.toTranscriptome.sorted.bam", recursive=T, full.names=T)
    out = NULL
    if(!is.null(bam.files))
        out = foreach(bam.file = bam.files) %dopar% {
            run_trust(bam.file)
        }
    out.all[[file1]] = out 
    bam.files = NULL
    system("rm -rf /rna-seq/trust4/dbExtract/*", wait=T)
    print(sprintf("%s finished", file1))
}

# tar -xf $file --wildcards --no-anchored '*.toTranscriptome.sorted.bam' -C /rna-seq/trust4/dbExtract

# /liulab/asahu/projects/bcr/bcr_git/run-trust4 -b  /rna-seq/Pre_20.out/broad/hptmp/Moshe_S/dbGAP/Pre_P20/A10_P1_M53_L001/A10_P1_M53_L001.Aligned.toTranscriptome.sorted.bam -f /liulab/asahu/projects/bcr/bcrtcr.hg19.fa --ref /liulab/asahu/projects/bcr/IMGT+C.fa

```

# to account for location tar file  
```{r}


bam.files = list.files(path="/liulab/asahu/projects/bcr/", pattern = "*.toTranscriptome.sorted.bam", recursive=T, full.names=T)
require(doMC)
require(foreach)
registerDoMC(cores = 48)
out = foreach(bam.file = bam.files) %dopar% {
    run_trust(bam.file)
}



# run in revrese order 
run_trust1 <- function(file) {
    filename = basename(file)
    dir.prefix = gsub('*.toTranscriptome.sorted.bam', filename, replacement = "")
    out = NULL
    dir.name = sprintf("/rna-seq/trust4/outputs/%s", dir.prefix)
    if(!dir.exists(dir.name)){
        dir.create(dir.name, recursive = T)
        setwd(dir.name)
        command1 = sprintf("cd %s; /liulab/asahu/projects/bcr/bcr_git/run-trust4 -b  %s -f /liulab/asahu/projects/bcr/bcrtcr.hg19.fa --ref /liulab/asahu/projects/bcr/IMGT+C.fa", dir.name, file)
        out = system(command1, wait = T)
        print(command1)
    }
    out 
}

bam.files = list.files(path="/liulab/asahu/projects/bcr/", pattern = "*.toTranscriptome.sorted.bam", recursive=T, full.names=T)
bam.files = rev(bam.files)
require(doMC)
require(foreach)
registerDoMC(cores = 48)
out = foreach(bam.file = bam.files) %dopar% {
    run_trust1(bam.file)
}


```

