---
output: html_notebook
author: Avinash Das Sahu
---
# this report test if the scRNA gene clusters are predictive of ICB response in Genentech cohorts.
# response_information

##Correcting scRNA annotation in using [https://github.com/dviraran/SingleR](singelR).


```{r}

#' Wrapper function to create a SingleR object
#'
#' @param counts single cell expression matrix with gene row and sample column. 
#' @param project.name the project name
#' @param min.genes Include cells where at least this many genes are detected (number non-zero genes).
#' @param technology The technology used for creating the single-cell data.
#' @param species The species of the sample ('Human' or 'Mouse').
#' @param citation a citation for the project.
#' @param ref.list a list of reference objects. If NULL uses the predefined reference objects - Mouse: ImmGen and Mouse.RNAseq, Human: HPCA and Blueprint+Encode. 
#' @param normalize.gene.length if a full-length method set to TRUE, if a 3' method set to FALSE.
#' @param variable.genes variable gene method to use - 'sd' or 'de'. Default is 'de'.
#' @param fine.tune perform fine tuning. Default is TRUE. Fine-tuning may take long to run.
#' @param do.signatures create signatures data
#' @param clusters input cluster id for each of the cells with at least min.genes, if NULL uses SingleR clusterings.
#' @param do.main.types run the SingleR pipeline for main cell types (cell types grouped together) as well.
#' @param reduce.file.size remove less used SingleR fields that increase the object size.
#' @param temp.dir used by the SingleR webtool.
#' @param numCores Number of cores to use.
#'
#' @return a SingleR object
mySCObject = function(counts, project.name,
                               min.genes=0,technology='10X',
                               species='Human',citation='',
                               ref.list=list(),normalize.gene.length=F,
                               variable.genes='de',fine.tune=T,
                               do.signatures=F,clusters=NULL,
                               do.main.types=T,reduce.file.size=T,
                               temp.dir=NULL,numCores = SingleR.numCores) {
  
  require(SingleR)
    
    
    print(paste0('Dimensions of counts data: ',
                 nrow(counts),'x',ncol(counts)))
    
    singler = list()
    
    
    N = colSums(counts>0)
    counts = counts[,N>=min.genes]
    orig.ident = colnames(counts)
    
    sc.data.gl = counts 
    
    if (length(ref.list)==0) {
        if (species == 'Mouse') {
            #if (!exists('immgen'))
            #  data('Immgen')
            #if (!exists('mouse.rnaseq'))
            #  data('Mouse-RNAseq')
            res = list(SingleR.CreateObject(sc.data.gl,immgen,clusters,species,
                                            citation,technology,
                                            do.main.types=do.main.types,
                                            variable.genes=variable.genes,
                                            fine.tune=fine.tune,numCores = numCores),
                       SingleR.CreateObject(sc.data.gl,mouse.rnaseq,clusters,
                                            species,citation,technology,
                                            do.main.types=do.main.types,
                                            variable.genes=variable.genes,
                                            fine.tune=fine.tune,numCores = numCores)
            )
        } else if (species == 'Human') {
            #if(!exists('hpca'))
            #  data ('HPCA')
            #if (!exists('blueprint_encode'))
            #  data('Blueprint_Encode')
            res = list(SingleR.CreateObject(sc.data.gl,hpca,clusters,species,
                                            citation,technology,
                                            do.main.types = do.main.types,
                                            variable.genes=variable.genes,
                                            fine.tune=fine.tune,numCores = numCores),
                       SingleR.CreateObject(sc.data.gl,blueprint_encode,
                                            clusters,species,citation,technology,
                                            do.main.types = do.main.types,
                                            variable.genes=variable.genes,
                                            fine.tune=fine.tune,numCores = numCores))
        }
    } else {
        res = lapply(ref.list, FUN=function(x) {
            SingleR.CreateObject(sc.data.gl,x,clusters,species,citation,technology,
                                 do.main.types=do.main.types,
                                 variable.genes=variable.genes,fine.tune=fine.tune,
                                 numCores = numCores)
        })
    }
    
    singler$singler = res
    
    if (do.signatures==TRUE) {
        signatures = calculateSingScores(sc.data.gl,species=species)
        singler$signatures = signatures
        
    }
    
    if (species == 'Human') {
        kang = SingleR.CreateKangAnnotations(sc.data.gl)
        singler$other = kang$kang_annotation
    }
    
    singler$meta.data = list(project.name=project.name,orig.ident=orig.ident)
    
    if (reduce.file.size==T) {
        singler = remove.Unnecessary.Data.single(singler)
    }
    
    singler
    
}

```

```{r}
# load("/liulab/asahu/data/ssgsea/xiaoman/getz/SKCM_ICB_SeuratObj.rds")
# write.table(file="/liulab/asahu/data/ssgsea/xiaoman/getz/GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO_corrected.txt", x=icb.expression.matched[1:100,1:100], col.names = T, row.names = T, quote = F, sep="\t")

singler = mySCObject(counts=t(icb.expression.matched), 
                                    project.name = 'GSE120575', min.genes = 100,
technology= 'Smart-seq', species = 'Human', citation = 'Moshe et al. 2018', reduce.file.size = T, variable.genes = 'de', normalize.gene.length = F)

# singler$singler

```

## scMatch : https://github.com/forrest-lab/scMatch

```{r}
library(data.table)
source("/homes6/asahu/project/deeplearning/icb/deepImmune/source.R")
cwd= "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/getz/epoch-52/"
# cwd= "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/getz/best/"
setwd(cwd)
getwd()
dataset.sample.name = fread("/homes6/asahu/project/deeplearning/icb/data/Getz_scRNA//scrna.v4.genes/samples_name.txt")$x
icb.phenotype = fread(sprintf("%s/val_prediction.csv", cwd))
icb.phenotype = icb.phenotype[unlist(icb.phenotype$sample_name) +1]
## correct the headers 
header = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/best_val_0.csv",nrows = 1)
colnames(icb.phenotype) = colnames(header)
icb.phenotype.col.dt = data.table(colnames(icb.phenotype), seq(ncol(icb.phenotype))) 

```


```{r}
load("~/project/deeplearning/icb/data/Getz_scRNA/phenotype_sel.mod.RData")
resp = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/GSE120575_patient_ID_single_cells.txt", skip=19)
xx = paste0("V",1:35)
colnames(resp)[1:35] = xx
resp$V2 = gsub(resp$V2, pattern="-", replacement=".")
# resp.patient
resp.matched=resp[match(phenotype_sel.mod$sample.name, resp$V2)]
response = resp.matched$V6
response.bin = ifelse(response=="Responder", 1, 0)

library(pROC)
cell.types = unique((phenotype_sel.mod$assign.ident.2) )
pre_post = resp$V5
pretreatment.samples = grep(pre_post, pattern="^Pre")
posttreatment.samples = grep(pre_post, pattern="^Post")

```


```{r}
### extra phenotypes 
get.checkpoint.genes = function(){
T_Cell_extra = c("IL10", "IDO", "TGFB1", "TGFB2", "TGFBR1", "TGFBR1", "CD37", "TLR", "Arginase")
APC_2 = c("A2AR", "VISTA", "B7_h3", "PDL1", "PDL2", "CD80", "CD86", "Galectin_9", "Ox40L", "CD40", "B7RP1", "CD70", "HVEM", "GITRL", "TNFSF9", "CD155", "CD112")
T_Cell_1 = c("CTLA4", "TIM3", "OX40", "CD40L", "ICOS", "CD27", "BTLA", "LAG3", "TCR", "KIR", "GITR", "TNFRSF9", "CD226", "TIGIT")
checkpoint.genes = unique(c(T_Cell_extra, APC_2, T_Cell_1))

load(file="/liulab/asahu/data/ssgsea/xiaoman/getz/all.tcga.genes.RData")
setdiff(checkpoint.genes, all.tcga.genes)
all.genes = all.tcga.genes
checkpoint.genes.1 = intersect(checkpoint.genes, all.genes)

#  [1] "IDO"        "TLR"        "Arginase"   "A2AR"       "VISTA"      "B7_h3"      "PDL1"       "PDL2"       "Galectin_9" "Ox40L"      "B7RP1"      "HVEM"       "GITRL"      "CD155"      "CD112"
# [16] "TIM3"       "OX40"       "CD40L"      "TCR"        "KIR"        "GITR"
checkpoint.genes.rescue = c("IDO1", "IDO2", "ARG1", "ARG2", "ADORA2A", "ADORA1", "VSIR", "CD276", "VTCN1", "JAK2", "STAT3", "CD80", "ICOSLG", "ICOS", "PVR", "CD226", "HAVCR2", "CD4", "PRF1", "FOXP3", "CD28", "LCK", "B2M")
pd1.genes = grep("^PDCD1",  all.genes, value=T)
Galectin_9.genes = grep("^LGALS",  all.genes, value=T)
jak.genes =c("JAK1", "JAK2", "JAK3")
stat.genes = grep("^STAT[0-9]",  all.genes, value=T)
TNF.genes  =c(grep("^TNFR",  all.genes, value=T), grep("^TNFS",  all.genes, value=T))
il2.gene = c("IL2", "PTPN2", grep("^IL2R",  all.genes, value=T))
il7.gene = grep("^IL7",  all.genes, value=T)
il4.gene = grep("^IL7",  all.genes, value=T)
il6.gene = grep("^IL6",  all.genes, value=T)
il10.gene = grep("^IL10",  all.genes, value=T)
HAVC.gene =grep("HAVC",  all.genes, value=T)
gzm.genes  =grep("^GZM",  all.genes, value=T)
traf.genes = grep("^TRAF",  all.genes, value=T)
nfk.genes = grep("^NFK",  all.genes, value=T)
cd40.genes = grep("^CD40",  all.genes, value=T)
igh.genes = grep("^IGH",  all.genes, value=T)
cd3.genes = grep("^CD3[A-Z]*$",  all.genes, value=T)
tra.genes = grep("^TR[A-B][C,D,V]",  all.genes, value=T)
kir.genes = grep("^KIR",  all.genes, value=T)
tgf.genes =grep("^TGF",  all.genes, value=T)
antigen.presentation.genes = grep("^HLA",  all.genes, value=T)
traf.genes = grep("^TR[A-B]F",  all.genes, value=T)
serpin.genes = grep("^SERPINB[1-9]$",  all.genes, value=T)
vegf.genes = grep("^VEGF",  all.genes, value=T)
tap.genes = c("TAP1", "TAP2", "TAPBP")

checkpoint.genes.semifinal = unique(c(checkpoint.genes.1, checkpoint.genes.rescue, pd1.genes, Galectin_9.genes, stat.genes, TNF.genes, il2.gene, il7.gene, il4.gene, il6.gene, il10.gene, HAVC.gene, gzm.genes, traf.genes, nfk.genes, cd40.genes, igh.genes, cd3.genes,tra.genes, kir.genes, tgf.genes, antigen.presentation.genes, traf.genes,serpin.genes, vegf.genes))
checkpoint.genes.final = checkpoint.genes.semifinal
setdiff(checkpoint.genes.final, all.genes)
checkpoint.genes.final = intersect(checkpoint.genes.final, all.genes)
checkpoint.genes.final
}
checkpoint.genes = get.checkpoint.genes()
```

## raw expression

```{r}
load("~/project/deeplearning/icb/data/Getz_scRNA/dataset_ssgsea_temp.RData")
load("~/project/deeplearning/icb/data/Getz_scRNA/headers.RData")
icb.expression = t(dataset_ssgsea_temp[,2:16292, with=F])
colnames(icb.expression) =  dataset_ssgsea_temp$gene_name
rownames(icb.expression) =  gsub(unlist(headers[1])[-1], pattern="-", replacement=".")
phenotype_sel.mod$sample.name = gsub(phenotype_sel.mod$sample.name, pattern="-", replacement=".")
length(intersect(rownames(icb.expression), phenotype_sel.mod$sample.name))
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
identical(phenotype_sel.mod$sample.name, dataset.sample.name)
rm(dataset_ssgsea_temp, icb.expression)
```

## plot AUCs 
```{r}
#stored top genes predictive of responder vs. non responders
# load("/liulab/asahu/data/ssgsea/xiaoman/getz/topaucs.genes.RData")
# load("/liulab/asahu/data/ssgsea/xiaoman/getz/topaucs.genes.list.RData")
load("/liulab/asahu/data/ssgsea/xiaoman/getz/chenfei1.marker.topaucs.genes.list.RData")
# save(file="/liulab/asahu/data/ssgsea/xiaoman/getz/chenfei1.marker.topaucs.final.RData", topaucs.final)
load("/liulab/asahu/data/ssgsea/xiaoman/getz/chenfei1.marker.topaucs.genes.RData")
aa = topaucs.genes.list$Monocyte[treat=="Pre"]

```


```{r}
calc.stat.new = function(response.curr, value){
  aa = tryCatch(
    as.numeric(auc(response.curr, value, levels=c(0,1))),
    error = function(e) NA
  )
  bb = tryCatch(
    wilcox.test(value[response.curr==0], value[response.curr==1], levels=c(0,1))$p.value,
    error = function(e) NA
  )
  c(aa,bb)
}

plot.aucs.hist = function(inx, indexes, filename, title, aucs.dt = NULL){
  
  require(ggthemes)
  response.curr = response.bin[inx]
  if(is.null(aucs.dt)){
    cor.monocytes = cor(icb.expression.matched[inx,] ,response.bin[inx])
    genes.sel = order(abs(cor.monocytes),decreasing=T)[1:500]
    gene.select = unique( c(genes.sel, sample.int(length(cor.monocytes), 1000)))
    # aa = auc( response.bin[inx], icb.expression.matched[inx, "HLA-G"]  )
    out = mclapply(gene.select, function(tt){
      value.all = icb.expression.matched[,tt]
      calc.stat.new(response.curr, value.all[inx]) 
    }, mc.cores=32
    )
    
    ###############
    # for each cell type create a figure for comparison of expression 
    ###############
    aucs.dt = do.call(rbind, out)
    aucs.dt = data.table(aucs.dt)
    aucs.dt$marker = dataset_ssgsea_temp$gene_name[gene.select]
    
    aucs.dt[,aucs:=ifelse( V1 < 0.5, 1-V1, V1)]
  }
  aucs.dt = aucs.dt[,.(V1, V2, marker, label, aucs)]
  aucs.dt$label = "gene"
  aucs.dt$alpha = 0.35
  
  out = mclapply(indexes, function(tt){
    value.all = icb.phenotype[[tt]]
    calc.stat.new(response.curr, value.all[inx]) 
  }, mc.cores=32
  )
  di.aucs.dt = do.call(rbind, out)
  di.aucs.dt = data.table(di.aucs.dt)
  di.aucs.dt$marker = gsub(indexes, pattern = ".output$", replacement = "")
  di.aucs.dt$label = "signature"
  di.aucs.dt = di.aucs.dt[!is.na(V1)]
  di.aucs.dt[,aucs:=ifelse( V1 < 0.5, 1-V1, V1)]
  di.aucs.dt$alpha = 0.8
  pre_treatment.aucs = rbind(aucs.dt, di.aucs.dt)
  pre_treatment.aucs = pre_treatment.aucs[order(aucs)]
  setnames(pre_treatment.aucs, "V2", "P")
  pre_treatment.aucs[,logP:=-log10(P)]
  require(ggrepel)
  m1 = di.aucs.dt[which(aucs > 0.7)]
  if(nrow(m1) > 20) m1 = di.aucs.dt[order(aucs,decreasing=T)][1:25]
  if(nrow(m1) < 2) m1 = di.aucs.dt[order(aucs,decreasing=T)[1:5]]
  m2 = aucs.dt[which(aucs > 0.7)]
  if(nrow(m2) > 20) m2 = aucs.dt[order(aucs,decreasing=T)[1:20]]
  if(nrow(m2) < 2) m2 = aucs.dt[order(aucs,decreasing=T)[1:5]]
  
  
  pre_treatment_subset = pre_treatment.aucs[marker %in% c(m1$marker, m2$marker)]
  p = ggplot(pre_treatment.aucs, aes(x = aucs, y = logP)) +
    geom_point(aes(color=as.factor(label), alpha = alpha)) +
    
    theme_minimal(base_size = 12) + theme(legend.position = "bottom") +
    ggthemes::scale_colour_tableau() + 
    labs(x="AUC", y="Significance", title=title)+
    geom_text_repel(
      data = pre_treatment_subset,
      aes(x = aucs, y = logP, label = marker),
      size = 3,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )  
  
  ggsave(p, file=filename, width =7, height = 7)
  p
}

auc.dir = sprintf("%s/aucs",cwd)
dir.create(auc.dir)

pheno.inx.start = icb.phenotype.col.dt[V1=="B_cells_naive.output"]$V2
pheno.inx.end = tail(grep(icb.phenotype.col.dt$V1, pattern="output$"),1)
indexes.output = colnames(icb.phenotype)[pheno.inx.start:pheno.inx.end]
survival.inx.start = icb.phenotype.col.dt[V1=="OS.time.output"]$V2
indexes1 = grep(colnames(icb.phenotype), pattern = "embedding", value = T)
indexes.survival = colnames(icb.phenotype)[seq(survival.inx.start, survival.inx.start+3)]
selected.outputs = indexes = c( indexes1, indexes.output, indexes.survival)
             
require(doMC)
require(foreach)
registerDoMC(cores = 32)
Pre.p = Post.p = All.p = list()
library(parallel)
# out = foreach(cell.type = cell.types) %dopar% { 
for(cell.type in cell.types) { 
  print(cell.type)
  
  cell.type.str = gsub(cell.type, pattern="/", replacement=".or.")
  cell.type.str = gsub(cell.type.str, pattern="-", replacement=".or.")
  inx = intersect(which(phenotype_sel.mod$assign.ident.2==cell.type), pretreatment.samples)
  # aucs.dt = topaucs.genes[treat=="Pre" & label==cell.type]
  aucs.dt = (topaucs.genes.list[[cell.type]])[treat=="Pre"][order(V2)][1:2000]
  Pre.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename =sprintf("%s/pretreatment_%s.pdf", auc.dir, cell.type.str), title= sprintf("Pretreatment %s", cell.type.str), aucs.dt = aucs.dt)
  
  inx = intersect(which(phenotype_sel.mod$assign.ident.2==cell.type), posttreatment.samples)
  aucs.dt = topaucs.genes.list[[cell.type]][treat=="Post"][order(V2)][1:2000]
  Post.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename = sprintf("%s/posttreatment_%s.pdf", auc.dir, cell.type.str), title= sprintf("Posttreatment %s  ", cell.type.str), aucs.dt = aucs.dt)
  
  
  inx =which(phenotype_sel.mod$assign.ident.2==cell.type)
  aucs.dt = topaucs.genes.list[[cell.type]][treat=="All"][order(V2)][1:2000]
  All.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename = sprintf("%s/Alltreatment_%s.pdf", auc.dir, cell.type.str), title= sprintf("All %s  ", cell.type.str), aucs.dt = aucs.dt)
  
}


```

```{r}

pdf(paste0(auc.dir, "/Pre.pdf"), width=20, height=15)
multiplot(plotlist = Pre.p, cols = 3)
dev.off()
```

```{r}
pdf(paste0(auc.dir,"/Post.pdf"), width=15, height=15)
multiplot(plotlist = Post.p, cols = 3)
dev.off()
pdf(paste0(auc.dir,"/All.pdf"), width=15, height=15)
multiplot(plotlist = All.p, cols = 3)
dev.off()
```
## finding direction of plot 
```{r}
plot.biauc <- function(ps, dir, indexes, cell.types, mat, ncol=3) {
  require(ggrepel)
  names.ps = names(ps)
  biauc.ps = dts = list() 
  for (tt in seq_along(ps) ) {
    p = ps[[tt]]$data
    name = p$name = names.ps[tt]
    title = gsub(ps[[tt]]$label$title, pattern = " ", replacement =  ".")
    p$title = title 
    xx = title
    curr.marker = p$marker
    responder.index.curr = intersect(indexes, which(cell.types==name & response=="Responder"))
    nonresponder.index.curr = intersect(indexes, which(cell.types==name & response!="Responder"))
    auc.pos = colMeans(mat[responder.index.curr,curr.marker], na.rm = T) > colMeans(mat[nonresponder.index.curr,curr.marker], na.rm = T)
    p[,biaucs:=ifelse(auc.pos, aucs-0.5, -(aucs-0.5))]
    dts[[name]] = p 
    
    di.aucs.dt = p[label=="signature"]
    aucs.dt = p[label=="gene"]
    
    m1 = di.aucs.dt[which(aucs > 0.7)]
    if(nrow(m1) > 20) m1 = di.aucs.dt[order(aucs,decreasing=T)][1:25]
    if(nrow(m1) < 2) m1 = di.aucs.dt[order(aucs,decreasing=T)[1:5]]
    m2 = aucs.dt[which(aucs > 0.7)]
    if(nrow(m2) > 20) m2 = aucs.dt[order(aucs,decreasing=T)[1:20]]
    if(nrow(m2) < 2) m2 = aucs.dt[order(aucs,decreasing=T)[1:5]]
    
    p[,sub:=ifelse(marker %in% c(m1$marker, m2$marker), 1, 0)]
    p1 = ggplot(p, aes(x = biaucs, y = logP)) +
      geom_point(aes(color=as.factor(label), alpha = alpha)) +
      
      theme_minimal(base_size = 12) + theme(legend.position = "bottom") +
      labs(x="AUC - 0.5", y="Significance", title=title)+
      geom_text_repel(
        data = subset(p, sub==1 ),
        aes(x = biaucs, y = logP, label = marker),
        size = 3,
        box.padding = unit(0.35, "lines"),
        point.padding = unit(0.3, "lines")
      )  
    
    biauc.ps[[name]] = p1
    filename = sprintf("%s/bi_aucs_%s.pdf",dir, title)
    ggsave(p1, file=filename, width =7, height = 7)
    
  }
  
  dts = do.call(rbind, dts)
  
  title1 = strsplit(title, split = "\\.")[[1]][1]
  auc.p = ggplot(dts, aes(x = aucs, y = logP)) +
    geom_point(aes(color=as.factor(label), alpha = alpha)) +
    
    theme_minimal(base_size = 12) + theme(legend.position = "bottom") +
    labs(x="AUC", y="Significance", title=title1)+
    geom_text_repel(
      data = subset(dts, sub==1 ),
      aes(x = aucs, y = logP, label = marker),
      size = 3,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )   + facet_wrap(~name, ncol=ncol, scales = "free")
  filename = sprintf("%s/aucs_%s.pdf",dir, title1)
  ggsave(auc.p, file=filename, width =20, height = 15)
  
    
  biauc.p = ggplot(dts, aes(x = biaucs, y = logP)) +
    geom_point(aes(color=as.factor(label), alpha = alpha)) +
    
    theme_minimal(base_size = 12) + theme(legend.position = "bottom") +
    labs(x="AUC - 0.5", y="Significance", title=title1)+
    geom_text_repel(
      data = subset(dts, sub==1 ),
      aes(x = biaucs, y = logP, label = marker),
      size = 3,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )   + facet_wrap(~name, ncol=ncol, scales = "free")
  filename = sprintf("%s/biaucs_%s.pdf",dir, title1)
  ggsave(biauc.p, file=filename, width =20, height = 15)
  
  list(biauc.p =biauc.p, biauc.ps=biauc.ps,auc.p=auc.p,  dts=dts)
}
```

```{r}
# options(error = recover)

biauc.dir = sprintf("%s/biaucs",cwd)
dir.create(biauc.dir)
undebug(plot.biauc)
icb.phenotype.output = icb.phenotype[,selected.outputs, with=F]
setnames(icb.phenotype.output, colnames(icb.phenotype.output), gsub(colnames(icb.phenotype.output), pattern=".output$", replacement="") )
mat = cbind(icb.expression.matched, as.matrix(icb.phenotype.output))
Pre.all.aucs = plot.biauc(ps=Pre.p,  mat=mat, dir = biauc.dir, cell.types = phenotype_sel.mod$assign.ident.2, indexes =pretreatment.samples)
Pre.all.aucs[["auc.ps"]] = Pre.p
Post.all.aucs = plot.biauc(ps=Post.p,  mat=mat, dir = biauc.dir, cell.types = phenotype_sel.mod$assign.ident.2, indexes =posttreatment.samples)
Post.all.aucs[["auc.ps"]] = Post.p
All.all.aucs = plot.biauc(ps=All.p,  mat=mat, dir = biauc.dir, cell.types = phenotype_sel.mod$assign.ident.2, indexes = seq(nrow(mat)))
All.all.aucs[["auc.ps"]] = All.p


```

```{r}
epoch_52 = list(Pre.all.aucs, Post.all.aucs,All.all.aucs)
save(file=sprintf("%s/AUC.RData",biauc.dir),epoch_52)
```


```{r}
library(heatmap3)
plot.heatmap = function(dat, filename, height =7, width =7){
  hc = hclust(as.dist(1-cor(dat, method="spearman", use="pairwise.complete.obs")), method="complete")
  hr = hclust(as.dist(1-cor(t(dat), method="spearman", use="pairwise.complete.obs")), method="complete")
  
  require(heatmap3)
  heatmap3(dat, Rowv=as.dendrogram(hr),  Colv=as.dendrogram(hc), scale="none", balanceColor=T, showRowDendro=F ,   showColDendro=F)
  
}

tcga.phenotype = fread(sprintf("%s/tcga_val_prediction.csv", cwd))
tcga.dataset = fread("~/project/deeplearning/icb/data/tcga/scrna.v4.genes/dataset.txt")
tcga.phenotype = tcga.phenotype[unlist(tcga.phenotype$sample_name) +1]
# tcga.phenotype.skcm = tcga.phenotype[which(tcga.dataset$cancertype=="SKCM"),]
col.pheno = colnames(tcga.phenotype)
embedding.matrix = as.matrix(tcga.phenotype[,indexes1,with=F])
pheno.matrix = as.matrix(tcga.phenotype[,indexes.output,with=F])
colnames(pheno.matrix) = gsub(colnames(pheno.matrix), pattern=".output$", replacement = "")
aa = cor(embedding.matrix, pheno.matrix, use="pairwise.complete.obs")
# plot.heatmap(aa, filename="/liulab/asahu/data/ssgsea/xiaoman/getz/dca/emb_pheno_skcm.pdf") 
library(ComplexHeatmap)
    Heatmap(aa, 
            name = "Embedding vs Phenotypes", #title of legend
            column_title = "Variables", row_title = "Samples",
            row_names_gp = gpar(fontsize = 7), # Text size for row names
            column_names_gp = gpar(fontsize = 6) # Text size for row names
    )
```


## Characterization of embedding 
1. Correlation 
2. High /low wilcox test
a. Immune factors
b. genes 

```{r}
library(Hmisc)
# tcga.dataset$sample_name= seq(0, nrow(tcga.dataset)-1)

# cor.curr = calc.cor(x=embedding.matrix, y=pheno.matrix)
cor.curr =WGCNA::corAndPvalue(x=embedding.matrix,y=pheno.matrix)

# aa = data.table(temp.corr$t[2,], temp.corr$cor[2,], colnames(temp.corr$cor))[order(abs(V2))]
# cor.curr = cor(x=embedding.matrix, y=pheno.matrix, method = "spearman")
agg.cor.volcano = function(cors, Ps, markers, filename=NULL, title="emb"){
  
  df.val = data.table(val=cors, P=Ps, marker = markers)
  # setnames(df.cors, "V2", "P")
  df.val[,logP:=-log10(P)]
  require(ggrepel)
  m2 = df.val[which(P < 1E-20 & abs(val) > 0.3)]
  if(nrow(m2) > 20) m2 = df.val[order(P)[1:5]]
  if(nrow(m2) < 2) m2 = df.val[order(P)[1:2]]
  thr = ifelse(sum(df.val$P < 1E-3) > 5, 1E-3, 1E-2)
  df.val[,Significant:=ifelse(P < thr, "Significant", "Not-significant")]
  df.val$title = title
  # df_subset = df.val[marker %in%  m2$marker]
  df.val[,repel:=ifelse(marker %in%  m2$marker, T, F)]
  df.val
}

agg.tvalue.volcano = function(cors, tvalues, markers, filename=NULL, title="emb"){
  
  df.val = data.table(val=cors, tvalue=tvalues, marker = markers)
  # setnames(df.cors, "V2", "P")
  require(ggrepel)
  m2 = df.val[which(abs(val) > 0.3)]
  if(nrow(m2) > 5) m2 = df.val[order(tvalue,decreasing = T)[1:5]]
  if(nrow(m2) < 2) m2 = df.val[order(tvalue,decreasing = T)[1:2]]
  thr = sort(df.val$tvalue, decreasing = T )[10]
  df.val[,topten:=ifelse(tvalue > thr, "top", "not-top")]
  df.val$title = title
  # df_subset = df.val[marker %in%  m2$marker]
  df.val[,repel:=ifelse(marker %in%  m2$marker, T, F)]
  df.val
}

dfs.val = dfs.subset = plots.p = list()
for (ii in seq_along(rownames(cor.curr$cor))) {
  embedding = rownames(cor.curr$cor)[ii]
  dfs.val[[embedding]] = agg.tvalue.volcano(cors=cor.curr$cor[ii,], tvalues=cor.curr$t[ii,], markers =colnames(cor.curr$cor), title=embedding)
}

dfs.val = do.call(rbind, dfs.val)



  p = ggplot(dfs.val, aes(x = val, y = tvalue)) +
    geom_point(aes(color=as.factor(topten))) +
    scale_color_manual(values=c("grey", "red")) + 
    theme_minimal(base_size = 7) +
    # theme_minimal() + 
    theme(legend.position = "bottom") +
    labs(x="correlation coeff", y="Student t value")+
    #scale_color_discrete()
    geom_text_repel(
      data = subset(dfs.val, repel),
      aes(x = val, y = tvalue, label = marker),
      size = 2,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )  + facet_wrap(~title, ncol=8, scales = "free")
  # 
  # if (!is.null(filename)) {
  #   ggsave(p, file=filename, width =7, height = 7)
  # }
  # p

  
  p2 = ggplot(dfs.val[title=="embedding16"],  aes(x = val, y = tvalue))  +
    geom_point(aes(color=as.factor(topten)),size=1, alpha = 0.7) +
    scale_color_manual(values=c("grey", "red")) + 
    theme_minimal(base_size = 7) +
    # theme_minimal() + 
    theme(legend.position = "bottom") +
    labs(x="correlation coeff", y="Student t value", title="embedding 16")+
    #scale_color_discrete()
    geom_text_repel(
      data = subset(dfs.val[title=="embedding16"], repel),
      aes(x = val, y = tvalue, label = marker),
      size = 3,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    ) 
```

## <span style="color:red"> *TODO* </span> control for cancer types in the analysis 
1. Only in Melanoma 
2. Cancer type specific property
3. Survival analysis
```{r}

```


```{r}

library(dplyr)
avi.dt =dfs.val[repel==TRUE]
avi.dt <- avi.dt %>%
  mutate(angle = 90 * sample(c(0, 1), n(), replace = TRUE, prob = c(60, 40)))
# [title=="embedding3"]
# Create wordcloud 
library("wordcloud")
library("RColorBrewer")
library(ggwordcloud)
       
set.seed(42)
p = ggplot(
  avi.dt,
  aes(
    label = marker, size = tvalue,
    # color = factor(sample.int(10, nrow(love_words_small), replace = TRUE)), ## for random 
    angle = angle,
    color = val
    # angle = angle
  )
) +
  scale_color_gradient2(high="darkred", low="darkgreen", midpoint = 0) + 
  # geom_text_wordcloud_area() +
   geom_text_wordcloud_area(rm_outside = TRUE) +
  scale_size_area(max_size = 4) +
  # scale_size_area(max_size = 24) +
  theme_minimal()  +  facet_wrap(~title, ncol=8) 

# ggsave(p, filename = "~/temp.pdf",width = 16, height = 8)
ggsave(p, filename = sprintf("%s/embedding_correlation_pancancer_uncontrolled.pdf", cwd),width = 16, height = 8)


```

 
## Plot association with immune genes 
```{r}

immunegene.matrix = as.matrix(tcga.phenotype[,paste0(input.output.genes, ".output"),with=F])

colnames(immunegene.matrix) = gsub(colnames(immunegene.matrix), pattern=".output$", replacement = "")
cor.curr =WGCNA::corAndPvalue(x=embedding.matrix,y=immunegene.matrix)
dfs.val = dfs.subset = plots.p = list()
for (ii in seq_along(rownames(cor.curr$cor))) {
  embedding = rownames(cor.curr$cor)[ii]
  dfs.val[[embedding]] = agg.tvalue.volcano(cors=cor.curr$cor[ii,], tvalues=cor.curr$t[ii,], markers =colnames(cor.curr$cor), title=embedding)
}

dfs.val = do.call(rbind, dfs.val)
avi.dt =dfs.val[repel==TRUE]
avi.dt <- avi.dt %>%
  mutate(angle = 90 * sample(c(0, 1), n(), replace = TRUE, prob = c(60, 40)))
set.seed(42)
p = ggplot(
  avi.dt,
  aes(
    label = marker, size = tvalue,
    # color = factor(sample.int(10, nrow(love_words_small), replace = TRUE)), ## for random 
    angle = angle,
    color = val
    # angle = angle
  )
) +
  scale_color_gradient2(high="darkred", low="darkgreen", midpoint = 0) + 
  # geom_text_wordcloud_area() +
   geom_text_wordcloud_area(rm_outside = TRUE) +
  scale_size_area(max_size = 4) +
  # scale_size_area(max_size = 24) +
  theme_minimal()  +  facet_wrap(~title, ncol=8) 

ggsave(p, filename = "~/temp.pdf",width = 16, height = 8)
ggsave(p, filename = sprintf("%s/embedding_immunegenes_correlation_pancancer_uncontrolled.pdf", cwd),width = 16, height = 8)

```


## Create Seurat object

```{r}
library(Seurat)
library(dplyr)
# sco = readRDS("/liulab/asahu/data/ssgsea/xiaoman/getz/SKCM_ICB_SeuratObj.rds")
# sco.cell =  sco[,]
# sco <- RunUMAP(sco, dims = 1:10)
# DotPlot(sco)
#QC
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
sco = CreateSeuratObject(counts =t(icb.expression.matched), project = "nir", min.cells = 3, min.features = 200)
sco[["percent.mt"]] <- PercentageFeatureSet(sco, pattern = "^MT-")
VlnPlot(sco, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# We filter cells that have unique feature counts over 2,500 or less than 200
# We filter cells that have >5% mitochondrial counts
 sco<- subset(sco, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
 sco[["RNA"]]@data =  log(sco[["RNA"]]@counts + 1)
 xx = log(as.matrix(sco[["RNA"]]@counts) +1)
 sco[["RNA"]]@data = xx
 
 # Identification of highly variable features 
 sco <- FindVariableFeatures(sco, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(sco), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(sco)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
CombinePlots(plots = list(plot1, plot2))

# Scaling the data
all.genes <- rownames(sco)
sco <- ScaleData(sco, features = all.genes)

# dimensionality reduction 
sco <- RunPCA(sco, features = VariableFeatures(object = sco))

#clustering 
sco <- FindNeighbors(sco, dims = 1:10)
sco <- FindClusters(sco, resolution = 0.5)

# UMAP
sco <- RunUMAP(sco, dims = 1:10)
dir.create("/liulab/asahu/data/ssgsea/xiaoman/getz/Seurat")
saveRDS(sco, file = "/liulab/asahu/data/ssgsea/xiaoman/getz/Seurat/seuratObj.rds")

# Finding differentially expressed features 
sco.markers <- FindAllMarkers(sco, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
sco.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_logFC)

## heatmap plot 
top10 <- sco.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
p = DoHeatmap(sco, features = top10$gene) + NoLegend()

```
