---
output: html_notebook
author: Avinash Das Sahu
---
# this report test if the scRNA gene clusters are predictive of ICB response in Genentech cohorts.
# response_information
```{r}
library(data.table)
source("/homes6/asahu/project/deeplearning/icb/deepImmune/source.R")
setwd("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/getz/epoch-52/")
getwd()
icb.phenotype = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/getz/epoch-52/val_prediction.csv")
dataset.prefix = "Getz_scRNA_dca"
icb.phenotype = icb.phenotype[unlist(icb.phenotype$sample_name) +1]
## correct the headers 
header = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/best_val_0.csv",nrows = 1)
colnames(icb.phenotype) = colnames(header)

```
```{r}
load("~/project/deeplearning/icb/data/Getz_scRNA/dca/phenotype_data.RData")
resp = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/GSE120575_patient_ID_single_cells.txt", skip=19)
xx = paste0("V",1:35)
colnames(resp)[1:35] = xx
resp$V2 = gsub(resp$V2, pattern="-", replacement=".")
# resp.patient
resp.matched=resp[match(phenotype_sel.mod$patient.name, resp$V2)]
response = resp.matched$V6
response.bin = ifelse(response=="Responder", 1, 0)

library(pROC)
cell.types = unique((phenotype_sel.mod$assign.ident) )
pre_post = resp$V5
pretreatment.samples = grep(pre_post, pattern="^Pre")
posttreatment.samples = grep(pre_post, pattern="^Post")

```

## plot AUCs 
```{r}
#stored top genes predictive of responder vs. non responders
load("/liulab/asahu/data/ssgsea/xiaoman/getz/topaucs.genes.RData")
load("/liulab/asahu/data/ssgsea/xiaoman/getz/topaucs.genes.list.RData")
aa = topaucs.genes.list$Monocyte[treat=="Pre"]

```


```{r}
calc.stat.new = function(response.curr, value){
  aa = tryCatch(
    as.numeric(auc(response.curr, value, levels=c(0,1))),
    error = function(e) NA
  )
  bb = tryCatch(
    wilcox.test(value[response.curr==0], value[response.curr==1], levels=c(0,1))$p.value,
    error = function(e) NA
  )
  c(aa,bb)
}

plot.aucs.hist = function(inx, indexes, filename, title, aucs.dt = NULL){
  
  response.curr = response.bin[inx]
  if(is.null(aucs.dt)){
    cor.monocytes = cor(icb.expression.matched[inx,] ,response.bin[inx])
    genes.sel = order(abs(cor.monocytes),decreasing=T)[1:500]
    gene.select = unique( c(genes.sel, sample.int(length(cor.monocytes), 1000)))
    # aa = auc( response.bin[inx], icb.expression.matched[inx, "HLA-G"]  )
    out = mclapply(gene.select, function(tt){
      value.all = icb.expression.matched[,tt]
      calc.stat.new(response.curr, value.all[inx]) 
    }, mc.cores=32
    )
    
    ###############
    # for each cell type create a figure for comparison of expression 
    ###############
    aucs.dt = do.call(rbind, out)
    aucs.dt = data.table(aucs.dt)
    aucs.dt$marker = dataset_ssgsea_temp$gene_name[gene.select]
    
    aucs.dt[,aucs:=ifelse( V1 < 0.5, 1-V1, V1)]
  }
  aucs.dt = aucs.dt[,.(V1, V2, marker, label, aucs)]
  aucs.dt$label = "gene"
  aucs.dt$alpha = 0.35
  
  out = mclapply(indexes, function(tt){
    value.all = icb.phenotype[[tt]]
    calc.stat.new(response.curr, value.all[inx]) 
  }, mc.cores=32
  )
  di.aucs.dt = do.call(rbind, out)
  di.aucs.dt = data.table(di.aucs.dt)
  di.aucs.dt$marker = gsub(indexes, pattern = ".output$", replacement = "")
  di.aucs.dt$label = "signature"
  di.aucs.dt = di.aucs.dt[!is.na(V1)]
  di.aucs.dt[,aucs:=ifelse( V1 < 0.5, 1-V1, V1)]
  di.aucs.dt$alpha = 0.8
  pre_treatment.aucs = rbind(aucs.dt, di.aucs.dt)
  pre_treatment.aucs = pre_treatment.aucs[order(aucs)]
  setnames(pre_treatment.aucs, "V2", "P")
  pre_treatment.aucs[,logP:=-log10(P)]
  require(ggrepel)
  m1 = di.aucs.dt[which(aucs > 0.7)]
  if(nrow(m1) > 20) m1 = di.aucs.dt[order(aucs,decreasing=T)][1:25]
  if(nrow(m1) < 2) m1 = di.aucs.dt[order(aucs,decreasing=T)[1:5]]
  m2 = aucs.dt[which(aucs > 0.7)]
  if(nrow(m2) > 20) m2 = aucs.dt[order(aucs,decreasing=T)[1:20]]
  if(nrow(m2) < 2) m2 = aucs.dt[order(aucs,decreasing=T)[1:5]]
  
  
  pre_treatment_subset = pre_treatment.aucs[marker %in% c(m1$marker, m2$marker)]
  p = ggplot(pre_treatment.aucs, aes(x = aucs, y = logP)) +
    geom_point(aes(color=as.factor(label), alpha = alpha)) +
    
    theme_minimal(base_size = 12) + theme(legend.position = "bottom") +
    labs(x="AUC", y="Significance", title=title)+
    geom_text_repel(
      data = pre_treatment_subset,
      aes(x = aucs, y = logP, label = marker),
      size = 3,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )  
  
  ggsave(p, file=filename, width =7, height = 7)
  p
}

dir.create("/liulab/asahu/data/ssgsea/xiaoman/getz/dca/aucs/")
setnames(icb.phenotype, 493:503, gsub(colnames(icb.phenotype)[493:503], pattern="V", replacement = "ERV"))
indexes.output = colnames(icb.phenotype)[420:503]
indexes1 = grep(colnames(icb.phenotype), pattern = "embedding", value = T)
indexes = c(indexes1, indexes.output)
require(doMC)
require(foreach)
registerDoMC(cores = 32)
Pre.p = Post.p = All.p = list()
library(parallel)
# out = foreach(cell.type = cell.types) %dopar% { 
for(cell.type in cell.types) { 
  print(cell.type)
  inx = intersect(which(phenotype_sel.mod$assign.ident==cell.type), pretreatment.samples)
  # aucs.dt = topaucs.genes[treat=="Pre" & label==cell.type]
  aucs.dt = (topaucs.genes.list[[cell.type]])[treat=="Pre"][order(V2)][1:2000]
  Pre.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename =sprintf("%s/pretreatment_%s.pdf", cell.type), title= sprintf("Pretreatment %s", cell.type), aucs.dt = aucs.dt)
  
  inx = intersect(which(phenotype_sel.mod$assign.ident==cell.type), posttreatment.samples)
  aucs.dt = topaucs.genes.list[[cell.type]][treat=="Post"][order(V2)][1:2000]
  Post.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename = sprintf("%s/posttreatment_%s.pdf", cell.type), title= sprintf("Posttreatment %s  ", cell.type), aucs.dt = aucs.dt)
  
  
  inx =which(phenotype_sel.mod$assign.ident==cell.type)
  aucs.dt = topaucs.genes.list[[cell.type]][treat=="All"][order(V2)][1:2000]
  All.p[[cell.type]] = plot.aucs.hist(inx, indexes = indexes, filename = sprintf("%s/Alltreatment_%s.pdf", cell.type), title= sprintf("All %s  ", cell.type), aucs.dt = aucs.dt)
  
}

```

```{r}

pdf("/liulab/asahu/data/ssgsea/xiaoman/getz/dca/aucs/Pre.pdf", width=20, height=15)
multiplot(plotlist = Pre.p, cols = 3)
dev.off()
```

```{r}
pdf("/liulab/asahu/data/ssgsea/xiaoman/getz/dca/aucs/Post.pdf", width=15, height=15)
multiplot(plotlist = Post.p, cols = 3)
dev.off()
pdf("/liulab/asahu/data/ssgsea/xiaoman/getz/dca/aucs/All.pdf", width=15, height=15)
multiplot(plotlist = All.p, cols = 3)
dev.off()
```

```{r}
library(heatmap3)
plot.heatmap = function(dat, filename, height =7, width =7){
  hc = hclust(as.dist(1-cor(dat, method="spearman", use="pairwise.complete.obs")), method="complete")
  hr = hclust(as.dist(1-cor(t(dat), method="spearman", use="pairwise.complete.obs")), method="complete")
  
  require(heatmap3)
  heatmap3(dat, Rowv=as.dendrogram(hr),  Colv=as.dendrogram(hc), scale="none", balanceColor=T, showRowDendro=F ,   showColDendro=F)
  
}
tcga.phenotype = fread("~/project/deeplearning/icb/data/tcga/scrna.v2//val_prediction.csv")
tcga.dataset = fread("~/project/deeplearning/icb/data/tcga/scrna.v2/dataset.txt")
tcga.phenotype = tcga.phenotype[unlist(tcga.phenotype$sample_name) +1]
tcga.phenotype.skcm = tcga.phenotype[which(tcga.dataset$cancertype=="SKCM"),]
col.pheno = colnames(tcga.phenotype)
embedding.matrix = as.matrix(tcga.phenotype.skcm[,grep(col.pheno, pattern="embedding"),with=F])
pheno.matrix = as.matrix(tcga.phenotype.skcm[,493:503,with=F])
aa = cor(embedding.matrix, pheno.matrix, use="pairwise.complete.obs")
plot.heatmap(aa, filename="/liulab/asahu/data/ssgsea/xiaoman/getz/dca/emb_pheno_skcm.pdf") 

```
## Checking why indel prediction so low
```{r}
aa = tcga.dataset[,-c(1,111),with=F]
xx = cor(tcga.dataset$Indel.Neoantigens, aa,method = "spearman", use = "pairwise.complete.obs")
xx.dt = data.table(xx[1,], colnames(xx))
xx.dt = xx.dt[order(abs(V1),decreasing = T)]
```

## Characterization of embedding 
1. Correlation 
2. High /low wilcox test
a. Immune factors
b. genes 

```{r}
library(Hmisc)
tcga.dataset$sample_name= seq(0, nrow(tcga.dataset)-1)
# tcga.dataset.curr = tcga.dataset[cancertype=="SKCM"]
tcga.dataset.curr = tcga.dataset
tcga.phenotype.curr = tcga.phenotype[match(tcga.dataset.curr$sample_name,sample_name)]
embedding.matrix = as.matrix(tcga.phenotype.curr[,grep(colnames(tcga.phenotype.curr), pattern="embedding"),with=F])
pheno.matrix = as.matrix(tcga.dataset.curr[,66:137,with=F])

calc.cor = function(x, y){
  op <- options(warn = (-1)) 
  h = cor(x, y, method = "spearman", use = "pairwise.complete.obs")
  options(op) # reset the default value
  aa = (!is.na(x)) + 0
  bb = (!is.na(y)) + 0
  npair = t(aa) %*% bb
  P <- 2 * (1 - pt(abs(h) * sqrt(npair - 2)/sqrt(1 - 
                                                   h * h), npair - 2))
  P[abs(h) == 1] <- 0
  P[is.na(P)] <-1
  list(estimate=h, P = P)
}


cor.curr = calc.cor(x=embedding.matrix, y=pheno.matrix)

# cor.curr = cor(x=embedding.matrix, y=pheno.matrix, method = "spearman")
agg.cor.volcano = function(cors, Ps, markers, filename=NULL, title="emb"){
  
  df.val = data.table(val=cors, P=Ps, marker = markers)
  # setnames(df.cors, "V2", "P")
  df.val[,logP:=-log10(P)]
  require(ggrepel)
  m2 = df.val[which(P < 1E-3 & abs(val) > 0.2)]
  if(nrow(m2) > 20) m2 = df.val[order(P)[1:20]]
  if(nrow(m2) < 2) m2 = df.val[order(P)[1:5]]
  thr = ifelse(sum(df.val$P < 1E-3) > 5, 1E-3, 1E-2)
  df.val[,Significant:=ifelse(P < thr, "Significant", "Not-significant")]
  df.val$title = title
  # df_subset = df.val[marker %in%  m2$marker]
  df.val[,repel:=ifelse(marker %in%  m2$marker, T, F)]
  df.val
}
dfs.val = dfs.subset = plots.p = list()
for (ii in seq_along(rownames(cor.curr$estimate))) {
  embedding = rownames(cor.curr$estimate)[ii]
  dfs.val[[embedding]] = agg.cor.volcano(cors=cor.curr$estimate[ii,], Ps=cor.curr$P[ii,], markers =colnames(cor.curr$estimate), title=embedding)
}

dfs.val = do.call(rbind, dfs.val)



  p = ggplot(dfs.val, aes(x = val, y = logP)) +
    geom_point(aes(color=as.factor(Significant))) +
    scale_color_manual(values=c("grey", "red")) + 
    theme_minimal(base_size = 7) +
    # theme_minimal() + 
    theme(legend.position = "bottom") +
    labs(x="correlation coeff", y="Significance (log10 p-value)")+
    #scale_color_discrete()
    geom_text_repel(
      data = subset(dfs.val, repel),
      aes(x = val, y = logP, label = marker),
      size = 1.5,
      box.padding = unit(0.35, "lines"),
      point.padding = unit(0.3, "lines")
    )  + facet_wrap(~title, ncol=8, scales = "free")
  
  if (!is.null(filename)) {
    ggsave(p, file=filename, width =7, height = 7)
  }
  p
multiplot(plotlist = plots.p[1:9], cols = 3)
library(gridExtra)
library(grid)
library(ggplot2)
library(lattice)
grid.arrange(grobs=plots.p[1:9], ncol=3)

```


## create tsnes 

```{r}
library(Rtsne)
library(ggplot2)
plotSNE <- function(data= data_tsne.merge,col = c(2:114),color.col = "condition",
                    title="t-SNE",size=0.25,do.discrete=T, filename=NULL, perplexity=30, theta=0.5, pca = FALSE, max_iter=5000, num_threads=32, tsne =NULL){
  set.seed(9)
  require(ggplot2)
  
  if(is.null(tsne)) tsne <- Rtsne(as.matrix(data[,col]), check_duplicates = FALSE, 
                                  pca = pca, perplexity=perplexity, theta=theta, dims=2, max_iter = max_iter, num_threads = num_threads)
  
  d_tsne_1 = as.data.frame(tsne$Y)
  d_tsne_1=cbind(d_tsne_1,data[,color.col,drop=F])
  
  ## plotting the results without clustering
  p=ggplot(d_tsne_1, aes(x=V1, y=V2)) +
    geom_point(size=size,aes_string(color=color.col), alpha=0.8) +
    guides(colour=guide_legend(override.aes=list(size=2))) +
    xlab("tSNE_1") + ylab("tSNE_2") +
    ggtitle(label = title) +
    theme_light(base_size=20) +
    theme(axis.text.x=element_blank(),
          axis.text.y=element_blank()) 
  if (do.discrete) {
    p<- p+ scale_colour_brewer(palette = "Set2")
  }
  ##theme(legend.position = "none")
  if(!is.null(filename)) ggsave(file=filename, p)
  list(tsne, p)
}

znorm = function(xx) (xx - mean(xx,na.rm=T))/sd(xx,na.rm=T)

plotSNE.array <- function(data= data_tsne.merge,col = c(2:114), color.cols = "condition",
                          title="t-SNE",size=0.25,do.discrete=T, filename=NULL, perplexity=30, theta=0.5, pca = FALSE, max_iter=5000, normalize=TRUE, num_threads=32){
  set.seed(9)
  
  tsne <- Rtsne(as.matrix(data[,col]), check_duplicates = FALSE, 
                pca = pca, perplexity=perplexity, theta=theta, dims=2, max_iter = max_iter, num_threads = num_threads)
  dt1 = as.data.frame(tsne$Y)
  ps = list()
  for (color.col in color.cols) {
    if(normalize) data[[color.col]] = znorm(data[[color.col]])
    d_tsne_1=cbind(dt1,col=data[[color.col]], shape=data$shape)
    title.curr = sprintf("%s_%s", title, color.col)
    
    ## plotting the results without clustering
    p=ggplot(d_tsne_1, aes(x=V1, y=V2)) +
      geom_point(size=size,aes(color=col, shape=as.factor(shape)), alpha=0.8) +
      scale_color_gradient2(low = "blue", mid = "white",
                            high = "red", space = "Lab" ) + 
      guides(colour=guide_legend(override.aes=list(size=2))) +
      xlab("tSNE_1") + ylab("tSNE_2") +
      ggtitle(label = title.curr) +
      theme_light(base_size=20) +
      theme(axis.text.x=element_blank(),
            axis.text.y=element_blank()) 
    if (do.discrete) {
      p<- p+ scale_colour_brewer(palette = "Set2")
    }
    ##theme(legend.position = "none")
    if(!is.null(filename)) {
      filename.curr = sprintf("%s_%s.pdf", filename, color.col)
      
      ggsave(file=filename.curr, p)
      ps[[color.col]]  = p
    }
  }
  list(d_tsne_1, ps)
}


color.clusters.features <- function(data, cluster,  color.cols = "condition",
                                    title="t-SNE",size=0.25,do.discrete=T, filename=NULL, normalize=TRUE){
  require(viridis)
  
  dt1 = as.data.frame(cluster)
  colnames(dt1) = c("V1", "V2")
  ps = list()
  for (color.col in color.cols) {
    color.col = gsub(color.col, pattern = "-", replacement = ".")
    if(normalize) data[[color.col]] = znorm(data[[color.col]])
    d_cluster_1=cbind(dt1,col=data[[color.col]], shape=data$shape)
    title.curr = sprintf("%s_%s", title, color.col)
    ## plotting the results without clustering
    p=ggplot(d_cluster_1, aes(x=V1, y=V2)) +
      geom_point(size=size,aes(color=col, shape=as.factor(shape)), alpha=0.7) +
      # scale_color_gradient2(low = "blue", mid = "white",
      # high = "red", space = "Lab" ) +
      
      # guides(colour=guide_legend(override.aes=list(size=2))) +
      xlab("Dim1") + ylab("Dim2") +
      ggtitle(label = title.curr) +
      theme_light(base_size=20) +
      theme(axis.text.x=element_blank(),
            axis.text.y=element_blank()) 
    if (do.discrete) {
      p<- p+ scale_colour_brewer(palette = "Set2")
    }else{
      # p <- p+  scale_color_viridis() 
      p <- p+ scale_color_gradientn(colours = heat.colors(20, alpha=0.7, rev=T))
      
    }
    ##theme(legend.position = "none")
    if(!is.null(filename)) {
      filename.curr = sprintf("%s_%s.pdf", filename, color.col)
      
      ggsave(file=filename.curr, p)
    }
    ps[[color.col]]  = p
    
  }
  ps
}


```



```{r}
tsne.dir = "/liulab/asahu/data/ssgsea/xiaoman/getz/dca/"
dir.create(tsne.dir)
tsne.dir = "/liulab/asahu/data/ssgsea/xiaoman/getz/dca/tsnes"
dir.create(tsne.dir)
tsne.dir = "/liulab/asahu/data/ssgsea/xiaoman/getz/dca/tsnes/Monocyte_emb"
dir.create(tsne.dir)
loc=grep(pattern = ".output",colnames(icb.phenotype))
loc2=grep(pattern = "embedding",colnames(icb.phenotype))
d1=icb.phenotype[,c(loc,loc2),with=F]
d2 = resp.matched[,.(V2, V5, V6,V7)]
setnames(d2, 1:4,c("sample.name", "patient.name", "response", "treatment"))
d3 = phenotype_sel.mod[,.(assign.ident, CD274, PDCD1, oxphos_score)]
data_tsne.merge = as.data.frame(cbind(d3, d2, d1))

locx=which(data_tsne.merge$assign.ident=="Monocyte")
all.p = plotSNE(data = data_tsne.merge[locx,], col=58:121, size = 2,do.discrete=F, title="Monocyte Response",
                color.col = "response", perplexity = 30, theta = 0, pca=TRUE, 
                filename=sprintf("%s/%s_%s_final.pdf", tsne.dir, "Monocyte", "response"), max_iter=5000)

```
```{r}
print(all.p[[2]])
```
```{r}
load( "~/project/deeplearning/icb/data/Getz_scRNA/dataset_ssgsea_temp.RData")
load( "~/project/deeplearning/icb/data/Getz_scRNA/headers.RData")
icb.expression = t(dataset_ssgsea_temp[,2:16292, with=F])
colnames(icb.expression) =  dataset_ssgsea_temp$gene_name
rownames(icb.expression) =  gsub(unlist(headers[1])[-1], pattern="-", replacement=".")
phenotype_sel.mod$patient.name = gsub(phenotype_sel.mod$patient.name, pattern="-", replacement=".")
length(intersect(rownames(icb.expression), phenotype_sel.mod$patient.name))
icb.expression.matched = icb.expression[match(phenotype_sel.mod$patient.name, rownames(icb.expression)),]
icb.expression.matched = icb.expression[match(phenotype_sel.mod$patient.name, rownames(icb.expression)),]

```




## find marker genes 


```{r}
library(scater)
filename = "/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/tsne_clustering.pdf"
data.tsne= as.data.frame(all.p[[1]]$Y)
data.exp = t(icb.expression.matched)
# # Defining clusters and markers:
library(scran)
data.DI = data_tsne.merge[locx,58:121]
# data.DI = data.tsne
snn.gr <- buildSNNGraph(t(data.DI))
data.tsne$clust1 = factor(igraph::cluster_walktrap(snn.gr)$membership)
data.tsne.dt = data.table(data.tsne)
data.summ = data.tsne.dt[,.(clust.m=median(V2)),by=clust1]
data.summ = data.summ[order(clust.m)]
data.summ$clust = seq(nrow(data.summ))
data.tsne$clust = as.factor(data.summ[match(data.tsne$clust1, clust1)]$clust)
data.tsne$cell_name = colnames(data.exp)[locx]
p=ggplot(data.tsne, aes(x=V1, y=V2)) +
  geom_point(size=2,aes(color=clust), alpha=0.8) +
  guides(colour=guide_legend(override.aes=list(size=4))) +
  xlab("Dim1") + ylab("Dim2") +
  # ggtitle(label = title) +
  theme_light(base_size=20) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank()) 
# if(!is.null(filename)) ggsave(file=filename, p)

print(p)
ggsave(file=filename, p)

```

# Cell cycle analysis

```{r}
cell.cycle = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/Gad_cell_cycle_score.csv")
data.tsne$cell.cycle = as.factor(cell.cycle[match(data.tsne$cell_name, V1)]$Phase)
p=ggplot(data.tsne, aes(x=V1, y=V2)) +
  geom_point(size=2,aes(color=cell.cycle), alpha=0.8) +
  guides(colour=guide_legend(override.aes=list(size=4))) +
  xlab("Dim1") + ylab("Dim2") +
  # ggtitle(label = title) +
  theme_light(base_size=20) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank()) 
# if(!is.null(filename)) ggsave(file=filename, p)
print(p)
ggsave("/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/tsne_clustering_cell_cycle.pdf",p)

cell.cycle.monocyte = cell.cycle[match(data.tsne$cell_name, V1)]
cell.cycle.monocyte$clust = data.tsne$clust
p=ggplot(cell.cycle.monocyte, aes(x=S.Score, y=G2M.Score)) +
  geom_point(size=2,aes(color=clust), alpha=0.8) +
  guides(colour=guide_legend(override.aes=list(size=4))) +
  # xlab("Dim1") + ylab("Dim2") +
  # ggtitle(label = title) +
  theme_light(base_size=20) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank()) 
ggsave("/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/tsne_clustering_cell_cycle.pdf",p)
p=ggplot(cell.cycle.monocyte, aes(x=S.Score, y=G2M.Score)) +
  geom_point(size=2,aes(color=clust), alpha=0.8) +
  guides(colour=guide_legend(override.aes=list(size=4))) +
  # xlab("Dim1") + ylab("Dim2") +
  # ggtitle(label = title) +
  theme_light(base_size=20) +
  theme(axis.text.x=element_blank(),
        axis.text.y=element_blank())+
  xlim(c(-.27, 0)) +  ylim(c(-.27, 0)) 
```



```{r}
proj="/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/Pagoda";
dir.create(proj)
proj; min.c = 3; min.g = 200; nPCs = 100; nKs = 30
#=========QC========
message("Check gene and cell coverage ...")
# genes x cells matrix
library(pagoda2)
tpmMat = t(icb.expression.matched[locx,]);
tpmMat = Matrix(tpmMat)
nGene = apply(tpmMat, 2, function(x) length(x[x>0]))
nCell = apply(tpmMat, 1, function(x) length(x[x>0]))


cm = tpmMat
par(mfrow=c(1,2), mar = c(3.5,3.5,2.0,0.5), mgp = c(2,0.65,0), cex = 1.0)
hist(log10(colSums(cm)+1),main='molecules per cell',col='cornsilk',xlab='log10(molecules per cell)')
counts <- gene.vs.molecule.cell.filter(cm,min.cell.size=500)
hist(log10(rowSums(counts)+1),main='Molecules per gene',xlab='molecules (log10)',col='cornsilk')
abline(v=1,lty=2,col=2)
hist(log10(rowSums(counts)+1),main='Molecules per gene',xlab='molecules (log10)',col='cornsilk')


counts <- counts[rowSums(counts)>=5,]
rownames(counts) <- make.unique(rownames(counts))
r <- Pagoda2$new(counts,log.scale=TRUE, n.cores=16)
r$adjustVariance(plot=T,gam.k=10)
r$calculatePcaReduction(nPcs=nPCs,n.odgenes=3e3,maxit=1000)
r$makeKnnGraph(k=40,type='PCA',center=T,distance='cosine');
require(igraph)
r$getKnnClusters(method=infomap.community,type='PCA')
M <- 30; r$getEmbedding(type='PCA',embeddingType = 'largeVis', M=M,perplexity=30,gamma=1/M,alpha=1)
r$plotEmbedding(type='PCA',show.legend=F,mark.clusters=T,min.group.size=50,shuffle.colors=F,mark.cluster.cex=1,alpha=0.1,main='clusters (largeVis)')
r$getDifferentialGenes(type='PCA',verbose=T,clusterType='community')

## calculate tsne
r$getEmbedding(type='PCA',embeddingType='tSNE',perplexity=50,verbose=F,n.cores=30)
# plot tsne
r$plotEmbedding(type='PCA',embeddingType='tSNE',show.legend=F,mark.clusters=T,min.group.size=1,shuffle.colors=F,mark.cluster.cex=1,alpha=0.1,main='clusters (tSNE)')

## differential expression analysis
r$getDifferentialGenes(type='PCA',verbose=T,clusterType='community')
de <- r$diffgenes$PCA[[1]][['2']];
r$plotGeneHeatmap(genes=rownames(de)[1:15],groups=r$clusters$PCA[[1]])

```
```{r}

pagoda.exp = r$counts
library(scran)
data.combined = data_tsne.merge
data.combined$sample.name =d2$sample.name
sce = SingleCellExperiment(
  assays = list(counts =t(pagoda.exp)))
sce1 = normalize(sce)
data.tsne.dt = data.table(data.tsne) 
setdiff(rownames(pagoda.exp), data.tsne.dt$cell_name)
data.tsne.dt = data.tsne.dt[match(rownames(pagoda.exp), data.tsne.dt$cell_name)]
data.tsne.dt$response = data.combined[match(data.tsne.dt$cell_name, data.combined$sample.name),]$response
my.clusters = as.numeric(data.tsne.dt$clust)
clust.col <- rainbow(max(my.clusters))
markers <- findMarkers(sce1, data.tsne.dt$clust)


genes.names = rownames(markers[[1]])[1:10]
exp = cbind(data.frame(pagoda.exp[,genes.names, drop=F]), shape=as.factor(data.tsne.dt$response))
ps = color.clusters.features(data = exp, cluster = data.tsne.dt[,.(V1, V2)], color.cols = genes.names, size=2, do.discrete = F, normalize = F)
```



## Evaluate the markers in Genentech data
1. Individual gene expression
2. SSGSEA 
3. Estimate the fraction in each tumors 

```{r}
# library(IMvigor210CoreBiologies)
load("/liulab/asahu/data/ssgsea/xiaoman/Genentech/genentech_cds.RData") # data name is cds
# data(cds)
genentech.exp = counts(cds)
genentech.feat = fData(cds)
genentech.pheno = pData(cds)

calc.stat.new = function(response.curr, value){
  aa = tryCatch(
    as.numeric(auc(response.curr, value, levels=c(0,1))),
    error = function(e) NA
  )
  bb = tryCatch(
    wilcox.test(value[response.curr==0], value[response.curr==1], levels=c(0,1))$p.value,
    error = function(e) NA
  )
  c(aa,bb)
}

genentech.response = genentech.pheno$binaryResponse
levels(genentech.response) = c(1,0)
require(doMC)
require(foreach)
registerDoMC(cores = 8)
markers.auc = foreach(ii = seq(length(markers))) %dopar% {
  aa = markers[[ii]]
  genes.curr = rownames(aa)[1:40] 
  out = lapply(genes.curr, function(gene.curr) {
    exp.curr =genentech.exp[match(gene.curr, genentech.feat$symbol),]
    calc.stat.new(genentech.response, exp.curr)
  })
  out.dt = data.table(do.call(rbind, out))
  out.dt$genes = genes.curr
  out.dt$clust = ii
  out.dt
  
}

markers.auc.dt = do.call(rbind, markers.auc)
markers.auc.dt = markers.auc.dt[order(V1,decreasing =T)]

# proabably use SSGSEA to estiamte the fraction of tumors 

library(ggplot2)
require(ggrepel)
p = ggplot(data=markers.auc.dt, aes(x=as.factor(clust), y = V1)) +
  geom_boxplot() + geom_point() + 
  geom_boxplot(outlier.shape=10, outlier.size=8)  +
  geom_point(position="jitter", size=2) + 
  geom_text_repel(
    data = markers.auc.dt[V1>0.6],
    aes(x=as.factor(clust),  y = V1, label = genes),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  )  
ggsave("/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/Monocytes_marker_genentech_response.pdf")

```





4. Conduct tsne using the Markers to differentiante between responders and non-responders

```{r}

top.markers = foreach(ii = seq(length(markers))) %dopar% {
  aa = markers[[ii]]
  rownames(aa)[1:100] 
}

genes.curr = unique(unlist(top.markers[c(1,7,8)]))

# top.markers = rownames(markers[[1]])[1:10]
genes.curr = intersect(genentech.feat$symbol , genes.curr)
exp.curr =log( genentech.exp[match(genes.curr, genentech.feat$symbol),] +1)
data.curr = as.data.frame(t(exp.curr))
data.curr$response = genentech.pheno$binaryResponse
all.p = plotSNE(data = data.curr, col=seq(length(genes.curr)), size = 4,do.discrete=T, title="Genentech Response",
                color.col = "response", perplexity = 30, theta = 0.1, pca=TRUE, 
                filename="/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/Monocytes_marker_genentech_response_tsne.pdf", max_iter=1000)

all.p[[2]]

data.curr$response1 = genentech.pheno$`Best Confirmed Overall Response`
data.curr$response = ifelse(data.curr$response1=="PR", "PR", "PR-NOT")

all.p = plotSNE(data = data.curr, col=seq(length(genes.curr)), size = 4,do.discrete=T, title="Genentech Response",
                color.col = "response", perplexity = 30, theta = 0.1, pca=TRUE, 
                filename="/liulab/asahu/data/ssgsea/xiaoman/getz/tsnes/Monocyte_emb/Monocytes_marker_genentech_response_tsne.pdf", max_iter=1000)

all.p[[2]]
```


```{r}
library(scater)
monocyte.exp  = data.exp[,locx]
sce = SingleCellExperiment(
  assays = list(counts = monocyte.exp))
ave.counts <- rowMeans(counts(sce))
keep <- rowMeans(counts(sce)) >= 0.2
sum(keep)
sce1 = sce[keep, ]
sce1 = normalize(sce1)
```


```{r}
PipelinePagoda <- function(tpmMat, proj, min.c = 3, min.g = 200, nPCs = 100, nKs = 30)
{
  library(pagoda2)
  
  #=========QC========
  message("Check gene and cell coverage ...")
  nGene = apply(tpmMat, 2, function(x) length(x[x>0]))
  nCell = apply(tpmMat, 1, function(x) length(x[x>0]))
  pdf(paste0(proj,"_QC_Coverage.pdf"),width=8,height=4.5)
  par(mfrow=c(1,2))
  plot(1:ncol(tpmMat),sort(nGene),pch=16,col="blue",ylab="Number of Genes Expressed",xlab="Cells",main="Cell Filter")
  abline(h=min.g,lwd=2,lty=2);text(ncol(tpmMat)/2,min.g+max(nGene)*0.05,paste0("n = ",min.g));legend("topleft",paste0("ave G = ",round(mean(nGene))),box.lty=0)
  plot(1:nrow(tpmMat),sort(nCell),pch=16,col="blue",ylab="Number of Cells Expressed",xlab="Genes",main="Gene Filter")
  abline(h=min.c,lwd=2,lty=2);text(nrow(tpmMat)/2,min.c+max(nCell)*0.05,paste0("n = ",min.c));legend("topleft",paste0("ave C = ",round(mean(nCell))),box.lty=0)
  dev.off()
  
  #=========PCA & tSNE=======
  message("Filter cells, variance shrinkage, PCA and tSNE analysis ...")
  tpmMat <- tpmMat[!duplicated(rownames(tpmMat)),]
  r <- Pagoda2$new(tpmMat,modelType='plain',trim=10,log.scale=T)
  r$adjustVariance(plot=T,do.par=T,gam.k=10)
  r$calculatePcaReduction(nPcs=nPCs,n.odgenes=3e3,maxit=1000)
  r$makeKnnGraph(k=nKs,type='PCA',center=T,distance='cosine');
  require(igraph)
  r$getKnnClusters(method=multilevel.community,type='PCA',name='multilevel')
  r$getEmbedding(type='PCA',embeddingType='tSNE',perplexity=50,verbose=T)
  png(file.path(paste0("tSNE_origIdent_", proj, "_cluster.png")),res=300, width=4, height=4, units = "in")
  r$plotEmbedding(type='PCA',embeddingType='tSNE',show.legend=F,mark.clusters=T,min.group.size=10,shuffle.colors=F,mark.cluster.cex=1,alpha=0.3,main=proj)
  dev.off()
  
  #=========identify marker===========
  message("Find marker genes ...")
  deSets <- get.de.geneset(r, groups = r$clusters$PCA[[1]], prefix = 'de_')
  saveRDS(deSets, file.path(paste0(proj, "_Pagoda_DiffMarkers.rds")))
  
  CellType <- AssignCellTypePagoda(deSets)
  r$clusters$PCA$annotation <- r$clusters$PCA$multilevel
  r$clusters$PCA$multilevel <- plyr::mapvalues(x = r$clusters$PCA$multilevel,
                                               from = levels(r$clusters$PCA$multilevel), to = CellType)
  png(file.path(paste0("tSNE_origIdent_", proj, "_annotated.png")),res=300, width=4, height=4, units = "in")
  r$plotEmbedding(type='PCA',embeddingType='tSNE',show.legend=F,mark.clusters=T,min.group.size=10,shuffle.colors=F,mark.cluster.cex=1,alpha=0.3,main=proj)
  dev.off() 
  
  saveRDS(r, file.path(paste0(proj, "_PagodaObj.rds")))
  return(list(PagodaObj=r, markers = deSets))
}

```



##Seurat

```{r}
tpmMat, proj, min.c = 3, min.g = 200, max.g = 20000, org="hsa",
normalization.method = NULL, do.scale = TRUE, do.center = FALSE, vars.to.regress = c("nUMI"),
dims.use = 1:15, findMarkers = TRUE, marker.use = markers.CIBERSORT, res = 0.6

```

```{r}
SeuratPipeline <- function(tpmMat, proj, min.c = 3, min.g = 200, max.g = 20000, org="hsa",
                           normalization.method = NULL, do.scale = TRUE, do.center = FALSE, vars.to.regress = c("nUMI"),
                           dims.use = 1:15, findMarkers = TRUE, marker.use = markers.CIBERSORT, res = 0.6)
{
  #=========QC========
  message("Check gene and cell coverage ...")
  nGene = apply(tpmMat, 2, function(x) length(x[x>0]))
  nCell = apply(tpmMat, 1, function(x) length(x[x>0]))
  pdf(paste0(proj,"_QC_Coverage.pdf"),width=8,height=4.5)
  par(mfrow=c(1,2))
  plot(1:ncol(tpmMat),sort(nGene),pch=16,col="blue",ylab="Number of Genes Expressed",xlab="Cells",main="Cell Filter")
  abline(h=min.g,lwd=2,lty=2);text(ncol(tpmMat)/2,min.g+max(nGene)*0.05,paste0("n = ",min.g));legend("topleft",paste0("ave G = ",round(mean(nGene))),box.lty=0)
  plot(1:nrow(tpmMat),sort(nCell),pch=16,col="blue",ylab="Number of Cells Expressed",xlab="Genes",main="Gene Filter")
  abline(h=min.c,lwd=2,lty=2);text(nrow(tpmMat)/2,min.c+max(nCell)*0.05,paste0("n = ",min.c));legend("topleft",paste0("ave C = ",round(mean(nCell))),box.lty=0)
  dev.off()
  
  SeuratObj <- CreateSeuratObject(tpmMat, project = proj, min.cells = min.c, min.genes = min.g)
  if(org=="hsa"){
    mito.genes <- grep("^MT-", rownames(SeuratObj@data), value = TRUE)
    ercc.genes <- grep("^ERCC", rownames(SeuratObj@data), value = TRUE)}
  else{
    mito.genes <- grep("^mt-", rownames(SeuratObj@data), value = TRUE)
    ercc.genes <- grep("^ercc", rownames(SeuratObj@data), value = TRUE)}   
  percent.mito <- colSums(SeuratObj@data[mito.genes, ])/colSums(SeuratObj@data)
  percent.ercc <- colSums(SeuratObj@data[ercc.genes, ])/colSums(SeuratObj@data)
  SeuratObj <- AddMetaData(SeuratObj, percent.mito, "percent.mito")
  SeuratObj <- AddMetaData(SeuratObj, percent.ercc, "percent.ercc")
  p1<-VlnPlot(SeuratObj, c("percent.mito","percent.ercc"), nCol = 2)
  ggsave(file.path(paste0(proj,"_QC_Spikein.pdf")), p1, width = 6, height = 4.5)
  
  #=========Filter========
  message("Filter cells and find variable genes ...")  
  SeuratObj <- SubsetData(SeuratObj, subset.name = "percent.mito", accept.high = 0.05)
  SeuratObj <- SubsetData(SeuratObj, subset.name = "percent.ercc", accept.high = 0.05)
  SeuratObj <- FilterCells(object = SeuratObj, subset.names = c("nGene"),
                           low.thresholds = min.g, high.thresholds = max.g) 
  
  SeuratObj <- NormalizeData(object = SeuratObj, normalization.method = normalization.method, scale.factor = 10000)
  SeuratObj <- FindVariableGenes(object = SeuratObj, mean.function = ExpMean, dispersion.function = LogVMR,
                                 x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, do.plot = FALSE)
  if(do.scale) SeuratObj <- ScaleData(object = SeuratObj, vars.to.regress = vars.to.regress, do.center = do.center)
  
  #=========PCA===========
  message("PCA analysis ...")
  SeuratObj <- RunPCA(object = SeuratObj, pc.genes = SeuratObj@var.genes,
                      do.print = FALSE, rev.pca = TRUE)
  # VizPCA(object = SeuratObj, pcs.use = 1:2)
  # PCAPlot(object = SeuratObj, dim.1 = 1, dim.2 = 2)
  # PCHeatmap(object = SeuratObj, pc.use = 1:12, cells.use = 500, do.balanced = TRUE,
  #           label.columns = FALSE, use.full = FALSE)
  # SeuratObj <- JackStraw(object = SeuratObj, num.replicate = 100, display.progress = FALSE)
  # JackStrawPlot(object = SeuratObj, PCs = 1:12)
  p2 = PCElbowPlot(object = SeuratObj )
  ggsave(file.path(paste0(proj,"_PCElbowPlot.pdf")), p2, width = 5, height = 4)
  
  #=========tSNE===========
  message("t-SNE analysis ...")
  SeuratObj <- FindClusters(object = SeuratObj, reduction.type = "pca", dims.use = dims.use,
                            resolution = res, print.output = 0, save.SNN = TRUE)
  SeuratObj <- RunTSNE(object = SeuratObj, dims.use = dims.use, do.fast = TRUE)
  pdf(file.path(paste0("tSNE_origIdent_", SeuratObj@project.name, "_cluster.pdf")),width=5, height=4)
  TSNEPlot(object = SeuratObj, do.label = TRUE, pt.size = 0.5, group.by = paste0("res.",res))
  dev.off()
  pdf(file.path(paste0("tSNE_origIdent_", SeuratObj@project.name, "_primary.pdf")), width=5, height=4)
  TSNEPlot(object = SeuratObj, do.label = TRUE, pt.size = 0.5, group.by = "orig.ident")
  dev.off()
  
  #=========identify marker===========
  cluster.markers <- NULL
  if (findMarkers){
    message("Find marker genes and annotation ...")
    cluster.markers <- FindAllMarkers(object = SeuratObj, only.pos = TRUE, min.pct = 0.1)
    cluster.markers <- cluster.markers[cluster.markers$p_val_adj<0.000001, ]
    
    current.cluster.ids = as.integer(levels(cluster.markers$cluster))
    new.cluster.ids = AssignCellTypeSeurat(cluster.markers, marker.use)
    SeuratObj@meta.data$assign.ident = SeuratObj@ident[rownames(SeuratObj@meta.data)]
    SeuratObj@meta.data$assign.ident = plyr::mapvalues(x = SeuratObj@meta.data$assign.ident,
                                                       from = current.cluster.ids, to = new.cluster.ids)
    pdf(file.path(paste0("tSNE_assignIdent_", SeuratObj@project.name, "_annotated.pdf")), width=6, height=4)
    TSNEPlot(object = SeuratObj, do.label = TRUE, pt.size = 0.5, group.by = "assign.ident")
    dev.off()}
  
  saveRDS(cluster.markers, file.path(paste0(proj, "_DiffMarkers.rds")))
  saveRDS(SeuratObj, file.path(paste0(proj, "_SeuratObj.rds")))
  return(list(SeuratObj = SeuratObj, markers = cluster.markers))
}

```



