---
title: "T-cell receptor clustering"
output: html_notebook
---

## Identify CDR3 regions in T-cells 

```{r}
# list all the files in tar
system("echo $HOME", intern = TRUE)

tar.files = list.files(path="/rna-seq/dbGAP_tar", pattern = "*.tar", full.names = T, recursive = T)
file.list = list()
for(tar.file in tar.files){
    print(tar.file)
        xx = system(sprintf("tar -tvf %s  \'*.Aligned.toTranscriptome.sorted.bam\'", tar.file), intern = T ) 
        print(length(xx))
    file.list[[tar.file]]  = xx
}

out = list()
for(fi in seq_along(file.list)){
    fx = file.list[[fi]]
    # out  = NULL
    if(length(fx) > 0){
    xx = sapply(fx,  function(tt){
        uu  = unlist(strsplit(tt, split= " "))
        uu[length(uu)]
        })
    out[[names(file.list)[fi]]] = data.table(unlist(xx), names(file.list)[fi])
    }
    
}
out.dt = do.call(rbind, out)

```

Run in Kraken 
```{r}
library(data.table)
require(doMC)
require(foreach)
registerDoMC(cores = 32)

files = list.files(path="/rna-seq/trust4/outputs", pattern = "*.cdr3.out", full.names = T, recursive = T)
filenames = gsub("TRUST_", gsub("_cdr3.out", basename(files), replacement = ""), replacement = "")
cdr3.all = foreach( file1=files) %dopar%  fread(file1, header=F)
names(cdr3.all) = filenames 

files = list.files(path="/rna-seq/trust4/outputs", pattern = "*annot.out", full.names = T, recursive = T)
filenames = gsub("TRUST_", gsub("_annot.out", basename(files), replacement = ""), replacement = "")
annot.all = foreach( file1=files) %dopar%  fread(file1, header=F)
names(annot.all) = filenames 

trust4.out = list(cdr3.all, annot.all)
save(file="/liulab/asahu/data/ssgsea/xiaoman/getz/trust4.out.V2.RData", trust4.out)

```

## ICB expression

```{r}
load("~/project/deeplearning/icb/data/Getz_scRNA/dataset_ssgsea_temp.RData")
load("~/project/deeplearning/icb/data/Getz_scRNA/headers.RData")
load("~/project/deeplearning/icb/data/Getz_scRNA/phenotype_sel.mod.RData")
icb.expression = t(dataset_ssgsea_temp[,2:16292, with=F])
colnames(icb.expression) =  dataset_ssgsea_temp$gene_name
rownames(icb.expression) =  gsub(unlist(headers[1])[-1], pattern="-", replacement=".")
phenotype_sel.mod$sample.name = gsub(phenotype_sel.mod$sample.name, pattern="-", replacement=".")
length(intersect(rownames(icb.expression), phenotype_sel.mod$sample.name))
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
icb.expression.matched = icb.expression[match(phenotype_sel.mod$sample.name, rownames(icb.expression)),]
# identical(phenotype_sel.mod$sample.name, dataset.sample.name)
rm(dataset_ssgsea_temp, icb.expression)
```

## Create the features for embedding prediction 

```{r}
load("/liulab/asahu/data/ssgsea/xiaoman/getz/trust4.out.V2.RData")
names(trust4.out) = c("cdr3", "annot")
substr.id = function(tt){
    paste(strsplit(tt, split="_")[[1]][1:3], collapse = "_")
}

trust4.cellid = sapply(names(trust4.out[[1]]), substr.id)
tcell.id = sapply(rownames(icb.expression.matched),  substr.id)
sum(trust4.cellid %in% tcell.id) ## only 3196 
only.tcells = setdiff(tcell.id, trust4.cellid)
common.cellids = intersect(trust4.cellid,tcell.id)
cdr3.tcell = trust4.out$cdr3[match(common.cellids, trust4.cellid)]
sample.name.map =data.table(sample.name.short=common.cellids)
sample.name.map$sample.name = rownames(icb.expression.matched)[match(common.cellids, tcell.id)]
sample.name.map$trust.id = names(trust4.out[[1]])[match(common.cellids, trust4.cellid)]

# cdr3.tcell1 = lapply(cdr3.tcell, function(tt) {
#     if(nrow(tt) > 0) out = tt[V9> 0]
#     out
# })
# xx = unlist(sapply(cdr3.tcell, nrow))
# yy = unlist(sapply(cdr3.tcell1, nrow))
# translate 

# Biostrings::translate(Biostrings::DNAStringSet(trust4.out[[1]][[2]]$V8))
# Biostrings::translate(Biostrings::DNAStringSet("TGTGCCAGCAGTTTCGGGCCGAACACTGAAGCTTTCTTTNNNNNNNACTGAAGCTTTCTT"))
# 
# dna3 <- DNAStringSet(c("ATC", "GCTG", "CGACT"))

# head(names(trust4.out[[1]]))


```


```{r}
library(stringr)
library(stringi)
translate.seq = function(tt) {
    ifelse(stringr::str_length(tt) %% 3 ==0, 
        as.character(Biostrings::translate(Biostrings::DNAStringSet(tt),if.fuzzy.codon = "X")), NA)
}
get.aminoacid = function(CDR3, V , J){
    missing = "X"
    out = NA
    str.len = stringr::str_length(CDR3)
    if(V!="*" & J=="*"){
        CDR3 = substr(CDR3,1, floor(str.len/3) *3)
        out = translate.seq(CDR3)
        out = stringr::str_c( out, missing)
    }else if(V=="*" & J!="*"){
        # CDR3 = stringi::stri_reverse(CDR3)
        CDR3 = substr(CDR3,1 + ( str.len %%3), str.len)
        out = translate.seq(CDR3)
        out = stringr::str_c(missing, out)
        # out = stringi::stri_reverse(out)
    }else if(V!="*" & J!="*" & str.len %%3 == 0){
        out = translate.seq(CDR3)
    }
    out
}
get.aminoacid.vec = Vectorize(get.aminoacid)

my.translate = function(dt, name=NULL, min.len= 6, max.len = 25){
    out = NULL 
    if(nrow(dt)){
    cols = c("consesus_id", "the_index_of_CDR3_in_that_consensus", "V", "J", "C", "CDR1", "CDR2", "CDR3", "score_of_CDR3", "abundance")
    setnames(dt,1:10, cols)
    dt$name = name 
    dt[,aa:=get.aminoacid.vec(CDR3, V , J)]
    dt[,aa.len:=stringr::str_length(aa)]
    dt[,is.alpha:=ifelse(grepl("TRA", V)|grepl("TRA", J)| grepl("TRA", C), T, F)]
    dt[,abs.score:= abundance * (score_of_CDR3+.001)]
    if(!is.na(min.len)) dt = dt[aa.len>= min.len]
    if(!is.na(max.len)) dt = dt[aa.len< max.len]
    dt.alpha = dt[is.alpha==T][abs.score == max(abs.score)]
    if(nrow(dt.alpha) > 1) dt.alpha = dt.alpha[which.min(aa.len)]
    dt.beta = dt[is.alpha==F][abs.score == max(abs.score, na.rm = T)]
    if(nrow(dt.beta) > 1) dt.beta = dt.beta[which.min(aa.len)]
    out = rbind(dt.alpha, dt.beta)
    }
    out
}




```

# ```{r}
# # trust4.filtered = for(dt in trust4.out$cdr3)
#     # uu = my.translate(dt)
# # require(doMC)
# # require(foreach)
# # registerDoMC(cores = 32)
# # trust4.filtered = foreach(dt=trust4.out$cdr3) %dopar%{
# #     my.translate(dt)
# # }
# ## convert to unicode 
# ```

## run in data table to make it faster 
```{r}

trust4.cdr3  =do.call(rbind,  lapply(seq_along(trust4.out$cdr3), function(tt) {
    dt = trust4.out$cdr3[[tt]]
    if(nrow(dt) ==0 ){ 
        dt = NULL
    }else{
        dt$cellid = names(trust4.out$cdr3)[[tt]]
    }
    dt
}))
min.len = 7*3
max.len =25*3
cols = c("consesus_id", "the_index_of_CDR3_in_that_consensus", "V", "J", "C", "CDR1", "CDR2", "CDR3", "score_of_CDR3", "abundance")
setnames(trust4.cdr3,1:10, cols)
trust4.cdr3$sample.name.short = sample.name.map[match(trust4.cdr3$cellid,trust.id)]$sample.name.short
trust4.cdr3$sample.name = sample.name.map[match(trust4.cdr3$cellid,trust.id)]$sample.name
trust4.cdr3[,str.len:=stringr::str_length(CDR3)]
trust4.cdr3[,is.alpha:=ifelse(grepl("TRA", V)|grepl("TRA", J)| grepl("TRA", C), T, F)]
trust4.cdr3[,is.beta:=ifelse(grepl("TRB", V)|grepl("TRB", J)| grepl("TRB", C), T, F)]
trust4.cdr3[,abs.score:= abundance * (score_of_CDR3+.001)]
# trust4.cdr3[, cellid.short:=sapply(cellid, substr.id)]
if(!is.na(min.len)) trust4.cdr3 = trust4.cdr3[str.len>= min.len]
if(!is.na(max.len)) trust4.cdr3 = trust4.cdr3[str.len<= max.len]

tcell.cdr3 = trust4.cdr3[is.alpha|is.beta]
xx = trust4.cdr3[!(is.alpha|is.beta)]
tcell.cdr3$aa = tcell.cdr3[,{
    start = ifelse(V=="*", 1 + ( str.len %%3), 1)
    end = ifelse(J=="*", floor(str.len/3) *3, str.len)
   aa = translate.seq(stringr::str_sub(CDR3,start, end))
   ifelse(V=="*", stringi::stri_c("X", aa), 
          ifelse(J=="*", stringi::stri_c("X", aa), aa))
}]
tcell.cdr3[,aa.len:=stringr::str_length(aa)]
tcell.cdr3 = tcell.cdr3[!is.na(aa.len)]
tcell.cdr3 = tcell.cdr3[!grepl(pattern="\\*", tcell.cdr3$aa)]
tcell.cdr3 = tcell.cdr3[aa.len <=22]
# ww = tcell.cdr3[1:100]
# ww[,aa:=mapply(get.aminoacid,CDR3, V , J)]
# 
# ww[,start:=ifelse(V=="*", 1 + ( str.len %%3), 1)]
# ww[,end:=ifelse(J=="*", floor(str.len/3) *3, str.len)]
# cdr3.new = ww[,translate.seq(stringr::str_sub(CDR3,start, end))]
# tt = ww[,{
#     start = ifelse(V=="*", 1 + ( str.len %%3), 1)
#     end = ifelse(J=="*", floor(str.len/3) *3, str.len)
#    translate.seq(stringr::str_sub(CDR3,start, end))
# }]

# trust4.cdr3.copy = copy(trust4.cdr3)
# 
tcell.cdr3.filtered = tcell.cdr3[, .SD[abs.score == max(abs.score)],
                                 by=.(is.alpha, cellid)][, .SD[which.min(str.len)], 
                                                         by = .(is.alpha, cellid)]
tcell.cdr3.beta = tcell.cdr3.filtered[is.beta==T & aa.len <=20 ][sample.name.short %in% tcell.id ][abs.score>=1]

# tcell.cdr3.beta[,aa.short]


```

## "Exhausted/Cellcytle T"
```{r}

data.ex = data.table(data.curr[data.curr$assign.ident.2=="Exhausted/Cellcytle T",])
trust4.cdr3[,samp.id:=sapply(cellid , substr.id)]
data.ex[,samp.id:=sapply(sample.name , substr.id)]
trust4.cdr3.ex =trust4.cdr3[samp.id %in% data.ex$samp.id]

tcell.cdr3.beta[,samp.id:=sapply(cellid , substr.id)]
sum(tcell.cdr3.beta$cellid.short %in% data.ex$samp.id)
```


## one hot encoding
```{r}

## convert aa sequence to column
## convert each to the ordered factors
## get one hot

# xx = stringi::stri_pad(tcell.cdr3.beta$aa,  pad = "X", width = 20, side="right")
# xx =do.call(rbind, strsplit(xx, split=""))
# xx[xx=="X"] = NA
# aa.uniq = sort(unique(c(xx)))
# aa.uniq = aa.uniq[!is.na(aa.uniq)]
# xx = data.table(xx)
# xx =xx[, lapply(.SD, factor, levels=aa.uniq)]
# aa.one = mltools::one_hot(xx, sparsifyNAs = T)
# tt = colSums(aa.one)
# aa.one.one = colSums(aa.one) 
# sum(aa.one.one >=10)
# aa.one.sel = aa.one[,aa.one.one >=10, with=F]
aa.one.sel = get_one_hot_aa(tcell.cdr3.beta$aa, width = 20)
```

## Create datasets 
Dataset is ordered as : 
1. First column is "fake" cancertype :) 
1. One hot of AA (400)
2. 64 embeddings
3. Immune factors 
4. Immune genes (deepImmune outputs)
5. Immune gene (actual expression)
5. All genes 

```{r}

cwd1= "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/nopipeline_vae_20190819-161104/epoch-142/"
dataset.sample.name = fread("/homes6/asahu/project/deeplearning/icb/data/Getz_scRNA//scrna.v4.genes/samples_name.txt")$x
icb.phenotype = fread(sprintf("%s/val_prediction.csv", cwd1))
icb.phenotype = icb.phenotype[unlist(icb.phenotype$sample_name) +1]
header1 = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/nopipeline_vae_20190819-161104/best_val_0.csv",nrows = 1)
colnames(icb.phenotype) = colnames(header1)
icb.phenotype.col.dt = data.table(colnames(icb.phenotype), seq(ncol(icb.phenotype))) 

load("~/project/deeplearning/icb/data/Getz_scRNA/phenotype_sel.mod.RData")
resp = fread("/liulab/asahu/data/ssgsea/xiaoman/getz/GSE120575_patient_ID_single_cells.txt", skip=19)
xx = paste0("V",1:35)
colnames(resp)[1:35] = xx
resp$V2 = gsub(resp$V2, pattern="-", replacement=".")
# resp.patient
resp.matched=resp[match(phenotype_sel.mod$sample.name, resp$V2)]
response = resp.matched$V6
response.bin = ifelse(response=="Responder", 1, 0)

library(avinash)
library(ggplot2)
output.inx=grep(pattern = ".output",colnames(icb.phenotype))
embedding.inx=grep(pattern = "embedding",colnames(icb.phenotype), value = T)
# d1=icb.phenotype[,c(1, 10:338, 942:1005), with=F]
d1=icb.phenotype
d2 = resp.matched[,.(V2, V5, V6,V7)]
setnames(d2, 1:4,c("sample.name", "patient.name", "response", "treatment"))
d3 = phenotype_sel.mod[,.(assign.ident.2)]
data.curr1 = cbind(d3, d2, d1)

```



```{r}

data.curr1.match = data.curr1[match(tcell.cdr3.beta$sample.name,sample.name)]
if(FALSE){
sco.curr = readRDS("/liulab/asahu/data/ssgsea/xiaoman/getz/Seurat/seuratObj.rds")
# tcell.id = sapply(colnames(sco.curr),  substr.id) this is for expression from sco.curr
# exp.match = t(sco.curr[["RNA"]]@scale.data)
# exp.match = exp.match[match(tcell.cdr3.beta$cellid.short, tcell.id),]
exp.match = icb.expression.matched[tcell.cdr3.beta$sample.name ,]
exp.match.genes = colnames(exp.match)
# rm(sco.curr)
}
sample.name  = cbind(tcell.cdr3.beta[,.(sample.name)], data.curr1.match[,.(patient.name)])


embedding.inx = grep("embedding", colnames(data.curr1.match))
surv.inx = 70:73
if.genes.output.inx = 74:402
if.genes.output= colnames(data.curr1.match)[if.genes.output.inx]
if.factor.inx = 403:537
# if.genes = intersect(gsub(".output", if.genes.output, replacement = ""), exp.match.genes)
if.all.genes = 15:343
# dataset.tcr = cbind("SKCM", aa.one.sel,
# data.curr1.match[,.(response.bin, UMAP1, UMAP2, G2M.Score, S.Score, diffusionmap1, diffusionmap2, dpt)], 
# data.curr1.match[,c(embedding.inx, surv.inx, if.factor.inx, if.genes.output.inx), with=F], exp.match[,if.genes]) 
# exp.match[,setdiff(exp.match.genes, if.genes)])
dataset.tcr = cbind("SKCM", aa.one.sel,
                    data.curr1.match[,c(embedding.inx,  if.all.genes), with=F])                     
output.dir = "~/project/deeplearning/icb/data/Getz_scRNA/TCR.AA.patient.independent/"
write.dataset(output.dir = output.dir, dataset = dataset.tcr, sample.name = sample.name, use.sample = sample.name$patient.name, frac = 0.6)
file.copy("~/project/deeplearning/icb/data/Getz_scRNA/TCR.AA.V2/params.json", output.dir)
file.copy("~/project/deeplearning/icb/data/Getz_scRNA/TCR.AA.V2/datasets_tsne_list.txt", output.dir)
file.copy("~/project/deeplearning/icb/data/Getz_scRNA/TCR.AA.V2/datasets_test_list.txt", output.dir)

save(file = sprintf("%s/tcell.cdr3.beta.RData", output.dir), tcell.cdr3.beta)
dataset.cols = data.table(colnames(dataset.tcr), seq(ncol(dataset.tcr)))

# uu = cor(aa.one.sel, dataset.tcr[,253:1119, with=F] )


```

CUDA_VISIBLE_DEVICES=0 python evaluate.py  --data_dir  ../data/oxphos/scrna.v4.genes/datasets_tsne_list.txt --model_dir ../data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/. --restore_file  ../data/tcga/scrna.v4.genes/tensorboardLog/no_pipeline_20190724-201216/epoch-52.pth.tar

## unsupervised clustering of TCRs from it autoencoder ouput
```{r}
tcr.emb = fread("~/project/deeplearning/icb/data/Getz_scRNA/TCR.AA/tensorboardLog/autoencoder_20190820-140204/best.csv")

# vv = cor(tcr.emb, dataset.tcr[,253:1119, with=F] )
 xx = tcr.emb[, 504:511,with=F]
 setnames(xx, 1:8, paste0("tcr.",colnames(xx)))
data.tcr = cbind(dataset.tcr,xx)
data.tcr$response = data.curr.match$response
tcr.embedding.inx = grep("tcr.embedding", colnames(data.tcr))

n_neighbors =20; learning_rate =1; min_dist = 0.01; pca = NULL
umap.all.p = plotUMAP(data = as.data.frame(data.tcr), col=c(tcr.embedding.inx), size=1.5, do.discrete=F, 
                                n_neighbors = n_neighbors, learning_rate = learning_rate, min_dist = min_dist, pca=pca,
                      color.col = "response", 
                      filename=NULL, n_epochs = 500, metric = "cosine")


```


### Top 100 clonal tcrs + all input to transcriptomic deepImmune 
```{r}
# save(file="/liulab/asahu/data/ssgsea/xiaoman/aa.uniq.RData", aa.uniq)


```




### Get post filenames of Tregs and TCRs

```{r}
files1 = list.files(path="/liulab/asahu/projects/bcr/", pattern = "*.toTranscriptome.sorted.bam", full.names = T, recursive = T)
# files1 = fread("/liulab/asahu/all.files.txt")
filenames = gsub("TRUST_", gsub(".Aligned.toTranscriptome.sorted.bam", basename(files1), replacement = ""), replacement = "")


load("/liulab/asahu/files.RData") # out.dt
out.dt$V1
out.dt[,filenames:= gsub("TRUST_", gsub(".Aligned.toTranscriptome.sorted.bam", basename(V1), replacement = ""), replacement = "")]

treg.id = sample.name.map[sample.name %in% data.curr1[assign.ident.2=="RegulatoryTcell"]$sample.name]
treg.files = out.dt[filenames %in%  treg.id$trust.id]
write.table(file="/liulab/asahu/treg.files.txt", x=treg.files, sep="\t", row.names = F, col.names=T)

cd8t.id = sample.name.map[sample.name %in% data.curr1[assign.ident.2=="Cytotoxicity CD8T"]$sample.name]
cd8t.files = out.dt[filenames %in%  cd8t.id$trust.id]
write.table(file="/liulab/asahu/cd8t.files.txt", x=cd8t.files, sep="\t", row.names = F, col.names=T)

```


```{r}

```

