---
title: "R Notebook"
output: html_notebook
---

## Read differential expressed genes
```{r}
aa = fread("/liulab/xmwang/oxphos_proj/loading_data/cistrome/differential expression of genes in all datasets.txt")
aa[,P_value:=as.numeric(P_value)]
aa.sub = aa[which(P_value < 1E-2)][(Gene %in% genes.analyzed)]
# library(reshape2)
aa.mat = data.table::dcast(data = aa.sub, Sample_ID~Gene,value.var = "Log2FC", fill=0)

missing.genes =  setdiff(genes.analyzed, colnames(aa.mat))
missing.genes.mat =  matrix(NA, nrow=length(aa.mat$Sample_ID), ncol= length(missing.genes), dimnames = list(aa.mat$Sample_ID, missing.genes))
tfko.mat = cbind(aa.mat, missing.genes.mat)

chemo.dat = readRDS("/homes6/asahu/liulab_home/data/ssgsea/xiaoman/wubin/ChemoTherapy_AllGenes_LogFC_0929.rds")
chemo.mat = t(chemo.dat[match(colnames(tfko.mat)[-1],rownames(chemo.dat)),])
chemo.mat = data.table(Sample_ID=rownames(chemo.mat), chemo.mat)
setnames(chemo.mat, seq(ncol(chemo.mat)), colnames(tfko.mat))

targeted.dat = readRDS("/homes6/asahu/liulab_home/data/ssgsea/xiaoman/wubin/TargetedTherapy_AllGenes_LogFC_0929.rds")
targeted.mat = t(targeted.dat[match(colnames(tfko.mat)[-1],rownames(targeted.dat)),])
targeted.mat = data.table(Sample_ID=rownames(targeted.mat), targeted.mat)
setnames(targeted.mat, seq(ncol(targeted.mat)), colnames(tfko.mat))

all.mat = rbind(rbind(tfko.mat,chemo.mat),targeted.mat)

setcolorder(all.mat, c("Sample_ID", genes.analyzed))

write.table(file = "/homes6/asahu/liulab_home/data/ssgsea/xiaoman/wubin/combined.drug.log2fc.txt", x = all.mat, quote = F, sep = "\t", row.names = F, col.names = T)

```
## write a directory for xin 
file that was transferred to xin : ll ../data/tcga/tf/binarize/tensorboardLog/Neo_20200217-233533/last.pth.tar
```{r}
tcga.dataset.curr = tcga.dataset[cancertype=="SKCM"]
output.dir = "/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/skcm"
tcga.samplename = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/samples_name.txt")$x
sample.name.curr = tcga.samplename[tcga.dataset$cancertype=="SKCM"]
write.dataset(output.dir =output.dir , dataset = tcga.dataset.curr, sample.name = sample.name.curr)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/tensorboardLog/Neo_20200217-233533/params.json", output.dir)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/datasets_test_list.txt", output.dir)
```
```{r}
tcga.dataset.curr = tcga.dataset[cancertype=="LUSC"]
output.dir = "/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/lusc"
tcga.samplename = fread("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/samples_name.txt")$x
sample.name.curr = tcga.samplename[tcga.dataset$cancertype=="LUSC"]
write.dataset(output.dir =output.dir , dataset = tcga.dataset.curr, sample.name = sample.name.curr)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/tensorboardLog/Neo_20200217-233533/params.json", output.dir)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/datasets_test_list.txt", output.dir)
```


## Find overlap of genes in ICB and TCGA dataset
```{r}
icb.files = list.files(path="/homes6/asahu/liulab_home/data/links/data_private/ICB/expression", full.names =T)
genes.icb = lapply(icb.files, function(tt) {
	genes = tt %>% fread(select=1) %>% unlist %>% c
	if( sum(grepl("^[0-9]", genes)) > 100 ){
		aa = genes
		gene.eg.dt = clusterProfiler::bitr(aa, toType="SYMBOL", fromType="ENTREZID", OrgDb="org.Hs.eg.db")
		gene.eg.dt1 = gene.eg.dt[match(aa,gene.eg.dt$ENTREZID),]
		genes = gene.eg.dt1$SYMBOL
	}
	genes  
})

icb.phenotypes = lapply(icb.files, function(tt) {
	sprintf( "/homes/asahu/liulab_home/data/links/data_private/ICB/follow_up/%s", tt %<>% basename)%>%
	 fread
})
out = data.table(sapply(genes.icb, length),
sapply(icb.phenotypes, nrow))
genes.overlap =rowData(project.exp)$external_gene_name
for (genes.curr  in genes.icb[-c(1,2)]) {
	genes.overlap %<>% intersect(genes.curr)
	print(length(genes.overlap))

}

```
## Write TCGA dataset with nanostring output 
```{r}
project.exp = readRDS("/homes6/asahu/liulab_home/data/tcga/TCGA_gdc/tcgabiolinks/all.tcga.expression.RDS")
library(SummarizedExperiment)
project.exp.mat = assay(project.exp) %>%
set_rownames(rowData(project.exp)$external_gene_name) %>% t %>% 
scale(., scale=T, center=T) %>% t 
colnames(project.exp.mat) %<>% substring(.,1,16)
project.sample = colnames(project.exp.mat)

tcga.dataset = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.pcs/dataset.txt")
tcga.sample = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.pcs/samples_name.txt")
tcga.sample[,sample.name:=gsub("\\.", x, replacement="-")]
setdiff(tcga.sample$sample.name, project.sample)
if.start.inx = which(colnames(tcga.dataset)== "B_cells_naive")
dataset.col = colnames(tcga.dataset)
surv.factors = tcga.dataset[,seq(if.start.inx, ncol(tcga.dataset)), with =F]

exp.tcga = project.exp.mat[genes.overlap, tcga.sample$sample.name] %>% t 
dataset.curr = cbind(cancertype=tcga.dataset$cancertype, exp.tcga) %>% 
cbind(., surv.factors) 
output.dir = "/homes6/asahu/liulab_home/projects/icb/data/tcga/nanostring"
write.dataset(output.dir =output.dir, dataset = dataset.curr, sample.name = tcga.sample$x)
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/tf/binarize/tensorboardLog/Neo_20200217-233533/params.json", output.dir)
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/tf/binarize/datasets_test_list.txt", output.dir)
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/tf/binarize/datasets_tsne_list.txt", output.dir)
aa = data.table(colnames(dataset.curr) %>% cbind(seq(ncol(dataset.curr))) ) %>% 
set_colnames(c("V1", "V2"))
aa[V1=="B_cells_naive"]
aa[grep("Neoantigen", V1)]
```
```{bash}
CUDA_VISIBLE_DEVICES=1 python train.py  --data_dir  ~/project/deeplearning/icb/data/tcga/nanostring/datasets_tsne_list.txt --model_dir ~/project/deeplearning/icb/data/tcga/nanostring/. --tensorboard_prefix survival
```

## Write ICB dataset
```{r}
## just write OS PFS response 
icb.sel = seq(3,22)
pheno.header = c("Response", "OS", "OS.Event", "PFS", "PFS.Event")
patient.all = icb.dataset.mats = list()
for (ii in icb.sel) {
	curr.dataset = icb.files[[ii]] %>% basename
	exp.curr = icb.files[[ii]] %>% fread
	exp.mat = exp.curr[,-1, with=F] %>% as.matrix %>% t %>%  
	set_colnames(genes.icb[[ii]])
	pheno.curr = icb.phenotypes[[ii]]
	col.present = intersect(pheno.header, colnames(pheno.curr))
	na.cols = setdiff(pheno.header, colnames(pheno.curr))
	exp.mat = exp.mat[pheno.curr$Patient,genes.overlap]
	na.matrix = matrix(NA, nrow=nrow(pheno.curr), ncol = length(na.cols)) %>% 
	set_colnames(na.cols)
	patient.all[[ii]] = paste(curr.dataset, pheno.curr$Patient )
	icb.dataset.mats[[ii]] = cbind(pheno.curr[,col.present,with=F], na.matrix) %>% 
	.[,pheno.header, with = F] %>% cbind(exp.mat,.) %>% 
	cbind(dataset=curr.dataset, .) 
}
icb.dataset.mats %<>%  do.call(rbind, .)

output.dir = "/homes6/asahu/liulab_home/projects/icb/data/tcga/nanostring/all.icb/"
write.dataset(output.dir =output.dir, dataset = icb.dataset.mats, sample.name = unlist(patient.all))
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/nanostring//params.json", output.dir)
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/nanostring//datasets_test_list.txt", output.dir)
file.copy("/homes6/asahu/liulab_home/projects/icb/data/tcga/nanostring//datasets_tsne_list.txt", output.dir)
aa = data.table(colnames(icb.dataset.mats) %>% cbind(seq(ncol(icb.dataset.mats))) ) %>% 
set_colnames(c("V1", "V2")
```

```{bash}
CUDA_VISIBLE_DEVICES=1 python train.py  --data_dir  ~/project/deeplearning/icb/data/tcga/nanostring/all.icb/datasets_tsne_list.txt --model_dir ~/project/deeplearning/icb/data/tcga/nanostring/all.icb/. --restore_file ~/project/deeplearning/icb/data/tcga/nanostring/tensorboardLog/20200708-225753/best.pth.tar
```
## Outputs are 
survivals
PD1/PDL1
sum(PRF1, GZMB, CD8A)
neoantigen
```{r}
output.genes = c("PDCD1" , "CD274", "PRF1", "GZMB", "CD8A")
aa[V1 %in% output.genes]
aa[V1 %in% output.genes]
```
## this does not work as anticapted

## Using scrna.v4.genes epoch-39 again.
```{bash}
CUDA_VISIBLE_DEVICES=1 python evaluate.py  --data_dir  ../data/tcga/scrna.v4.genes/datasets_test_list.txt --model_dir ../data/tcga/scrna.v4.genes/tensorboardLog/nopipeline_vae_20190819-161104/epoch-39/. --restore_file  ../data/tcga/scrna.v4.genes/tensorboardLog/nopipeline_vae_20190819-161104/epoch-39.pth.tar --output_dir ../data/tcga/scrna.v4.genes/tensorboardLog/nopipeline_vae_20190819-161104/epoch-39/.
```
### Above works 

## function to calculate pca 
```{r}


match.deepimmune.inputs  = function(exp, ref.pca.obj, cancertype=NULL, genes=NULL) {
	source("/homes6/asahu/project/deeplearning/icb/deepImmune/source.R")
	pca_sel_obj.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/pca_sel_obj.RData"
	pca_obj.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/pca_obj.RData"
	ref.expression.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/ref.expression.RData"
	ref.cancertype.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/ref.cancertype.RData"
	load(ref.expression.RData)
	load(ref.cancertype.RData)
	ref.expression.cancertype = ref.expression
	if(!is.null(cancertype)){
		ref.expression.cancertype = ref.expression[ref.cancertype==cancertype, ] 
	}
	load("/liulab/asahu/data/ssgsea/xiaoman/commmon.genes.RData")
	if(is.null(genes)) genes = common.genes
	all.genes = union(common.genes, genes)
	dataset_ssgsea_mat1 = impute.closest.gene(all.genes,exp)
	dataset_ssgsea_sel = dataset_ssgsea_mat1[ ,common.genes] 
	dataset_ssgsea_sel[is.na(dataset_ssgsea_sel)] = 0
	dataset_ssgsea_norm = normalize.expression(dataset_ssgsea_sel)

	dataset_ssgsea_matched = match.expression.distribution(dataset_ssgsea_sel, ref.expression.cancertype)

	load(pca_obj.RData)
	temp_out = get_pca(dataset_ssgsea_matched, pca_obj = pca_obj, scale=F) 
	pca_obj = temp_out$pca_obj
	pca_obj$len_selected = 50
	# save(file=paste0(output.dir, "/pca_obj.RData"), pca_obj)
	general.pcs = temp_out$pca_out[,seq(pca_obj$len_selected)]

	load("/liulab/asahu/data/ssgsea/xiaoman/ICB_datasets/top_correlated_genes_with_ICB_biomarkers.RData")

	load("/liulab/asahu/data/ssgsea/xiaoman/ICB_datasets/correlated_genes_with_ICB_biomarkers.RData")
	top.genes = unique(unlist(lapply(top.cors, names)))
	top.genes.extra = c("PMS2", "MSH6", "EPCAM", "MSH2")
	top.genes = unique(c(top.genes, top.genes.extra))
	load(pca_sel_obj.RData)

	temp_out = get_sel_pca(dataset_ssgsea_matched, top.genes, pca_sel_obj, scale=F)
	pca_sel_obj = temp_out$pca_obj
	pca_sel_obj$len_selected = 10
	pca_top = temp_out$pca_out[,seq(pca_sel_obj$len_selected)]
	colnames(pca_top) = paste0(colnames(pca_top), ".sel")

	list(dataset_ssgsea_matched = dataset_ssgsea_mat1[, genes], general.pcs=general.pcs, pca_top=pca_top)
}

read.icb2mat = function(file) {
	exp.curr = file %>% fread
	gene.names = exp.curr[,1, with=F] %>% unlist %>% c
	exp.mat = exp.curr[,-1, with=F] %>% as.matrix %>% t %>%  
	set_colnames(gene.names)
}

impute.and.match.expression = function(exp, genes, reference, ref.cancertype= NULL, cancertype = NULL){
	source("/homes6/asahu/project/deeplearning/icb/deepImmune/source.R")
	if(!is.null(cancertype))
	{
		if(cancertype %in% ref.cancertype)
			reference = reference[ref.cancertype==cancertype,]
	}
	exp2.dt =t(reference) %>% as.data.table  %>% 
	cbind(gene_name = colnames(reference), .) 
	dataset_ssgsea_mat1 = impute.closest.gene(genes, exp, exp2.dt)
	dataset_ssgsea_sel = dataset_ssgsea_mat1[ ,genes] 
	dataset_ssgsea_norm = normalize.expression(dataset_ssgsea_sel)
	reference.sub = reference[,genes]
	dataset_ssgsea_matched = match.expression.distribution(dataset_ssgsea_sel, reference.sub)
	dataset_ssgsea_matched
}
match.cols = function(dt1, cols){
	na.cols = setdiff(cols, colnames(dt1))
	matrix(NA, nrow = nrow(dt1), ncol = length(na.cols)) %>% 
	set_colnames(na.cols) %>% 
	cbind(dt1, .) %>% 
	.[, cols, with=FALSE] 
}
```

## Write genentech cohort
```{r}
tcga.dataset = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.genes/dataset.txt")
tcga.sample.name = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.genes/samples_name.txt")
tcga.headers = colnames(tcga.dataset)
pheno.start =which(tcga.headers == "B_cells_naive")
tcga.headers.pheno = tcga.headers[seq(pheno.start, length(tcga.headers))]
tcga.genes.inps = tcga.headers[seq(52, pheno.start-1)]
tcga.exp = tcga.dataset[,tcga.genes.inps,with=F] 
# temp.tcga.dataset = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v2/dataset.txt", select=1:60)
genentech.dataset = fread("/homes6/asahu/liulab_home/projects/icb/data/genentech.tpm/tf/dataset.txt")
genentech.sample = fread("/homes6/asahu/liulab_home/projects/icb/data/genentech.tpm/tf/samples_name.txt")
genentech.expression = read.icb2mat("/liulab/asahu/data/ssgsea/xiaoman/Genetech_expression_TPM.txt")
genentech.expression=genentech.expression[genentech.sample$x,]
genentech.deepimmune.ips = match.deepimmune.inputs(genentech.expression, cancertype="BLCA")
genentech.headers = colnames(genentech.dataset)
genentech.headers.pheno = genentech.headers[seq(which(genentech.headers == "B_cells_naive"), length(genentech.headers))]
genentech.followup = fread("/homes/asahu/liulab_home/data/links/data_private/ICB/follow_up/Mariathasan2018_PDL1_Bladder_RNASeq_mUC") 
setnames(genentech.followup, 2:3, c("OS.time", "OS.filtered"))
genentech.followup=genentech.followup[match(genentech.sample$x, Patient),]

phenotypes.present =  intersect(genentech.headers.pheno, tcga.headers.pheno) 
new.phenotypes = genentech.headers.pheno[51:54]
genentech.new.phenotype = cbind(Response = genentech.followup$Response, genentech.dataset[,new.phenotypes,with=F])

ref.expression.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/ref.expression.RData"
ref.cancertype.RData = "/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v1/ref.cancertype.RData"
load(ref.expression.RData)
load(ref.cancertype.RData)

new.reference.exp = cbind(
	ref.expression[,setdiff(colnames(ref.expression), tcga.genes.inps)], 
	tcga.exp[match(rownames(ref.expression), tcga.sample.name$x),]) %>% as.matrix
genentech.expression.matched = impute.and.match.expression(genentech.expression, genes = tcga.genes.inps, reference = new.reference.exp, ref.cancertype = ref.cancertype, cancertype="BLCA")

genentech.di.dataset = cbind(
	genentech.deepimmune.ips$general.pcs,
	genentech.deepimmune.ips$pca_top,
	genentech.expression.matched[,tcga.genes.inps]) %>% 
cbind(genentech.followup[,c("OS.time", "OS.filtered"), with=F]) 

genentech.di.dataset = match.cols(genentech.di.dataset, cols=tcga.headers)
genentech.di.dataset %<>% cbind(., response=genentech.followup$Response, genentech.new.phenotype)
genentech.di.dataset$cancertype = "BLCA_genentech"
genentech.di.dataset = genentech.di.dataset[,c(ncol(genentech.di.dataset), seq(1,ncol(genentech.di.dataset)-1)),with=F]
output.dir = "/homes6/asahu/liulab_home/projects/icb/data/all.icb/scrna.v4/genentech/"
dir.create(output.dir, recursive=T)
write.dataset(output.dir =output.dir, dataset = genentech.di.dataset , samples.names =genentech.sample$x)

aa = data.table(colnames(genentech.di.dataset) %>% cbind(seq(ncol(genentech.di.dataset)))) %>% set_colnames(c("V1", "V2"))
```


## All other ICB cohorts
```{r}
icb.files.sel = which(
	sapply(genes.icb, length) > 10000 & 
!grepl("Maria", icb.files))
icb.cancertypes = sapply(icb.files, function(tt) {
	aa = basename(tt) %>%
	strsplit(., split="_") %>% unlist 
	aa[3]
})
cancer.code = c("SKCM", "STAD", "BLCA", "KIRC", "NULL", "GBM" ) 
names(cancer.code) =c("Melanoma","Gastric","Bladder","Kidney","NSCLC-HNSC-Melanoma","Glioblastoma")
icb.cancercode = cancer.code[icb.cancertypes]

require(doMC)
require(foreach)
registerDoMC(cores = 32)
all.icb.di.datasets = foreach(ii = icb.files.sel, .inorder=T) %dopar%{
# all.icb.di.datasets =list()
# for (ii in icb.files.sel) {
# for (ii in seq(18:22)) {
	all.icb.di.dataset = NULL
	try(
		{
			all.icb.followup = icb.phenotypes[[ii]]
	all.icb.expression = read.icb2mat(icb.files[[ii]])
	all.icb.expression=all.icb.expression[all.icb.followup$Patient,]
	all.icb.deepimmune.ips = match.deepimmune.inputs(all.icb.expression, cancertype=icb.cancercode[[ii]])
	if("OS" %in% colnames(all.icb.followup))
		setnames(all.icb.followup, c("OS", "OS.Event"), c("OS.time", "OS.filtered"))
	if("PFS" %in% colnames(all.icb.followup))
		setnames(all.icb.followup, c("PFS", "PFS.Event"), c("PFI.time", "PFI.filtered"))
	all.icb.headers.pheno  = colnames(all.icb.followup)

	phenotypes.present =  intersect(all.icb.headers.pheno, tcga.headers.pheno) 

	all.icb.expression.matched = impute.and.match.expression(all.icb.expression, genes = tcga.genes.inps, reference = new.reference.exp, ref.cancertype = ref.cancertype, cancertype=icb.cancercode[[ii]])

	cancertype.curr = paste(icb.cancercode[[ii]], ii)
	all.icb.di.dataset = cbind(
		all.icb.deepimmune.ips$general.pcs,
		all.icb.deepimmune.ips$pca_top,
		all.icb.expression.matched[,tcga.genes.inps]) %>% 
	cbind(all.icb.followup[,phenotypes.present, with=F]) 

	all.icb.di.dataset = match.cols(all.icb.di.dataset, cols=tcga.headers)
	response.curr = ifelse(all.icb.followup$Response > 0, 1,0)
	all.icb.di.dataset %<>% cbind(., response=response.curr)
})
	# all.icb.di.datasets[[ii]] = all.icb.di.dataset
	all.icb.di.dataset
}
for (ii in seq(18)) {
	try(
		{
			all.icb.di.datasets[[ii]]$cancertype = NULL 
			all.icb.di.datasets[[ii]] %<>% .[,cancertype:=ii]
		})
}

all.icb.di.datasets.comb = do.call(rbind, all.icb.di.datasets)
all.icb.di.datasets.comb = all.icb.di.datasets.comb[,c(ncol(all.icb.di.datasets.comb), seq(1,ncol(all.icb.di.datasets.comb)-1)),with=F]
output.dir = "/homes6/asahu/liulab_home/projects/icb/data/all.icb/scrna.v4/all.icb/"
dir.create(output.dir, recursive=T)
write.dataset(output.dir =output.dir, dataset =all.icb.di.datasets.comb, samples.names =all.icb.di.datasets.comb$cancertype)

aa = data.table(colnames(all.icb.di.datasets.comb ) %>% cbind(seq(ncol(all.icb.di.datasets.comb )))) %>% set_colnames(c("V1", "V2"))
```


```{bash}

CUDA_VISIBLE_DEVICES=3 python train.py  --data_dir  ../data/tcga/scrna.v4.genes/datasets_tsne_list.txt --model_dir ../data/tcga/scrna.v4.genes/. --tensorboard_prefix mtl_

mkdir /homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25

CUDA_VISIBLE_DEVICES=3 python train.py  --data_dir  ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/datasets_tsne_list.txt --model_dir ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/.   --restore_file ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25.pth.tar

CUDA_VISIBLE_DEVICES=3 python train.py  --data_dir  ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/datasets_tsne_list.txt --model_dir ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/. --tensorboard_prefix genen


CUDA_VISIBLE_DEVICES=3 python train.py  --data_dir  ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/datasets_tsne_list.txt --model_dir ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25/. --tensorboard_prefix allICB
```

```{r}
tcga.dataset = fread("/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.genes/dataset.txt")
tcga.dataset.curr = tcga.dataset[cancertype=="SKCM"]
output.dir = "/homes6/asahu/liulab_home/projects/icb/data/tcga/scrna.v4.genes/skcm"
sample.name.curr = tcga.sample.name [tcga.dataset$cancertype=="SKCM"]
write.dataset(output.dir =output.dir , dataset = tcga.dataset.curr, samples.names = sample.name.curr)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300//params.json", output.dir)
file.copy("/homes6/asahu/project/deeplearning/icb/data/tcga/tf/binarize/datasets_test_list.txt", output.dir)
```

```{bash}
CUDA_VISIBLE_DEVICES=3 python evaluate.py  --data_dir  ../data/tcga/scrna.v4.genes/skcm/datasets_test_list.txt --model_dir ../data/tcga/scrna.v4.genes/skcm/. --restore_file ../data/tcga/scrna.v4.genes/tensorboardLog/20200713-005300/epoch-25.pth.tar
```
