---
title: "LEE process data Notebook"
output:
  pdf_document: default
  html_notebook: default
---

```{r}
install.github <- function(lib) {
    require(devtools)
    require(withr)
    with_libpaths(new = "/homes6/asahu/R/x86_64-pc-linux-gnu-library/3.6", install_github(lib))
}

```


```{r}
library(data.table)
library(magrittr)
library(ggplot2)
# library(SingleCellExperiment)
# library(Matrix)
library(Seurat)
library(cowplot)
# library(harmony)
library(uwot)
library(parallel)
library(EnhancedVolcano)
library(tidyverse)
setwd("~/project/deeplearning/icb/deepImmune/data_processing/ssgsea/imputation/simple")
lee.sco = readRDS(file="~/liulab_home/data/single_cell/Lee_data/lee.seurat.cd3.RDS")
# var.genes = VariableFeatures(lee.sco)
# saveRDS(file = "~/liulab_home/data/single_cell/Lee_data/lee.seurat.cd3.vargenes.RDS", var.genes)
load(".figs/lee/lee.cd3.pre.deg.RData")
```

```{r}
lee.sco.pre =lee.sco[, lee.sco$Treatment.Cycle=="C1D1"]
p = FeaturePlot(lee.sco.pre, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", 
                                          "CD8A", "GZMB", "PRF1", "IL7R", "CCR7", "CD4")) 
ggsave(file=".figs/lee/lee.all.features.pdf", p, width=15, height=10)

p = FeaturePlot(lee.sco.pre[,lee.sco.pre$response=="CR"], features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", 
                                                                       "CD8A", "GZMB", "PRF1", "IL7R", "CCR7", "CD4")) 
ggsave(file=".figs/lee/lee.completeresponders.features.pdf", p, width=15, height=10)
p = FeaturePlot(lee.sco.pre[,lee.sco.pre$response=="PD"], reduction = "umap", features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP", 
                                                                                           "CD8A", "GZMB", "PRF1", "IL7R", "CCR7", "CD4")) 
ggsave(file=".figs/lee/lee.partialdisease.features.pdf", p, width=15, height=10)

p = DimPlot(lee.sco.pre, reduction = "umap", group.by = "seurat_clusters", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/lee/lee.clusters.pdf", p, width=15, height=10)
p


lee.sco$response = factor(lee.sco$response, levels=c("CR", "PR", "PD"))

p1 = DimPlot(lee.sco[,lee.sco$response %in% c("CR", "PD", "PR")], reduction = "umap", group.by = "patient.name", split.by = "response", pt.size = .1) 
ggsave(file=".figs/lee/lee.clusters.patient.pdf", p1, width=15, height=10)


p1 = DimPlot(lee.sco[,(lee.sco$response %in% c("CR", "PD", "PR")) & (lee.sco$Treatment.Cycle=="C1D1")], reduction = "umap", group.by = "patient.name", split.by = "response", pt.size = .1) 
ggsave(file=".figs/lee/lee.clusters.patient.pre.pdf", p1, width=15, height=10)

p1 = DimPlot(lee.sco[,(lee.sco$response %in% c("CR", "PD", "PR")) & (lee.sco$Treatment.Cycle=="C4D1")], reduction = "umap", group.by = "patient.name", split.by = "response", pt.size = .1) 
ggsave(file=".figs/lee/lee.clusters.patient.post.pdf", p1, width=15, height=10)
```
## Pseudobulk for cluster 0
```{r}

newpatient.name 
dt1= lee.sco@meta.data %>% 
    as.data.table() %>%
    .[,new.patientname:= as.factor(dt1$response)]

levels(dt1$new.patientname) = c("CR", "xxna", "xPD", "PR")
dt1[,new.patient:=paste(new.patientname,patient.name)] 
lee.sco <- AddMetaData(
    object = lee.sco,
    metadata = dt1$new.patient,
    col.name = 'new.patient'
)

lee.sco.pre.clust = lee.sco[,(lee.sco$Treatment.Cycle=="C1D1") & (lee.sco$seurat_clusters==0) & (lee.sco$response %in% c("CR", "PR", "PD"))]
exp.curr1 = lee.sco.pre.clust@assays$RNA@counts
meta.dt1 = lee.sco.pre.clust@meta.data %>%
    as.data.table() %>%
    .[,.(binaryResponse=ifelse(response %in% c("CR", "PR"),1 ,0) , patient=patient.name)] 
meta.curr = list()
exp.curr2 = list()
for(patient in unique(meta.dt1$patient)){
    inx = which(meta.dt1$patient==patient)
    exp.curr2[[patient]] = rowSums(exp.curr1[,inx],na.rm=T)
    meta.curr[[patient]] = meta.dt1[inx[1],]
}

meta.dt = do.call(rbind, meta.curr)
exp.curr = t(do.call(rbind, exp.curr2))
meta.dt$total_count = colSums(exp.curr)
f1 <- 'expr ~ log_total_count'
f2 <- 'expr ~ log_total_count + binaryResponse'
f3 <- 'expr ~ log_total_count + 1|binaryResponse'
f4 <- 'expr ~ log_total_count + 1|patient'
f5 <- 'expr ~ log_total_count + 1|binaryResponse/patient'
library(lmtest)
# meta.dt[,log_total_count:=avinash::znorm(log(total_count))]
meta.dt[,log_total_count:=log(total_count)]

eval.glm.poisson = function(tmp_df, val){
    tryCatch({
        if(sum(val>0) <=1 ) stop()
        tmp_df = tmp_df %>%
            mutate(expr=val)
        mres1 <- glm(f1, data = tmp_df, family = poisson())
        mres2 <- glm(f2, data = tmp_df, family = poisson())
        lrt_res <- lrtest(mres1, mres2)
        # browser()
        pval <- lrt_res$`Pr(>Chisq)`[2]
        effect_size <- unname(coef(mres2)[2])
        c(pval = pval, effect_size = effect_size)
    }, error = function(e) c(pval = NA, effect_size = NA))
}

eval.glmer.poisson = function(tmp_df, val){
    tryCatch({
        if(sum(val>0) <=1 ) stop()
        tmp_df = tmp_df %>%
            mutate(expr=val)
        mres1 <- glm(f1, data = tmp_df, family = poisson())
        mres2 <- glm(f2, data = tmp_df, family = poisson())
        mres5 <- glmer(f5, data = tmp_df, family = poisson())
        
        lrt_res <- lrtest(mres1, mres2)
        browser()
        pval <- lrt_res$`Pr(>Chisq)`[2]
        effect_size <- unname(coef(mres2)[2])
        
        c(pval = pval, effect_size = effect_size)
    }, error = function(e) c(pval = NA, effect_size = NA))
}
out = mclapply(seq(nrow(exp.curr)), function(tt) 
    eval.glm.poisson(meta.dt, exp.curr[tt,]),
    mc.cores=16
)
bulk.pre.diff.dt = do.call(rbind, out)%>% 
    as.data.table() %>%
    .[,gene:=rownames(exp.curr)] %>%
    .[!is.na(pval)]%>%
    .[order(pval)]


p=VlnPlot(lee.sco.pre.clust, bulk.pre.diff.dt[1:10]$gene,group.by = 'response') 

lee.sco.pre.clust.norm = lee.sco[,((lee.sco$Treatment.Cycle=="C1D1") & (lee.sco$seurat_clusters==0)) & (lee.sco$response %in% c("CR", "PR", "PD")) ] 
p=VlnPlot(lee.sco.pre.clust.norm, bulk.pre.diff.dt[1:8]$gene, group.by = 'new.patient') 
p=VlnPlot(lee.sco.pre.clust.norm, c("JUNB", "CD69"), group.by = 'new.patient') 
dim(meta.dt)
responders = meta.dt[binaryResponse==1]$patient
nonresponders = meta.dt[binaryResponse==0]$patient
# library(aviansh.scRNA)
deseq.out = DESeq2DETest(data.use=exp.curr[,c(responders,nonresponders)], cells.1=responders, cells.2=nonresponders)
deseq.dt = deseq.out %>%
    as.data.frame() %>%
    mutate(gene=rownames(.)) %>%
    data.table() %>% 
    .[order(pvalue)]

```

```{r}
mixed.dt = pre.diff.dt[["0"]] %>%
    .[order(pval)]

VlnPlot(lee.sco.pre.clust, deseq.dt[1:4]$gene,group.by = 'new.patient') 
VlnPlot(lee.sco.pre.clust, mixed.dt[1:4]$gene,group.by = 'new.patient') 
```

```{r}

do.psudeobulkDEG <- function(clust, pretreatment=TRUE) {
    treatment.type = ifelse(pretreatment,"C1D1", "C4D1" )
    lee.sco.pre.clust = lee.sco[,(lee.sco$Treatment.Cycle==treatment.type) & (lee.sco$seurat_clusters %in% clust) & (lee.sco$response %in% c("CR", "PR", "PD"))]
    
    exp.curr1 = lee.sco.pre.clust@assays$RNA@counts
    meta.dt1 = lee.sco.pre.clust@meta.data %>%
        as.data.table() %>%
        .[,.(binaryResponse=ifelse(response %in% c("CR", "PR"),1 ,0) , patient=patient.name)] 
    meta.curr = list()
    exp.curr2 = list()
    for(patient in unique(meta.dt1$patient)){
        inx = which(meta.dt1$patient==patient)
        exp.curr2[[patient]] = rowSums(exp.curr1[,inx],na.rm=T)
        meta.curr[[patient]] = meta.dt1[inx[1],]
    }
    
    meta.dt = do.call(rbind, meta.curr)
    exp.curr = t(do.call(rbind, exp.curr2))
    
    responders = meta.dt[binaryResponse==1]$patient
    nonresponders = meta.dt[binaryResponse==0]$patient
    # library(aviansh.scRNA)
    deseq.out = DESeq2DETest(data.use=exp.curr[,c(responders,nonresponders)], cells.1=responders, cells.2=nonresponders)
    deseq.dt = deseq.out %>%
        as.data.frame() %>%
        mutate(gene=rownames(.)) %>%
        data.table() %>% 
        .[order(pvalue)]
    deseq.dt
}
desseq.dt.B = do.psudeobulkDEG(clust=c(6,8,15:18))
desseq.dt.B.post = do.psudeobulkDEG(clust=c(6,8,15:18), pretreatment = F)
clust = c(6,8,15:18)
lee.sco[,(lee.sco$Treatment.Cycle=="C1D1") &(lee.sco$seurat_clusters %in% clust) & (lee.sco$response %in% c("CR", "PR", "PD"))] %>%
    VlnPlot(., desseq.dt.B[1:6]$gene,group.by = 'new.patient') 
lee.sco@meta.data$binaryResponse = ifelse(lee.sco$response %in% c("CR", "PR"), 1 , ifelse(lee.sco$response == "PD", 0, NA))
lee.sco.temp = lee.sco[,(lee.sco$Treatment.Cycle=="C1D1") &(lee.sco$seurat_clusters %in% clust) & (lee.sco$response %in% c("CR", "PR", "PD"))] 
lee.sco.temp.markers <- FindMarkers(lee.sco.temp,  ident.1 = "1", group.by = 'binaryResponse', only.pos = FALSE, min.pct = 0.1, logfc.threshold = 0.1)

 aa = rownames(lee.sco.temp.markers)
gene.sel = c("CXCR5", "CXCL13", "TCF7")
aa.sel = aa[!(grepl("RP",aa) |grepl("MT",aa))] %>% head(.,20) %>%
    c(., gene.sel)
p = VlnPlot(lee.sco.temp, aa.sel,group.by = 'new.patient') 
ggsave(file=".figs/lee/lee.clusters.temp.pdf", p, width=25, height=20)
desseq.dt.B[gene=="CXCR5",]
desseq.dt.B.post[gene=="CXCR5",]
desseq.dt.B.post[gene=="TCF7",]
desseq.dt.B.all =do.psudeobulkDEG(clust=seq(0,25))

```
## Differential expresssion analysis
```{r}

# universe = clusterProfiler::bitr(desseq.dt.B.all$gene, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")


gene.eg.dt = clusterProfiler::bitr(unique(desseq.dt.B.all$gene), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
desseq.dt.B.all$ENTREZID = gene.eg.dt[match(desseq.dt.B.all$gene,universe$SYMBOL),]$ENTREZID

kk <- clusterProfiler::enrichKEGG(gene         = desseq.dt.B.all[padj<0.05 & stat>0 ]$ENTREZID,
                                  universe = gene.eg.dt$ENTREZID,
                                  organism     = 'hsa',
                                  pvalueCutoff = 0.05)
clusterProfiler::dotplot(kk, showCategory=30) 


# using LM22 signatures 
lm22  = fread("~/liulab_home/data/single_cell/LM22.txt")
lm22.exp = lm22%>%
    .[,2:23, with=F] %>%
    as.matrix() %>%
    set_rownames(lm22$`Gene symbol`) %>%
    apply(., c(1,2),as.numeric)
desseq.dt.B.all.match = desseq.dt.B.all[match(rownames(lm22.exp),gene)]
lm22.enrich = apply(lm22.exp,2, function(tt) {
    aa = cor.test(tt, desseq.dt.B.all.match$log2FoldChange, method = "spearman")[c("estimate","p.value")]
    unlist(aa)
}) %>% t() %>%
     data.table() %>%
    .[,type:=colnames(lm22.exp)] %>%
    .[,padj:=p.adjust(p.value,method="fdr")]


EnhancedVolcano(lm22.enrich,
	lab = lm22.enrich$type,
	x = 'estimate.rho',
	y = 'padj',
	pCutoff = 5e-2,
	FCcutoff = .1,
	pointSize = c(ifelse(lm22.enrich$padj< 5E-2, 4, 3)),
	labSize = 4.0,
	legend=c('NS','Log (base 2) fold-change','Adj.P value',
		'Adj.P value & Log (base 2) fold-change'),
	legendPosition = 'right',
	legendLabSize = 8,
	legendIconSize = 4.0,
	drawConnectors = TRUE,
	widthConnectors = 0.2,
	colAlpha = 0.8,
	colConnectors = 'grey30'
	)


```
## implmentation of myhippo

```{r}
library(profvis)
lee.small.mat = assay(lee.small)

profvis({


myRowVar <- function(x) {
 (proxyC::rowSds(x))^2
 

}
mypreprocess_heterogeneous = function(X) {

    gene_var = myRowVar(X) 
    gene_var = ifelse(gene_var > 0, gene_var, NA)
    df = data.frame(gene = rownames(X), gene_mean = rowMeans(X),
                    zero_proportion = proxyC::rowZeros(X)/ncol(X),
                    gene_var = gene_var)
    df$samplesize = ncol(X)
    df = compute_test_statistic(df)
    return(df)
}

	

myone_level_clustering = function(subX, z_threshold, subsample= NULL) {
  require(bigpca)
  subdf = mypreprocess_heterogeneous(subX)
  # browser()
  # subdf = compute_test_statistic(subdf)
  features = subdf[subdf$zvalue > z_threshold, ]
  nullfeatures = data.frame(matrix(ncol = 11, nrow = 0))
  colnames(nullfeatures) = c("gene", "gene_mean", "zero_proportion",
                             "gene_var", "samplesize", "expected_pi", "se",
                             "minus_logp","zvalue", "subsetK", "K")
  if (nrow(features) < 10) {
    return(list(features = nullfeatures, pcs = NA, km = NA))
  }
  if (nrow(features) < 10) {
    return(list(features = nullfeatures, pcs = NA, km = NA,
                unscaled_pcs = NA,subdf = NA))
  }
     # browser()
     features.all = features
     if(nrow(features) >100 )
      features = features[order(features$zvalue,decreasing=T)[1:100],] 

    # inx = which(rownames(subX) %in% features$gene)
    # mat = log(subX[inx, ] + 1)
    submat = mat = log(subX[features$gene,] + 1) %>% 
        scale(., center=T, scale=F)
    
    if(!is.null(subsample)){
        size.curr = max(250, nrow(mat) * subsample)
        if(size.curr< nrow(mat))
            submat = mat[sample(nrow(mat), size=size.curr),]  
    }
  # pcs = NA
  # try(expr = {
  eigenvec = big.PCA(submat, pcs.to.keep =min(9, nrow(features) -1, ncol(submat) - 1), center = FALSE,  return.loadings = T)$loadings
  # browser()
  pcs = mat %*% eigenvec
  # }, silent = TRUE)
  if (is.na(pcs[1])) {
      browser()
    return(list(features = nullfeatures, pcs = NA, km = NA,
                unscaled_pcs = NA,
                subdf = NA))
  } else {
    km = kmeans(pcs, 2, nstart = 10, iter.max = 50)
  }
  return(list(features = features.all,
              pcs = pcs,
              km = km,
              subdf = subdf))
}


myhippo = function(X, K = 20,
                 z_threshold = 2,
                 outlier_proportion = 0.001,
                 verbose = TRUE, subsample_pca = NULL) {

  if (outlier_proportion > 1 | outlier_proportion < 0) {
    stop("Outlier_proportion must be a number between 0 and 1.
         Default is 5%")
  }

  param = list(z_threshold = z_threshold,
               outlier_proportion = outlier_proportion,
               maxK = K)
  outlier_number = nrow(X) * outlier_proportion
  labelmatrix = matrix(NA, ncol(X), K)
  labelmatrix[, 1] = 1
  eachlevel = list()
  subX = X
  subXind = seq(ncol(X))
  withinss = rep(0, K)
  oldk = 1
  features = list()
  featuredata = list()
  for (k in 2:K) {
    thisk = myone_level_clustering(subX, z_threshold, subsample = subsample_pca)
    if (is.na(thisk$features$gene[1])) {
      if(verbose){
        message("not enough important features left; terminate the procedure")
      }
      labelmatrix = labelmatrix[, seq((k - 1))]
      break
    }
    if (nrow(thisk$features) < outlier_number) {
      if(verbose){
        message("not enough important features; terminate the procedure")
      }
      labelmatrix = labelmatrix[, seq((k - 1))]
      break
    }

    if (verbose) {message(paste0("K = ", k, ".."))}
    labelmatrix[, k] = labelmatrix[, k - 1]
    labelmatrix[subXind[thisk$km$cluster == 2], k] = k
    oneind = thisk$km$cluster == 1
    twoind = thisk$km$cluster == 2

    withinss[c(oldk,k)] = thisk$km$withins/thisk$km$size

    ind = which(table(thisk$km$cluster) <= 5)
    if (length(ind) >= 1){
      valid_indices = seq(k-1)[-ind]
      oldk = which(withinss == max(withinss[valid_indices]))
    }else{
      oldk = which.max(withinss[seq(k-1)])
    }
    if (sum(labelmatrix[, k] == oldk) < 2) {
      if(verbose){
        message("too few cells in one cluster; terminating the procedure")
      }
      labelmatrix = labelmatrix[, seq(k)]
      break
    }
    # subX = X[thisk$features$gene, which(labelmatrix[, k] == oldk)]
    subX = X[, which(labelmatrix[, k] == oldk)]
    subXind = which(labelmatrix[, k] == oldk)
    thisk$features$subsetK = oldk
    thisk$features$K = k
    features[[k - 1]] = thisk$features
  }
  hippo = list(X = X,
                                features = features,labelmatrix = labelmatrix,
                                z_threshold = z_threshold, param = param,
                                outlier_proportion = outlier_proportion)
  return(hippo)
}



mylee.small.subsample1 = myhippo(lee.small, 
            K = 3, 
            z_threshold = 2, 
            outlier_proportion = 0.005,
            verbose=TRUE, subsample_pca = .1)
})


```

## Create genesets from TCGA 
```{r}
# immune factors 
# ERVs
# Mutations
immune.genes = avinash::get.checkpoint.genes()
cytokines = fread("/liulab/xmwang/oxphos_proj/loading_data/surface/cytokine.txt", header=F)$V2
surface.genes = fread("/liulab/xmwang/oxphos_proj/loading_data/surface/ExpressionLigRec.txt", header=T)$ApprovedSymbol
tcga.dataset = fread("~/project/deeplearning/icb/data/tcga/scrna.v4.allgenes.nopcs/dataset.txt")
if.start.inx = which(colnames(tcga.dataset)== "B_cells_naive")
if.factors = tcga.dataset[,seq(if.start.inx, ncol(tcga.dataset)),with=F] %>% 
    apply(.,2, rank, na.last = "keep")
exp.tcga = tcga.dataset[,seq(2,if.start.inx-1), with=F]%>%
    apply(.,2, rank, na.last = "keep")
if.exp.cor = WGCNA::cor(if.factors, exp.tcga, use="pairwise.complete.obs")

## Add other biomarkers.
#INFG11I 
#INFG
```


## Tree differential expression analysis
```{r}
#' Title
#'
#' @param sco 
#'
#' @return
#' @export
#'
#' @examples
mydeg <- function(sco) {
    exp.curr1 = sco@assays$RNA@counts
    meta.dt1 = sco@meta.data %>%
        as.data.table() %>%
        .[,.(binaryResponse=ifelse(response %in% c("CR", "PR"),1 ,0) , patient=patient.name)] 
    
    meta.curr = list()
    exp.curr2 = list()
    for(patient in unique(meta.dt1$patient)){
        inx = which(meta.dt1$patient==patient)
        exp.curr2[[patient]] = rowSums(exp.curr1[,inx],na.rm=T)
        meta.curr[[patient]] = meta.dt1[inx[1],]
    }
    meta.dt = do.call(rbind, meta.curr)
    exp.curr = t(do.call(rbind, exp.curr2))
    responders = meta.dt[binaryResponse==1]$patient
    nonresponders = meta.dt[binaryResponse==0]$patient
    deseq.out = DESeq2DETest(data.use=exp.curr[,c(responders,nonresponders)], cells.1=responders, cells.2=nonresponders)
    deseq.dt = deseq.out %>%
        as.data.frame() %>%
        mutate(gene=rownames(.)) %>%
        data.table() %>% 
        .[order(pvalue)]
    deseq.dt
}


load("~/liulab_home/data/single_cell/Lee_data/cd3_plus.hippo.clustering.RData") ## hippo.out
# head(hippo.out$labelmatrix)
for (xx in seq(2,ncol(hippo.out$labelmatrix))) {
    lee.sco$clust.temp = as.factor(hippo.out$labelmatrix[,xx])
    p= DimPlot(lee.sco, reduction = "umapunorm",  group.by = "clust.temp", pt.size = .1, label = TRUE) +  NoLegend()
        ggsave(file=sprintf(".figs/lee/lee.hippo.clusters1.%s.pdf", xx), p, width=15, height=10)
}



```

# create tree stucture
```{r}
create.tree <- function(object, reduction='pca', dims=NULL, oldk = 1, newk = 2) {
    
     embeddings <- Embeddings(object = object, reduction = reduction)[, dims]
     km = kmeans(embeddings, centers = 2, iter.max = 10)$cluster
     seurat.clust = object$seurat_clusters
     out = sapply(levels(seurat.clust), function(tt){
         aa = table(km[seurat.clust==tt]) %>% which.max() %>%
         names() 
         ifelse(aa=="1", oldk, newk)
     })
     if(length(unique(out))==1) out[] = oldk
     out
}
create.cluster.tree<- function(sco) {
    seurat.clusters = levels(sco$seurat_clusters)
    maxk = length(unique(sco$seurat_clusters))
    clust.mat = matrix(NA, ncol=maxk, nrow=maxk)
    rownames(clust.mat) = paste0("clust", seurat.clusters) 
    clust.mat[,1] = 1
    for (ii in seq(1,maxk-1)) {
        clust.mat.old = clust.mat[, ii] 
        temp =  table(clust.mat.old) 
        oldk = temp[temp > 1] %>% 
            names() %>% as.numeric() %>% min()
        newk=max(clust.mat[,ii]) +1
        if(oldk >= maxk) break()
        seurat.clust.old = seurat.clusters[clust.mat.old == oldk]
        inx = which(seurat.clusters == oldk)
        out = create.tree(sco[,sco$seurat_clusters %in% seurat.clust.old], reduction="harmony", oldk, newk, dims = 1:20)
        out = out[seurat.clust.old]
        clust.mat[, ii+1] = clust.mat[,ii]
        clust.mat[paste0("clust", names(out)),ii+1] = unlist(out)
    }
    clust.mat%>% t() %>%
        unique() %>% t()
}

clust.mat.all = create.cluster.tree(lee.sco)
library(data.tree)
# acme <- Node$new("Acme Inc.")
  # accounting <- acme$AddChild("Accounting")
clust.mat.tree = list()
clust.mat.tree[["1 1"]] = Node$new("root")
# parent = 1
for(ii in seq(2,ncol(clust.mat.all))){
    for(parent in unique(clust.mat.all[,ii-1])){
        parent.label = paste(ii-1, parent)
        parent.inx = which(clust.mat.all[,ii-1]==parent)
        childs = clust.mat.all[parent.inx,ii]
        for(child in unique(childs))
            clust.mat.tree[[paste(ii, child)]] = clust.mat.tree[[parent.label]]$AddChild(paste(ii, child))
    }
    
}

```


## calculate tree differential expression 
```{r}

cc = 1
lee.degs = list()
clust.inx = gsub("clust", rownames(clust.mat.all), replacement = "")
clust.mat.padded = cbind(clust.mat.all, as.factor(clust.inx))
# for (ii in seq(ncol(clust))) {
for (ii in c(1:6,17)) {
    clust.curr = clust.mat.padded[,ii]
    # for(jj in unique(clust.curr)) {
    for(jj in 8:12) {
        seurat.clust.curr = clust.inx[which(clust.curr==jj)]
        lee.degs[[cc]] = 
            tryCatch({
            lee.sco[,(lee.sco$seurat_clusters %in%seurat.clust.curr) & (lee.sco$response %in% c("CR", "PD"))] %>%
            mydeg()}, error = function(e) NA)
        cc  = cc +1
    }
    print(cc)
}
# lee.degs.back = lee.degs

lee.deg.mat = lapply(lee.degs, function(tt){ 
    if(is.na(tt)) return(NULL)
    tt[match(lee.degs[[1]]$gene, gene)]$stat
    }) %>%
    do.call(cbind, .) %>%
    set_rownames(lee.degs[[1]]$gene)

## only include lymphocyte populations. 
aa =lee.sco@reductions$umap@cell.embeddings 
lymp.deg = lee.sco[,(aa[,1] > -10 &  aa[,2] < 10) & (lee.sco$response %in% c("CR", "PD"))] %>% 
    mydeg()


sel.inx = which((aa[,1] > -10 &  aa[,2] < 10) & (lee.sco$response %in% c("CR", "PD")) & (lee.sco$Treatment.Cycle=="C1D1"))
lee.curr.pre  = lee.sco[,sel.inx]
lymp.deg.pre = lee.curr.pre %>% 
    mydeg()
```
##Bulk deg mat
```{r}
bulk.deg.dt = readRDS("/homes6/asahu/liulab_home/data/immunotherapy-trials/all.bulk.rnaseq.deg.Rds")
bulk.deg.temp = dcast(gene~type, data=bulk.deg.dt, value.var = "deg.effect", fill = NA) 
bulk.deg.mat = bulk.deg.temp[,-1,with=F] %>%
    as.matrix() %>%
    set_rownames(bulk.deg.temp$gene)
```


## Pathways 
<!-- 1. LM22
2. IF
3. Others 
4.INGs -->
```{r}
myFastCor.multicores = function(x,y, method="spearman", num=20000, nthreads=1){
    if(is.null(num))  num=0
    if(nthreads>1){
        
        out = mclapply(seq(ncol(y)), function(tt) myFastCor(x=x,y=y[,tt], method=method, num=num), mc.cores = nthreads) %>%
            do.call(cbind,.)
    }else{
        out = myFastCor(x=x,y=y, method=method, num=num)
    }
    out
}

myFastCor <- function(x,y, method="spearman", num=20000) {
    if (ncol(x) < num | ncol(y) < num){
        out = cor(x, y, method="spearman", use="pairwise.complete.obs")
    }else{
        if(method=="spearman"){
            x%<>%
                apply(.,2, rank, na.last="keep")
            y%<>%
                apply(.,2, rank, na.last="keep")
        }
        out = WGCNA::cor(x, y, use="pairwise.complete.obs")
    }
    out
}
calcCorEnrich <- function(x,y) {
    common = intersect(rownames(x), rownames(y))
    myFastCor(x[common,], y[common,])
}

calcLeeBulkEnrich <- function(signature, figFile, reduce.rownames = FALSE) {
    require(pheatmap)
    col1 = colorRampPalette(c("darkgreen", "white", "white", "white", "white", "darkred"))(300)
    col = col1[c(1:100, seq(101,200,2), 201:300)]
    lee.signature.enrich = calcCorEnrich(signature , lee.deg.mat)
    bulk.signature.enrich = calcCorEnrich(signature , bulk.deg.mat)
    combined.enrich =bulk.signature.enrich %>%
        set_colnames(substring(colnames(bulk.signature.enrich), 1, 25)) %>%
        cbind(lee.signature.enrich, .)
    enrich.mat = combined.enrich
    combined.enrich[abs(combined.enrich) <.22]=0
    combined.enrich[combined.enrich >.5]=.5
    combined.enrich[combined.enrich <-.5]=-.5
    combined.enrich = combined.enrich[which(rowSums(combined.enrich!=0,na.rm = T) > 1), colSums(combined.enrich!=0, na.rm=T) > 1]
    combined.enrich[is.na(combined.enrich)] = 0 
    if(!reduce.rownames) {
        p = pheatmap(t(combined.enrich), cluster_rows=F, cluster_cols=T, fontsize_col = 8, fontsize_row =8, color = col, width = 13, height = 7, filename = figFile)
    }else{
        rownames(combined.enrich)  = substring(rownames(combined.enrich), 1, 15)
        p = pheatmap(t(combined.enrich), cluster_rows=F, cluster_cols=T, fontsize_col = 8, fontsize_row =3, color = col, width = 16, height = 7, filename = figFile)
    }
    list(p=p, enrich.mat = enrich.mat, lee.signature.enrich = lee.signature.enrich, bulk.signature.enrich=bulk.signature.enrich)
    
}

lee.if.enrich = calcLeeBulkEnrich(t(if.exp.cor), figFile = ".figs/lee/lee.if.enrich.pdf")
```
## LM22 
```{r}
lm22.exp.norm = scale(lm22.exp, center = T, scale = T)
lee.lm22.enrich = calcLeeBulkEnrich(lm22.exp.norm, figFile = ".figs/lee/lee.lm22.enrich.pdf")
load('/liulab/asahu/data/immunotherapy-trials/icb/rdata/mariathasan.exp.RData')
mariathasan.exp.norm = mariathasan.exp %>%
    scale(.,center=T, scale = T)
lm22.marithasan = calcCorEnrich(lm22.exp.norm, t(mariathasan.exp.norm)) 
responders = icb.response.dt[dataset=="Mariathasan2018_PDL1_Bladder_RNASeq_mUC"][response==1]$Patient
nonresponders = icb.response.dt[dataset=="Mariathasan2018_PDL1_Bladder_RNASeq_mUC"][response==0]$Patient
summary(lm22.marithasan["T cells CD8",responders])
summary(lm22.marithasan["T cells CD8",nonresponders])
wilcox.test(lm22.marithasan["NK cells activated",responders], lm22.marithasan["NK cells activated",nonresponders], alternative = "greater")


lm22.marithasan["NK cells activated",nonresponders], alternative = "greater")
dt1 = bulk.deg.dt[type=="Mariathasan2018_PDL1_Bladder_RNASeq_mUC"]


```

## Loading all gene signatures
<!-- 1. Msigdb 
2. ScMatch 
3. Davis
4. Chenfei
5. Cluster markers -->
```{r}
msigdb = msigdbr::msigdbr(species = "Homo sapiens")

davis.signatures <- list(
    CD8T = c("CD8A","CD8B","CD3E","CD3D","CD3G"),
    CD4conv = c("CD4","CD3E","CD3D","CD3G"),
    Treg=c("CD3E","CD4","FOXP3","CTLA4"),
    Neutrophils = c("S100A9","MPO","CXCR2"),
    Monocytes = c("CD14","CD16","CCR2"),
    #Macrophages = c("ADGRE1","CD68"),
    MacrophagesM2 = c("ADGRE1","CD68",'ARG1'),
    MacrophagesM1 = c("ADGRE1","CD68"),
    Dendritic= c("CD1C","CD209","CCL22","CD86"),
    pDendritic = c("SIGLECH","IRF7"),
    NK = c("KLRK1", "KLRC1", "KLRF1"),
    Mast = c("CPA3", "ENPP3"),
    B=c("CD19","CD79A","CD79B"),
    Plasma=c("SLAMF7", "IGKC"),
    HSC=c("CD34","KIT","SOX4"),
    Epithelial=c("EPCAM","CDH1"),
    Endo=c("VWF","PECAM1"),
    Fibro=c("COL1A2","MMP2","MYL9"),
    CD8Tex=c("PDCD1","HAVCR2","LAG3","GZMB","CD8A"),
    CD8Tpexh=c("IFNG","PDCD1","CD8A"),
    CD8Tnaivememory=c("IL7R","CD8A"),
    CD4conv = c("CD4"),
    Treg=c("CD4","FOXP3","CTLA4"),
    CD8Tproliferating=c("CD8A","MKI67")
)
```
# scMatch and singleR
https://github.com/forrest-lab/scMatch.git
```{r}
fread.matrix.norm = function(filename){
    aa =fread(filename)
    rowname = aa[,1] %>%
        unlist %>%
        set_names(NULL)
    aa[,-1] %>% 
        as.matrix %>%
        set_rownames(rowname) %>%
        scale(., center = T, scale = T)
}
fandom.10090 =fread.matrix.norm("unzip -p ~/liulab_home/softwares/scMatch/refDB/FANTOM5/10090_symbol.csv.zip") 
fandom.9606 =fread.matrix.norm("unzip -p ~/liulab_home/softwares/scMatch/refDB/FANTOM5/9606_symbol.csv.zip") 

# load("~/liulab_home/softwares/SingleR/data/cell.type.classification.rda")

fandom.10090.enrich =  calcLeeBulkEnrich(signature = fandom.10090, figFile = ".figs/lee/fandom.10090.pdf", reduce.rownames = T)
sort(rowSums(fandom.10090.enrich$bulk.signature.enrich > 0.4, na.rm=T),decreasing = T)[1:100]
sort(rowSums(fandom.10090.enrich$bulk.signature.enrich < -0.4, na.rm=T),decreasing = T)[1:100]

fandom.9606.enrich =  calcLeeBulkEnrich(signature = fandom.9606, figFile = ".figs/lee/fandom.9606.pdf", reduce.rownames = T)
sort(rowSums(fandom.9606.enrich$bulk.signature.enrich > 0.1, na.rm=T),decreasing = T)[1:20]
sort(rowSums(fandom.9606.enrich$bulk.signature.enrich < -0.3, na.rm=T),decreasing = T)[1:100]
 
cur.order = order(rowSums(abs(fandom.10090.enrich$bulk.signature.enrich) > 0.3, na.rm=T),decreasing = T)
cur.order = order(rowSums((fandom.10090.enrich$bulk.signature.enrich) > 0.25, na.rm=T),decreasing = T)
cur.order = order(rowSums((fandom.10090.enrich$bulk.signature.enrich) < -0.25, na.rm=T),decreasing = T)
combined.enrich = fandom.10090.enrich$enrich.mat[cur.order[1:30],]
rownames(combined.enrich)  = substring(rownames(combined.enrich), 1, 10)
p = pheatmap(t(combined.enrich), cluster_rows=F, cluster_cols=F, fontsize_col = 8, fontsize_row =3, color = col, width = 20, height = 7, filename = ".figs/lee/fandom.10090.half.pdf")

```



## Hypothesis : Are the responders more differentiated 

```{r}
avi.dt = lee.sco@meta.data %>%
    as.data.table %>%
    .[,counts:= colSums(lee.sco@assays$RNA@counts)]%>%
    .[,logcounts:=log(counts)]
avi.dt[,patient.name.response:=paste0(response,patient.name)]
fac.order = c(
    unique(avi.dt[response=="CR"]$patient.name.response), 
    unique(avi.dt[response=="PD"]$patient.name.response), 
    unique(avi.dt[response=="PR"]$patient.name.response))
fac.order = c(fac.order, setdiff(unique(avi.dt$patient.name.response), fac.order))
avi.dt$patient.name.response = factor(avi.dt$patient.name.response, levels = fac.order)    
p = ggplot(avi.dt[(response %in% c("CR", "PR" "PD"))& (Treatment.Cycle=="C1D1") ], aes(x=patient.name.response, y=counts)) + 
    geom_boxplot(outlier.shape = NA) + 
theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
    ylim(c(100,5000))
ggsave(filename = ".figs/lee/counts.pdf", p)

p = ggplot(avi.dt[(response %in% c("CR", "PR", "PD") ) & (Treatment.Cycle=="C1D1")], aes(x=patient.name.response, y=log(counts))) + 
    geom_boxplot(outlier.shape = NA, aes(fill=response)) + 
theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) 

+
   facet_wrap(~seurat_clusters, scales = "free")


p = ggplot(avi.dt[(response %in% c("CR", "PD"))& (Treatment.Cycle=="C4D1") ], aes(x=patient.name.response, y=counts)) + 
    geom_boxplot(outlier.shape = NA) + 
theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +  ylim(c(100,5000))
```

## cell annotation 
```{r}
# lee.sco.sub = lee.sco[,sample.int(ncol(lee.sco), size=10000)]
# seuratfile = extractSeurat(lee.sco)
# sampTab = seuratfile$sampTab
# expDat = seuratfile$expDat
expQDat = lee.sco@assays$RNA@data
sampTab = lee.sco@meta.data
oTab = utils_loadObject("/liulab/asahu/data/single_cell/singleCellNet/human_mouse_genes_Jul_24_2018.rda")
aa = csRenameOrth(expQDat, expTMraw, oTab)
expQueryOrth = aa[['expQuery']]
expTrainOrth = aa[['expTrain']]

cts = c("B cell", "endothelial cell", "erythroblast", "granulocyte", "hematopoietic precursor cell", "late pro-B cell", "limb_mesenchymal", "macrophage", "mammary_basal_cell", "monocyte", "natural killer cell", "T cell", "trachea_epithelial", "trachea_mesenchymal", "alveolar macrophage", "erythrocyte", "hepatocyte", "keratinocyte")

stTM2 = filter(stTM, newAnn %in% cts)
stTM2 = droplevels(stTM2)
rownames(stTM2) = as.vector(stTM2$cell) # filter strips rownames

expTMraw2 = expTrainOrth[,rownames(stTM2)]
stList = splitCommon(stTM2, ncells=100, dLevel="newAnn")
stTrain = stList[[1]]
expTrain = expTMraw2[,rownames(stTrain)]

system.time(class_info2<-scn_train(stTrain = stTrain, expTrain = expTrain, nTopGenes = 10, nRand = 70, nTrees = 1000, nTopGenePairs = 25, dLevel = "newAnn", colName_samp = "cell"))

nqRand = 50 
scn.labels <- scn_predict(class_info2[['cnProc']], expQueryOrth, nrand=nqRand)


scn.out <- get_cate(classRes = scn.labels, sampTab = sampTab, dLevel = "seurat_clusters", sid = "cell.name", nrand = nqRand)
```
## Using singleR 
```{r}
library(SingleR)
lee.sco.sub = lee.sco[,sample.int(ncol(lee.sco), size=30000)]
library(SingleR)
hpca.se <- MonacoImmuneData()
if(all.data){
    pred.hesc.main.fine.all <- SingleR(test = lee.sco@assays$RNA@data, ref = hpca.se, labels = hpca.se$label.main)
    pred.hesc.fine.all <- SingleR(test = lee.sco@assays$RNA@data, ref = hpca.se, labels = hpca.se$label.fine)
    pred.hesc.all = list(pred.hesc.fine.all, pred.hesc.main.fine.all)
    saveRDS(file=".figs/lee/lee.MonacoImmuneData.all.RDS", pred.hesc.all)
    
}
pred.hesc <- SingleR(test = lee.sco.sub@assays$RNA@data, ref = hpca.se, labels = hpca.se$label.fine)
pred.hesc.main <- SingleR(test = lee.sco.sub@assays$RNA@data, ref = hpca.se, labels = hpca.se$label.main)
save(file=".figs/lee/singler.main.Rdata", pred.hesc.main)

lee.sco.curr = lee.sco[,rownames(pred.hesc)]
lee.sco.curr$singleR_cluster = as.factor(pred.hesc$labels)
p = DimPlot(lee.sco.curr, reduction = "umap", group.by = "singleR_cluster",  pt.size = .1, label = TRUE)

aa = lee.sco.sub@meta.data %>%
    as.data.table() %>%
    .[,list(singleR_cluster, seurat_clusters)] %>%
  .[,sc:=as.numeric(seurat_clusters)] %>%
 xtabs(~seurat_clusters+singleR_cluster, data=.)
aa = aa/rowSums(aa)
colx = colorRampPalette(c("white", "darkred"))(300)
p = pheatmap(aa[1:18,], cluster_rows=T, cluster_cols=T, fontsize_col = 8, fontsize_row =8, color = colx, width = 13, height = 7)
# ggsave(file=".figs/lee/lee.clusters.pdf", p, width=15, height=10)

lee.sco.curr = lee.sco[,rownames(pred.hesc.main)]
lee.sco.curr$singleR_cluster = pred.hesc.main$pruned.labels
non.normal.samp = grep("^P[0-9]*", unique(lee.sco.curr$patient.name),value=T)
lee.sco.curr$response.normal = ifelse(lee.sco.curr$patient.name %in% non.normal.samp,  as.character(lee.sco.curr$response), "Normal")
lee.sco.curr$response.normal = factor(lee.sco.curr$response.normal, levels = c("CR", "PR", "PD", "Normal"))
p = DimPlot(lee.sco.curr, reduction = "umap", group.by = "singleR_cluster",  pt.size = .1, label = TRUE) 
p2 = DimPlot(lee.sco.curr, reduction = "umap", group.by = "singleR_cluster", split.by = "singleR_cluster",  pt.size = .1, label = TRUE) 
p3 = DimPlot(lee.sco.curr, reduction = "umap", group.by = "singleR_cluster", split.by = "response.normal",  pt.size = .1, label = TRUE) 

p4 = DimPlot(lee.sco.curr[,lee.sco.curr$Treatment.Cycle=="C1D1"], reduction = "umap", group.by = "singleR_cluster", split.by = "response.normal",  pt.size = .1, label = TRUE) 
lee.sco.temp=lee.sco.curr[,which((lee.sco.curr$Treatment.Cycle %in% c("C4D1")) &(!is.na(lee.sco.curr$singleR_cluster))) ]
p5 = DimPlot(lee.sco.temp, reduction = "umap", group.by = "singleR_cluster", split.by = "response",  pt.size = .1, label = TRUE) 

p = FeaturePlot(lee.sco.curr,reduction = "umap", features = c( "CD3E", "CD3D", "C3DG")) 


```
##singleR 
```{r}
# lymp.deg1 =lymp.deg[padj<1E-1]
lymp.deg1 =lymp.deg.pre 
lymp.deg.mat  = matrix(lymp.deg1$stat, ncol=1)
rownames(lymp.deg.mat) = toupper(lymp.deg1$gene)
lymp.deg.mat2  = cbind(lymp.deg.mat, lymp.deg.mat)
bulk.signature.enrich = calcCorEnrich(aa , lymp.deg.mat2)
    temp = melt(bulk.signature.enrich)
        dt2= cbind(col.dat[temp$Var1,], temp)  %>% as.data.table()
        dt2 = dt2[Var2==1]
dt1 = dt2[,bulk.enrich:=value]
    dt.sum = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.main] %>%
        .[order(enrich.mean, decreasing =T)]
    dt.sum2 = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.fine] %>%
        .[order(enrich.mean, decreasing =T)]
  dt2[label.fine=="B_cell:Germinal_center"]  
  dt2[label.fine=="T_cell:gamma-delta"]
  dt2[label.fine=="MEP"]
  dt2[label.main=="T_cells"]
  
  meta.curr = lee.curr.pre@meta.data %>% as.data.table() %>%
      .[, new.patient.name:=paste(response, patient.name)]
  new.order = c(unique(meta.curr[response=="CR"]$new.patient.name),
  unique(meta.curr[response=="PD"]$new.patient.name))
  meta.curr$new.patient.name = factor( meta.curr$new.patient.name, levels = new.order)
  lee.curr.pre$new.patient.name = meta.curr$new.patient.name
  lymp.deg.pre.sel = lymp.deg.pre[gene %in% VariableFeatures(lee.curr.pre)]
  p=VlnPlot(lee.curr.pre, lymp.deg.pre[1:5]$gene,group.by = 'new.patient.name') 
  
library(SingleR)
rownames(bulk.deg.mat) = toupper(rownames(bulk.deg.mat))
plot.bulk.singleR.enrichment <- function(data.name, dirname, width=20, height = 16, dt2  = NULL) {
    # ref.se <- BlueprintEncodeData(rm.NA = "none")
    dir.create(dirname)
    prefix = sprintf("%s/bulk", dirname)
    if(is.null(dt2)){
        ref.se =  eval(parse(text=paste0(data.name, "()"))) #MonacoImmuneData()
        col.dat = colData(ref.se)
        aa = assay(ref.se) %>%
            t() %>% scale(., center=T, scale=T) %>% t()
        rownames(aa) = toupper(rownames(aa))
        # aa = aa[rownames(aa)%in%toupper(VariableFeatures(lee.sco)),]
        
        bulk.signature.enrich = calcCorEnrich(aa , bulk.deg.mat)
        temp = melt(bulk.signature.enrich)
        dt2= cbind(col.dat[temp$Var1,], temp)  %>% as.data.table()
    }
    # bulk.enrich = bulk.signature.enrich[,"combined"]
    # dt1 = cbind(col.dat, bulk.enrich) %>% as.data.table()
    dt1 = dt2[Var2=="combined"][,bulk.enrich:=value]
    dt.sum = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.main] %>%
        .[order(enrich.mean, decreasing =T)]
    dt.sum2 = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.fine] %>%
        .[order(enrich.mean, decreasing =T)]
    dt1$label.main = factor(dt1$label.main, level=dt.sum$label.main)
    dt1$label.fine = factor(dt1$label.fine, level=dt.sum2$label.fine)
    p1= ggplot(dt1, aes(y=bulk.enrich,x=label.main)) + geom_boxplot(aes(fill=label.main)) +   
        geom_point() + theme_bw()
    ggsave(p1,filename = sprintf("%s.combined.cor.main.pdf", prefix), width=width, height = height)
    p2= ggplot(dt1, aes(y=bulk.enrich,x=label.fine)) + geom_boxplot(aes(fill=label.main)) +   
        geom_point() + theme_bw() + theme(axis.text.x=element_text(angle=90, size = 4))
    ggsave(p2,filename = sprintf("%s.combined.cor.fine.pdf", prefix),  width=width, height = height)
    
    dt2$label.main = factor(dt2$label.main, level=dt.sum$label.main)
    dt2$label.fine = factor(dt2$label.fine, level=dt.sum2$label.fine)
    p3= ggplot(dt2, aes(y=value,x=label.main))  +   
        geom_boxplot(aes(fill=label.main)) + facet_wrap(~Var2, scales = "free") + 
         theme(axis.text.x = element_blank()) +
        theme_bw() 
    ggsave(p3,filename = sprintf("%s.cor3.main.pdf", prefix),  width=width, height = height)
    
    p4= ggplot(dt2, aes(y=value,x=label.fine))  +   
        geom_boxplot() + 
        theme(axis.text.x=element_blank()) +
        facet_wrap(~Var2) + 
    theme_bw()
    ggsave(p4,filename = sprintf("%s.cor3.fine.pdf", prefix),  width=2*width, height = 2*height, limitsize = F)

    dt2
}
singleR.data.names = grep("Data$", ls("package:SingleR"), value=T)
bulk.enrich.list = list()
for (ii in singleR.data.names) {
    bulk.enrich.list[[ii]] = plot.bulk.singleR.enrichment(ii, sprintf(".figs/lee/%s", ii))
}
bulk.enrich.dt = do.call(rbind, bulk.enrich.list)
bulk.enrich.dt[,label.main:=gsub("_cell", label.main, replacement = "-cell", ignore.case = T)]
bulk.enrich.dt[,label.main:=gsub(" cell", label.main, replacement = "-cell", ignore.case = T)]
bulk.enrich.dt[,label.main:=gsub("cells", label.main, replacement = "cell", ignore.case = T)]
bulk.enrich.dt[,label.fine:=gsub("_cell", label.fine, replacement = "-cell", ignore.case = T)]
bulk.enrich.dt[,label.fine:=gsub(" cell", label.fine, replacement = "-cell", ignore.case = T)]
bulk.enrich.dt[,label.fine:=gsub("cells", label.fine, replacement = "cell", ignore.case = T)]


aa = plot.bulk.singleR.enrichment("temp", ".figs/lee/SingleRcombined", width=25, heigh=20, dt2 = bulk.enrich.dt)

combined.bulk.enrich.dt = bulk.enrich.dt[Var2=="combined"] %>%
    .[order(abs(value), decreasing = T)]
quantile.signed = function(val, probs){
    m = mean(val)
    if(m<0) probs = 1-probs
    quantile(val, probs)
}
combined.bulk.enrich.sum = bulk.enrich.dt[Var2!="combined", .(enrich.mean=mean(value),
                                                        enrich.median=quantile(value,probs=0.5),
                                                        enrich.75=quantile(value,probs=0.75),
                                                        enrich.75.signed=quantile.signed(value,probs=0.75),
                                                        enrich.25=quantile(value,probs=0.25),
                                                        label.main=unique(label.main)
                                                        ), by=label.fine]
top_n(combined.bulk.enrich.sum, 10, enrich.75)

bulk.merged = combined.bulk.enrich.dt[,.(enrich.mean=mean(value)),  by=label.fine] %>%
merge(., combined.bulk.enrich.sum, by="label.fine")
bulk.merged.sub = bulk.merged[(abs(enrich.mean.x) > 0.223| abs(enrich.75.signed) > 0.18) & (sign(enrich.mean.x)==sign(enrich.75.signed))]
p1 = ggplot(bulk.merged, aes(y=enrich.mean.x,x=enrich.mean.y))  +   
    geom_point() + 
      geom_text_repel(
    data = bulk.merged.sub,
    aes(label =label.fine, color = factor(label.main)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
        theme_bw() 
ggsave(".figs/lee/bulk.merged.pdf",p1, width=15, height = 10)
# write.table(file="Pancancer.ASNS.drug.association.tab", pan.can.dt, row.names=F, col.names = T, quote=F, sep="\t")


# plot only gamma delta T-cells
p = bulk.enrich.dt[label.fine=="T-cell:gamma-delta"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.gdT.pdf",p)

#Vd2 gd T cells
# Non-Vd2 Vd2 gd T cells
bulk.enrich.dt[grep(" gd ", label.fine, ignore.case = T)]
p1 = bulk.enrich.dt[label.fine=="Vd2 gd T-cell"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.Vd2 gd T cells.pdf",p)

bulk.enrich.dt[grep(" gd ", label.fine, ignore.case = T)]
p2 = bulk.enrich.dt[label.fine=="Non-Vd2 gd T-cell"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.Non-Vd2 gd T cells.pdf",p)


p = bulk.enrich.dt[label.fine=="B-cell:Germinal_center"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.B_cell:Germinal_center.pdf",p)

p = bulk.enrich.dt[label.fine=="CD8+ Tcm"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.CD8+ Tcm.pdf",p)

p = bulk.enrich.dt[label.fine=="T-cell:effector"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.T_cell:effector.pdf",p)

p = bulk.enrich.dt[label.fine=="Pro-B-cell_CD34+"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/bulk.Pro-B-cell_CD34+.pdf",p)
```


## check in cd3- population 

```{r}
lee.cd3neg = readRDS(file="~/liulab_home/data/single_cell/Lee_data/lee.seurat.cd3.neg.RDS")

sel.inx = which( (lee.cd3neg$response %in% c("CR", "PD")) & (lee.cd3neg$Treatment.Cycle=="C1D1"))
cd3Neg.deg.pre = lee.cd3neg[,sel.inx] %>% 
    mydeg()
cd3Neg.deg.mat  = matrix(cd3Neg.deg.pre$stat, ncol=1)
rownames(cd3Neg.deg.mat) = toupper(cd3Neg.deg.pre$gene)
cd3Neg.deg.mat2  = cbind(cd3Neg.deg.mat, cd3Neg.deg.mat)
bulk.signature.enrich = calcCorEnrich(aa , cd3Neg.deg.mat2)
    temp = melt(bulk.signature.enrich)
        dt2= cbind(col.dat[temp$Var1,], temp)  %>% as.data.table()
        dt2 = dt2[Var2==1]
dt1 = dt2[,bulk.enrich:=value]
    dt.sum = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.main] %>%
        .[order(enrich.mean, decreasing =T)]
    dt.sum2 = dt1[,.(enrich.mean=mean(bulk.enrich)), by=label.fine] %>%
        .[order(enrich.mean, decreasing =T)]
  dt2[label.fine=="B_cell:Germinal_center"]  
  dt2[label.fine=="T_cell:gamma-delta"]
  dt2[label.fine=="MEP"]
  dt2[label.main=="T_cells"]

```
## 3 more scRNA dataset
```{r}
export.rdata <- function(rdata) {
    local({load(rdata);environment()})
}

scrna.dataset.list = readRDS("~/liulab_home/data/single_cell/scrna.dataset.deg.RData")
scrna.dataset.list = scrna.dataset.list[!is.na(scrna.dataset.list)]
all.genes = scrna.dataset.list %>% sapply(., function(tt) as.character(tt[["gene"]])) %>% unlist %>% unique
scrna.dataset.deg.mat = matrix(NA, nrow=length(all.genes), ncol=length(scrna.dataset.list),
	dimnames =list(all.genes, names(scrna.dataset.list)))
for (ii in seq(length(scrna.dataset.list))) {
	dt.curr = scrna.dataset.list[[ii]]
	scrna.dataset.deg.mat[dt.curr$gene,ii] = dt.curr$stat
}

scrna.cor.dataset = cor(scrna.dataset.deg.mat, method="spearman", use="pairwise.complete.obs")
which(scrna.cor.dataset < -0.07, arr.ind = T)
corrplot::corrplot(scrna.cor.dataset, is.corr = F, diag = F)
```
## 
Other Single cell data
-1. Correlate with singleR gene expression 
-2. Correlate with  LEE 
-3. Correlate with bulk 
-4. Evaluate the fraction of Tgd in responder vs. non-resp
```{r}

mytheme <- function() {
    theme(panel.border = element_blank(), panel.background = element_blank(), panel.grid.major = element_line(colour = "grey"),
              panel.grid.minor = element_line(colour = "grey"), axis.line = element_line(colour = "black"))
}
combined.singleR.cor.datasets =  function(singleR.list) {
    enrich.dt = do.call(rbind, singleR.list)
    enrich.dt[,label.main:=gsub("_cell", label.main, replacement = "-cell", ignore.case = T)]
    enrich.dt[,label.main:=gsub(" cell", label.main, replacement = "-cell", ignore.case = T)]
    enrich.dt[,label.main:=gsub("cells", label.main, replacement = "cell", ignore.case = T)]
    enrich.dt[,label.fine:=gsub("_cell", label.fine, replacement = "-cell", ignore.case = T)]
    enrich.dt[,label.fine:=gsub(" cell", label.fine, replacement = "-cell", ignore.case = T)]
    enrich.dt[,label.fine:=gsub("cells", label.fine, replacement = "cell", ignore.case = T)]
    enrich.dt
}

plot.mat.singleR.enrichment <- function(mat, data.name, dirname, prefix="", width=20, height = 16, dt2  = NULL) {
    # ref.se <- BlueprintEncodeData(rm.NA = "none")
    dir.create(dirname, recursive = T)
    prefix = sprintf("%s/%s", dirname, prefix)
    if(is.null(dt2)){
        ref.se =  eval(parse(text=paste0(data.name, "()"))) 
        col.dat = colData(ref.se)
        col.dat$Var1= paste(rownames(col.dat),seq(nrow(col.dat)))  
        col.dat = as.data.table(col.dat)
        aa = assay(ref.se) %>%
            t() %>% scale(., center=T, scale=T) %>% t()
        rownames(aa) %<>% toupper
        colnames(aa) = col.dat$Var1
        
        # aa = aa[rownames(aa)%in%toupper(VariableFeatures(lee.sco)),]
        
        mat.signature.enrich = calcCorEnrich(aa, mat)
        temp = melt(mat.signature.enrich) 
        dt2= col.dat[match(as.character(temp$Var1),Var1), .(label.main,label.fine)] %>%
        cbind(., temp) 
    }
    # mat.enrich = mat.signature.enrich[,"combined"]
    # dt1 = cbind(col.dat, mat.enrich) %>% as.data.table()
    
    dt.sum = dt2[,.(enrich.mean=mean(value)), by=label.main] %>%
        .[order(enrich.mean, decreasing =T)]
    dt.sum2 = dt2[,.(enrich.mean=mean(value)), by=label.fine] %>%
        .[order(enrich.mean, decreasing =T)]
    dt2$label.main = factor(dt2$label.main, level=dt.sum$label.main)
    dt2$label.fine = factor(dt2$label.fine, level=dt.sum2$label.fine)
    p3= ggplot(dt2, aes(y=value,x=label.main))  +   
        geom_boxplot(aes(fill=label.main)) + facet_wrap(~Var2, scales = "free") + 
         theme(axis.text.x = element_blank()) +
        mytheme() 
    ggsave(p3,filename = sprintf("%s_cor3.main.pdf", prefix),  width=width, height = height)
    
   p3= ggplot(dt2, aes(y=value,x=label.main))  +   
        geom_boxplot(aes(fill=Var2)) + 
         theme(axis.text.x = element_blank()) +
        mytheme() + 
            theme(axis.text.x = element_text(angle=60, size = 8))
    ggsave(plot = p3,filename = sprintf("%s_cor2.main.pdf", prefix),  width=width, height = height)
    
    p4= ggplot(dt2, aes(y=value,x=label.fine))  +   
        geom_boxplot(aes(fill=label.main)) + 
        theme(axis.text.x=element_blank()) +
       mytheme() + 
        facet_wrap(~Var2) + theme(axis.text.x = element_text(angle=60, size = 5)) 
    ggsave(p4,filename = sprintf("%s_cor3.fine.pdf", prefix),  width=2*width, height = 2*height, limitsize = F)
    
       p4= ggplot(dt2, aes(y=value,x=label.fine))  +   
        geom_boxplot(aes(fill=Var2)) + 
        theme(axis.text.x=element_blank()) +
        mytheme() + 
            theme(axis.text.x = element_text(angle=60, size = 5))
    ggsave(p4,filename = sprintf("%s_cor2.fine.pdf", prefix),  width=width, height = height, limitsize = F)

    dt2
}

rownames(scrna.dataset.deg.mat) %<>% toupper
singleR.data.names = grep("Data$", ls("package:SingleR"), value=T)
scrna.dataset.singleR.list = list()
for (ii in singleR.data.names) {
    scrna.dataset.singleR.list[[ii]] = plot.mat.singleR.enrichment(mat=scrna.dataset.deg.mat, data.name=ii, dirname= ".figs/lee/scrna.dataset", prefix = ii)
}


scrna.dataset.enrich.dt = combined.singleR.cor.datasets(scrna.dataset.singleR.list)

aa = plot.mat.singleR.enrichment(mat=NULL, data.name="temp", dirname=".figs/lee/scrna.dataset", prefix = "singleRcombined", width=25, heigh=20, dt2 = scrna.dataset.enrich.dt)

combined.scrna.dataset.enrich.sum = scrna.dataset.enrich.dt[, .(enrich.mean=mean(value),
                                                        enrich.median=quantile(value,probs=0.5),
                                                        enrich.75=quantile(value,probs=0.75),
                                                        enrich.75.signed=quantile.signed(value,probs=0.75),
                                                        enrich.25=quantile(value,probs=0.25),
                                                        label.main=unique(label.main)
                                                        ), by=label.fine]
top_n(combined.scrna.dataset.enrich.sum, 10, enrich.75)

scrna.dataset.merged = combined.scrna.dataset.enrich.sum
scrna.dataset.merged.sub = scrna.dataset.merged[enrich.mean > 0.1 | enrich.mean < -0.075| abs(enrich.median) > 0.1]
p1 = ggplot(scrna.dataset.merged, aes(y=enrich.mean,x=enrich.median))  +   
    geom_point() + 
      geom_text_repel(
    data = scrna.dataset.merged.sub,
    aes(label =label.fine, color = factor(label.main)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
        theme_bw() 
ggsave(".figs/lee/scrna.dataset/scrna.dataset.merged.pdf",p1, width=15, height = 10)
# write.table(file="Pancancer.ASNS.drug.association.tab", pan.can.dt, row.names=F, col.names = T, quote=F, sep="\t")


# plot only gamma delta T-cells
p = scrna.dataset.enrich.dt[label.fine=="T-cell:gamma-delta"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/scrna.dataset/scrna.dataset.gdT.pdf",p)
## GSE12909 has a gamma-delta correlation. 



scrna.bulk.merged.annotation = merge(scrna.dataset.merged, bulk.merged, by="label.fine", suffixes = c(".scrna", ".bulk"))  
scrna.bulk.dataset.merged.sub = scrna.bulk.merged.annotation[enrich.mean > 0.11 | enrich.mean < -0.074 | abs(enrich.mean.x) >= 0.223]
p1 = ggplot(scrna.bulk.merged.annotation, aes(y=enrich.mean,x=enrich.mean.x))  +   
    geom_point() + 
      geom_text_repel(
    data = scrna.bulk.dataset.merged.sub,
    aes(label =label.fine, color = factor(label.main.scrna)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
    xlab("bulk") + ylab("scrna") + 
        theme_bw() 
ggsave(".figs/lee/scrna.dataset/scrna.bulk.merged.annotation.pdf",p1, width=15, height = 10)
```

## Check gamma-delta in ICB cohorts
```{r}
"HumanPrimaryCellAtlasData"
ref.se =  eval(parse(text=paste0(data.name, "()"))) #MonacoImmuneData()
col.dat = colData(ref.se)
aa = assay(ref.se) %>%
    t() %>% scale(., center=T, scale=T) %>% t()
rownames(aa) = toupper(rownames(aa))
Tgd.exp = aa[,which(col.dat$label.fine=="T_cell:gamma-delta")]
considered.genes = intersect(rownames(Tgd.exp), bulk.deg.dt[type=="combined"][P<1E-3]$gene)

icb.exp = fread("~/liulab_home/data/ssgsea/xiaoman/icb/expression/Mariathasan2018_PDL1_Bladder_RNASeq_mUC")
icb.followup =  fread("~/liulab_home/data/ssgsea/xiaoman/icb/follow_up/Mariathasan2018_PDL1_Bladder_RNASeq_mUC")
icb.exp.mat = icb.exp[,-1] %>% as.matrix %>%
    set_rownames(icb.exp$Symbol) %>% t() %>%
     scale(., center=T, scale=T) %>% t()
genes.curr = intersect(rownames(icb.exp.mat), considered.genes)
aa = cor(icb.exp.mat[genes.curr,], Tgd.exp[genes.curr,], method="spearman", use="pairwise.complete.obs")
genes.curr = intersect(rownames(icb.exp.mat), rownames(Tgd.exp))
aa = cor(icb.exp.mat[genes.curr,], Tgd.exp[genes.curr,], method="spearman", use="pairwise.complete.obs")
pROC::roc(icb.followup$Response, aa[,1])
bb = pROC::roc(icb.followup$Response, aa[,2])
plot(bb)
pROC::roc(icb.followup$Response, icb.exp.mat["TRDV3",])
```
## Analyze the tcga survival genes 
```{r}
tcga.surv.genes = readRDS("~/liulab_home/data/tcga/tcga.cox.genes.Rds")
tcga.surv.deg.mat = lapply(tcga.surv.genes, function(tt) tt$z) %>%
    do.call(cbind, .) %>%
    set_rownames(tcga.surv.genes[[1]]$genes)

rownames(tcga.surv.deg.mat) %<>% toupper
singleR.data.names = grep("Data$", ls("package:SingleR"), value=T)
tcga.surv.singleR.list = list()
for (ii in singleR.data.names) {
    tcga.surv.singleR.list[[ii]] = plot.mat.singleR.enrichment(mat=tcga.surv.deg.mat, data.name=ii, dirname= ".figs/lee/tcga.surv", prefix = ii)
}
tcga.surv.enrich.dt = combined.singleR.cor.datasets(tcga.surv.singleR.list)

aa = plot.mat.singleR.enrichment(mat=NULL, data.name="temp", dirname=".figs/lee/tcga.surv", prefix = "singleRcombined", width=25, heigh=20, dt2 = tcga.surv.enrich.dt)

combined.tcga.surv.enrich.sum = tcga.surv.enrich.dt[, .(enrich.mean=mean(value),
                                                        enrich.median=quantile(value,probs=0.5),
                                                        enrich.75=quantile(value,probs=0.75),
                                                        enrich.75.signed=quantile.signed(value,probs=0.75),
                                                        enrich.25=quantile(value,probs=0.25),
                                                        label.main=unique(label.main)
                                                        ), by=label.fine]

tcga.surv.merged = combined.tcga.surv.enrich.sum
tcga.surv.merged.sub = tcga.surv.merged[enrich.mean > 0.3| enrich.mean < -.3| abs(enrich.median) > 0.2]

p1 = ggplot(tcga.surv.merged, aes(y=enrich.mean,x=enrich.median))  +   
    geom_point() + 
      geom_text_repel(
    data = tcga.surv.merged.sub,
    aes(label =label.fine, color = factor(label.main)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
        theme_bw() 
ggsave(".figs/lee/tcga.surv/tcga.surv.merged.pdf",p1, width=15, height = 10)
# write.table(file="Pancancer.ASNS.drug.association.tab", pan.can.dt, row.names=F, col.names = T, quote=F, sep="\t")


# plot only gamma delta T-cells
p = tcga.surv.enrich.dt[label.fine=="T-cell:gamma-delta"] %>%
ggplot(., aes(y=value,x=Var2))  +   
        geom_boxplot(aes(fill=label.fine)) +
         theme(axis.text.x = element_blank()) +
        theme_bw() + theme(axis.text.x=element_text(angle=90, size=6))
ggsave(".figs/lee/tcga.surv/tcga.surv.gdT.pdf",p)
## GSE12909 has a gamma-delta correlation. 



tcga.bulk.merged.annotation = merge(tcga.surv.merged, bulk.merged, by="label.fine", suffixes = c(".tcga", ".bulk"))  
tcga.bulk.dataset.merged.sub = tcga.bulk.merged.annotation[enrich.mean > 0.2 | enrich.mean < -0.2 | abs(enrich.mean.x) >= 0.223]
p1 = ggplot(tcga.bulk.merged.annotation, aes(y=enrich.mean,x=enrich.mean.x))  +   
    geom_point() + 
      geom_text_repel(
    data = tcga.bulk.dataset.merged.sub,
    aes(label =label.fine, color = factor(label.main.tcga)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
    xlab("bulk") + ylab("tcga") + 
        theme_bw() 
ggsave(".figs/lee/tcga.surv/tcga.surv.bulk.merged.annotation.pdf",p1, width=15, height = 10)


tcga.bulk.dataset.merged.sub = tcga.bulk.merged.annotation[(enrich.mean > 0.05  & enrich.mean.x > 0.05)  | (enrich.mean <  -0.05  & enrich.mean.x <  -0.05)]
p1 = ggplot(tcga.bulk.merged.annotation, aes(y=enrich.mean,x=enrich.mean.x))  +   
    geom_point() + 
      geom_text_repel(
    data = tcga.bulk.dataset.merged.sub,
    aes(label =label.fine, color = factor(label.main.tcga)),
    size = 3,
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines")
  ) +
    xlab("bulk") + ylab("tcga") + 
        theme_bw() 
ggsave(".figs/lee/tcga.surv/tcga.surv.bulk.merged.annotation.oppo.pdf",p1, width=15, height = 10)
```

## Analyze the seurat annotation GSE145281 data 
```{r}
gse145281 = readRDS("~/liulab_home/data/single_cell/GSE145281/seurat.RDS")
singleR.annot = readRDS("~/liulab_home/data/single_cell/GSE145281/singleR.annotations.RDS")
gse145281$HumanPrimaryCellAtlasData.fine=singleR.annot$HumanPrimaryCellAtlasData.fine[colnames(gse145281),]$pruned.labels
gse145281$HumanPrimaryCellAtlasData.main=singleR.annot$HumanPrimaryCellAtlasData.main[colnames(gse145281),]$pruned.labels
gse145281$MonacoImmuneData.fine=singleR.annot$MonacoImmuneData.fine[colnames(gse145281),]$pruned.labels
gse145281$MonacoImmuneData.main=singleR.annot$MonacoImmuneData.main[colnames(gse145281),]$pruned.labels
saveRDS(file = "~/liulab_home/data/single_cell/GSE145281/seurat.RDS", gse145281)
p = DimPlot(gse145281, reduction = "umap", group.by = "HumanPrimaryCellAtlasData.fine", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/HumanPrimaryCellAtlasData.fine.pdf", p, width=15, height=10)
p = DimPlot(gse145281, reduction = "umap", group.by = "MonacoImmuneData.fine", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/MonacoImmuneData.fine.pdf", p, width=15, height=10)

p = DimPlot(gse145281, reduction = "umap", group.by = "MonacoImmuneData.fine", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/MonacoImmuneData.fine.pdf", p, width=15, height=10)
p = DimPlot(gse145281, reduction = "umap", group.by = "seurat_clusters", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/seurat_clusters.pdf", p, width=15, height=10)


gse145281.lymp <- gse145281[,gse145281$seurat_clusters %in% c(0,16,13,2,7,10)] %>% 
    RunUMAP(reduction = "harmony", dims = 1:20) %>% 
    FindNeighbors(reduction = "harmony", dims = 1:20) %>% 
    FindClusters(resolution = 0.5) %>% 
    identity()

gse145281.lymp %<>% FindVariableFeatures(selection.method = "vst", nfeatures = 2000) 

gse145281.lymp %<>%  RunPCA(pc.genes = VariableFeatures(gse145281.lymp), npcs = 50, verbose = FALSE) %>%
    ScaleData(verbose = FALSE) 

gse145281.lymp %<>% RunUMAP(reduction = "pca", dims=1:50, reduction.name="myumap")
     
gse145281.lymp = RunTSNE(gse145281.lymp, reduction = "pca", dims=1:50, reduction.name="mytsne")





p = DimPlot(gse145281.lymp, reduction = "mytsne", group.by = "HumanPrimaryCellAtlasData.fine", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/lymph.HumanPrimaryCellAtlasData.fine.pdf", p, width=15, height=10)

p = DimPlot(gse145281.lymp, reduction = "mytsne", group.by = "HumanPrimaryCellAtlasData.main", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/lymph.HumanPrimaryCellAtlasData.main.pdf", p, width=15, height=10)

p = DimPlot(gse145281.lymp, reduction = "mytsne", group.by = "MonacoImmuneData.fine", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/lymph.MonacoImmuneData.fine.pdf", p, width=15, height=10)
p = DimPlot(gse145281.lymp, reduction = "mytsne", group.by = "seurat_clusters", pt.size = .1, label = TRUE) +  NoLegend() 
ggsave(file=".figs/gse145281/lymph.seurat_clusters.pdf", p, width=15, height=10)


p = FeaturePlot(gse145281.lymp, reduction = "mytsne", features = c( "CD3E",  "CD4", "CD8A", "NKG7", 
                                           "CD79A", "LYZ", "CD14", "FCGR3A", 
                                           "FCER1A", "PRF1", "GLNY", "CD27",
                                           "TRG-AS1",  "GZMB", "GZMK", "PRF1")) 
ggsave(file=".figs/gse145281/lymph.features.pdf", p, width=15, height=10)

```

```{r}

exp = gse145281.lymp@assays$RNA@data %>% as.matrix 
exp.norm = t(exp) %>% scale(.,scale = T, center = T) %>% t()

rownames(exp.norm) %<>% toupper

"HumanPrimaryCellAtlasData"
ref.se =  eval(parse(text=paste0(data.name, "()"))) #MonacoImmuneData()
col.dat = colData(ref.se)
aa = assay(ref.se) %>%
    t() %>% scale(., center=T, scale=T) %>% t()
rownames(aa) = toupper(rownames(aa))
T_NK_exp = aa[,which(col.dat$label.main %in% c("T_cells", "NK_cell"))]
col.dat.sub = col.dat[which(col.dat$label.main %in% c("T_cells", "NK_cell")),] %>% 
    as.data.table %>%
    .[,ID:=seq(.N)]

genes.curr = intersect(VariableFeatures(gse145281.lymp), intersect(rownames(exp.norm), rownames(aa)))
aa1 = myFastCor(exp.norm[genes.curr,], T_NK_exp[genes.curr,], method="spearman",  num =0)
exp.ref.cor = lapply(unique(col.dat.sub$label.fine), function(var){
    bb = aa1[,which(col.dat.sub$label.fine == var),drop=F]
    if(ncol(bb) > 1) bb = rowMeans(bb,na.rm=T)
    bb
}) %>% do.call(cbind, .) %>% set_colnames(unique(col.dat.sub$label.fine)) %>% as.data.frame()

gse145281.lymp@meta.data %<>% cbind(., exp.ref.cor)

p = FeaturePlot(gse145281.lymp, reduction = "newumap", features = colnames(exp.ref.cor)) 
ggsave(file=".figs/gse145281/lymph.HumanPrimaryCellAtlasData.fine.scores.pdf", p, width=15, height=10)
```

## Perform differential expression analysis between NK and gdT; T-cell and gdT
 Then calculate correlation of differential expression 

```{r}
bulk.deg = function(exp.curr, group1, group2) {
    exp.g1 = exp.curr[,group1]
    exp.g2 = exp.curr[,group2]
    xx = matrixTests::row_t_welch(exp.g1, exp.g2) 
    xx%>%as.data.table %>% 
        .[,gene:=toupper(rownames(xx))]
}
ref.exp = assay(ref.se) 

gdt.vs.nk = bulk.deg(ref.exp, 
                     group1 = rownames(col.dat)[col.dat$label.fine=="T_cell:gamma-delta"], 
                     group2 = rownames(col.dat)[col.dat$label.main=="NK_cell"]) 
gdt.vs.t = bulk.deg(ref.exp, 
                     group1 = rownames(col.dat)[col.dat.sub$label.fine=="T_cell:gamma-delta"], 
                     group2 = rownames(col.dat)[(col.dat.sub$label.main=="T_cells") & (col.dat.sub$label.fine!="T_cell:gamma-delta")])       

genes.curr = intersect(rownames(exp.norm), gdt.vs.nk[order(pvalue)[1:100]]$gene)
gse145281.lymp$gdt.vs.nk = cor(exp.norm[genes.curr,], gdt.vs.nk[match(genes.curr, gene)]$statistic, method="spearman",  use="pairwise.complete.obs")
genes.curr = intersect(rownames(exp.norm), gdt.vs.t[order(pvalue)[1:1000]]$gene)
gse145281.lymp$gdt.vs.t = cor(exp.norm[genes.curr,], gdt.vs.t[match(genes.curr, gene)]$statistic, method="spearman",  use="pairwise.complete.obs")
 

p = FeaturePlot(gse145281.lymp, reduction = "newumap", features = c("gdt.vs.nk", "gdt.vs.t")) 
ggsave(file=".figs/gse145281/lymph.HumanPrimaryCellAtlasData.fine.gdt.scores.pdf", p, width=15, height=10)

p=VlnPlot(gse145281.lymp, c("gdt.vs.nk", "gdt.vs.t"), group.by = 'patient') 

```



